// @swimlane/ngx-ui/esm2015/lib/directives/dbl-click-copy/dbl-click-copy.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Output, EventEmitter, HostListener, ElementRef } from '@angular/core';
// tslint:disable-next-line:directive-selector
let DblClickCopyDirective = class DblClickCopyDirective {
    constructor(element) {
        this.element = element;
        this.onCopy = new EventEmitter();
    }
    onDblClick() {
        const selection = getSelection();
        const range = document.createRange();
        range.selectNodeContents(this.element.nativeElement);
        selection.removeAllRanges();
        selection.addRange(range);
        document.execCommand('copy');
        this.onCopy.emit(range);
    }
};
DblClickCopyDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output(),
    __metadata("design:type", Object)
], DblClickCopyDirective.prototype, "onCopy", void 0);
__decorate([
    HostListener('dblclick'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], DblClickCopyDirective.prototype, "onDblClick", null);
DblClickCopyDirective = __decorate([
    Directive({
        selector: '[dbl-click-copy]',
        host: {
            title: 'Double click to copy to clipboard'
        }
    }),
    __metadata("design:paramtypes", [ElementRef])
], DblClickCopyDirective);
export { DblClickCopyDirective };
