// @wkoza/ngx-upload/esm5/ngx-upload.module.js
import * as tslib_1 from "tslib";
import { InjectionToken, isDevMode, NgModule, Optional, SkipSelf } from '@angular/core';
import { NgxDragAndDropDirective } from './directives/dropzone.directive';
import { NgxThumbnailDirective } from './directives/thumbnail.directive';
import { NgxInputFileDirective } from './directives/inputfile.directive';
import { NGX_DROP_TARGET_OPTIONS, NGX_LOGGER_OPTIONS, ngxDropTargetOptions, ngxloggerOptions } from './utils/configuration.model';
import { ConsoleLogger, NgxUploadLogger, NoOpLogger } from './utils/logger.model';
import { InputfileComponent } from './components/inputfile.component';
import { HttpClientModule } from '@angular/common/http';
var ngxDeclarations = [
    NgxDragAndDropDirective, NgxThumbnailDirective, InputfileComponent, NgxInputFileDirective
];
/**
 * Factory associated with internal logger
 * @param options
 * @returns {any}
 * @private
 */
export function _loggerFactory(options) {
    var enabled = options.enabled ? options.enabled : isDevMode();
    if (enabled) {
        var _console = typeof console === 'object' ? console : {};
        var debug = options.debug ? options.debug : true;
        return new ConsoleLogger(_console, debug);
    }
    return new NoOpLogger();
}
export var NGX_UPLOAD_ROOT_GUARD = new InjectionToken('Internal forRoot Guard');
export function createNgxUploadRootGuard(options) {
    if (options) {
        throw new TypeError('NgxUploadModule.forRoot() is called twice.');
    }
    return 'guarded';
}
var NgxUploadModule = /** @class */ (function () {
    function NgxUploadModule() {
    }
    NgxUploadModule_1 = NgxUploadModule;
    NgxUploadModule.forRoot = function (dropTargetOptions, loggerOptions) {
        return {
            ngModule: NgxUploadModule_1,
            providers: [
                { provide: NGX_LOGGER_OPTIONS, useValue: (loggerOptions) ? loggerOptions : ngxloggerOptions },
                {
                    provide: NGX_DROP_TARGET_OPTIONS,
                    useValue: (dropTargetOptions) ? dropTargetOptions : ngxDropTargetOptions
                },
                {
                    provide: NgxUploadLogger,
                    useFactory: _loggerFactory,
                    deps: [NGX_LOGGER_OPTIONS]
                },
                {
                    provide: NGX_UPLOAD_ROOT_GUARD,
                    useFactory: createNgxUploadRootGuard,
                    deps: [[NGX_LOGGER_OPTIONS, new Optional(), new SkipSelf()]]
                }
            ]
        };
    };
    ;
    var NgxUploadModule_1;
    NgxUploadModule = NgxUploadModule_1 = tslib_1.__decorate([
        NgModule({
            declarations: tslib_1.__spread(ngxDeclarations),
            exports: tslib_1.__spread(ngxDeclarations),
            imports: [HttpClientModule],
            entryComponents: [InputfileComponent]
        })
    ], NgxUploadModule);
    return NgxUploadModule;
}());
export { NgxUploadModule };
