// @swimlane/ngx-ui/esm5/lib/components/tree/tree-node.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, EventEmitter, Output, ViewEncapsulation, TemplateRef, ChangeDetectionStrategy } from '@angular/core';
var TreeNodeComponent = /** @class */ (function () {
    function TreeNodeComponent() {
        this.icons = {
            collapse: 'icon-tree-collapse',
            expand: 'icon-tree-expand'
        };
        this.activate = new EventEmitter();
        this.deactivate = new EventEmitter();
        this.selectNode = new EventEmitter();
        // backwards compatibility. Use selectNode
        this.select = this.selectNode;
        this.expand = new EventEmitter();
        this.collapse = new EventEmitter();
    }
    TreeNodeComponent.prototype.ngOnChanges = function () {
        this.data = {
            $implicit: this.node,
            label: this.label,
            children: this.children,
            model: this.model
        };
    };
    TreeNodeComponent.prototype.onExpandClick = function (event) {
        if (this.disabled || !this.expandable)
            return;
        event.stopPropagation();
        this.expanded = !this.expanded;
        if (this.expanded) {
            this.expand.emit(this.data);
        }
        else {
            this.collapse.emit(this.data);
        }
    };
    TreeNodeComponent.prototype.onClick = function () {
        if (!this.selectable || this.disabled)
            return;
        this.selectNode.emit(this.data);
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], TreeNodeComponent.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "model", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "node", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], TreeNodeComponent.prototype, "children", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], TreeNodeComponent.prototype, "disabled", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], TreeNodeComponent.prototype, "expandable", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], TreeNodeComponent.prototype, "expanded", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], TreeNodeComponent.prototype, "selectable", void 0);
    __decorate([
        Input(),
        __metadata("design:type", TemplateRef)
    ], TreeNodeComponent.prototype, "template", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "icons", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "activate", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "deactivate", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "selectNode", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "select", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "expand", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TreeNodeComponent.prototype, "collapse", void 0);
    TreeNodeComponent = __decorate([
        Component({
            exportAs: 'ngxTreeNode',
            selector: 'ngx-tree-node',
            template: "<li\n  class=\"ngx-tree-node\"\n  [class.selectable]=\"selectable\"\n  (focus)=\"activate.emit(this.data)\"\n  (blur)=\"deactivate.emit(this.data)\"\n  tabindex=\"-1\"\n>\n  <span\n    *ngIf=\"expandable\"\n    class=\"ngx-expander\"\n    (click)=\"onExpandClick($event)\"\n    [class.disabled]=\"disabled\"\n    [ngClass]=\"expanded ? icons.collapse : icons.expand\"\n    >\n  </span>\n  <span *ngIf=\"!template\" [innerHTML]=\"label\" [class.disabled]=\"disabled\" class=\"ngx-node-label\" (click)=\"onClick()\"> </span>\n  <ng-template *ngIf=\"template\" [ngTemplateOutlet]=\"template\" [ngTemplateOutletContext]=\"data\"> </ng-template>  \n  <ng-content *ngIf=\"expanded\"></ng-content>\n  <ngx-tree\n    *ngIf=\"children?.length && expandable && expanded\"\n    class=\"ngx-sub-tree\"\n    [nodes]=\"children\"\n    [template]=\"template\"\n    [icons]=\"icons\"\n    (selectNode)=\"selectNode.emit($event)\"\n  >\n  </ngx-tree>\n</li>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], TreeNodeComponent);
    return TreeNodeComponent;
}());
export { TreeNodeComponent };
