// ngx-awesome-uploader/esm5/lib/file-picker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { of } from 'rxjs';
import { DomSanitizer } from '@angular/platform-browser';
var FilePickerService = /** @class */ (function () {
    function FilePickerService(sanitizer) {
        this.sanitizer = sanitizer;
    }
    /**
     * @param {?} formData
     * @return {?}
     */
    FilePickerService.prototype.mockUploadFile = /**
     * @param {?} formData
     * @return {?}
     */
    function (formData) {
        /** @type {?} */
        var event = new CustomEvent('customevent', {
            detail: {
                type: 'UploadProgreess'
            }
        });
        return of(event.detail);
    };
    /**
     * @param {?} file
     * @return {?}
     */
    FilePickerService.prototype.createSafeUrl = /**
     * @param {?} file
     * @return {?}
     */
    function (file) {
        if (((/** @type {?} */ (window))).UPLOADER_TEST_MODE) {
            return;
        }
        try {
            /** @type {?} */
            var url = window.URL.createObjectURL(file);
            /** @type {?} */
            var safeUrl = this.sanitizer.bypassSecurityTrustResourceUrl(url);
            return safeUrl;
        }
        catch (er) {
            console.log(er);
        }
    };
    FilePickerService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FilePickerService.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    return FilePickerService;
}());
export { FilePickerService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    FilePickerService.prototype.sanitizer;
}
