// @swimlane/ngx-ui/esm5/lib/directives/resize-observer/resize-observer.directive.js
import { __decorate, __metadata, __values } from "tslib";
import { Directive, Output, EventEmitter, OnInit, OnDestroy, ElementRef } from '@angular/core';
import ResizeObserver from 'resize-observer-polyfill';
var ResizeObserverDirective = /** @class */ (function () {
    function ResizeObserverDirective(_el) {
        this._el = _el;
        this.resize = new EventEmitter();
    }
    ResizeObserverDirective.prototype.ngOnInit = function () {
        var _this = this;
        this._observer = new ResizeObserver(function (entries) {
            var e_1, _a;
            try {
                for (var entries_1 = __values(entries), entries_1_1 = entries_1.next(); !entries_1_1.done; entries_1_1 = entries_1.next()) {
                    var entry = entries_1_1.value;
                    _this.onResize(entry.contentRect);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (entries_1_1 && !entries_1_1.done && (_a = entries_1.return)) _a.call(entries_1);
                }
                finally { if (e_1) throw e_1.error; }
            }
        });
        this._observer.observe(this._el.nativeElement);
    };
    ResizeObserverDirective.prototype.ngOnDestroy = function () {
        this._observer.unobserve(this._el.nativeElement);
    };
    ResizeObserverDirective.prototype.onResize = function (e) {
        var _this = this;
        if (this._timer) {
            clearTimeout(this._timer);
            this._timer = undefined;
        }
        this._timer = setTimeout(function () { return _this.resize.emit(e); }, 100);
    };
    ResizeObserverDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output('resizeObserver'),
        __metadata("design:type", Object)
    ], ResizeObserverDirective.prototype, "resize", void 0);
    ResizeObserverDirective = __decorate([
        Directive({
            exportAs: 'resizeObserver',
            selector: '[resizeObserver]'
        }),
        __metadata("design:paramtypes", [ElementRef])
    ], ResizeObserverDirective);
    return ResizeObserverDirective;
}());
export { ResizeObserverDirective };
