// @swimlane/ngx-ui/esm5/lib/components/toolbar/toolbar.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, ViewChild, ViewEncapsulation, Renderer2, ElementRef, ChangeDetectionStrategy } from '@angular/core';
import { ToolbarTitleDirective } from './toolbar-title.directive';
import { ToolbarContentDirective } from './toolbar-content.directive';
var ToolbarComponent = /** @class */ (function () {
    function ToolbarComponent(renderer, elRef) {
        this.renderer = renderer;
        this.elRef = elRef;
        this.menu = [];
        this.menuClick = new EventEmitter();
    }
    Object.defineProperty(ToolbarComponent.prototype, "toolbarItems", {
        get: function () {
            return this.menu.filter(function (m) {
                return !m.dropdown;
            });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToolbarComponent.prototype, "dropdownItems", {
        get: function () {
            return this.menu.filter(function (m) {
                return m.dropdown;
            });
        },
        enumerable: true,
        configurable: true
    });
    ToolbarComponent.prototype.onMenuClicked = function (item, $event) {
        if (item.click) {
            item.click($event);
        }
    };
    ToolbarComponent.prototype.ngOnInit = function () {
        // backwards compatibility
        if (this.title) {
            this.mainTitle = this.title;
            this.renderer.removeAttribute(this.elRef.nativeElement, 'title');
        }
    };
    ToolbarComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], ToolbarComponent.prototype, "title", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], ToolbarComponent.prototype, "mainTitle", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], ToolbarComponent.prototype, "subtitle", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], ToolbarComponent.prototype, "menu", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], ToolbarComponent.prototype, "menuClick", void 0);
    __decorate([
        ViewChild(ToolbarTitleDirective),
        __metadata("design:type", ToolbarTitleDirective)
    ], ToolbarComponent.prototype, "toolbarTitle", void 0);
    __decorate([
        ViewChild(ToolbarContentDirective),
        __metadata("design:type", ToolbarContentDirective)
    ], ToolbarComponent.prototype, "toolbarContent", void 0);
    ToolbarComponent = __decorate([
        Component({
            selector: 'ngx-toolbar',
            template: "<header class=\"flex-container\" fxLayout=\"row\" fxLayoutWrap=\"nowrap\" fxFill fxLayoutGap=\"5px\">\n  <div class=\"ngx-toolbar-title-col\" fxFlex>\n    <ng-content *ngIf=\"!mainTitle\" select=\"ngx-toolbar-title\"></ng-content>\n    <h2 class=\"ngx-toolbar-title\" *ngIf=\"mainTitle\">\n      {{ mainTitle }}\n      <small *ngIf=\"subtitle\">{{ subtitle }}</small>\n    </h2>\n  </div>\n  <div class=\"ngx-toolbar-content-col\" fxFlex>\n    <ng-content *ngIf=\"!menu.length\" select=\"ngx-toolbar-content\"></ng-content>\n    <ul class=\"horizontal-list ngx-toolbar-menu\" *ngIf=\"menu.length\">\n      <li *ngFor=\"let item of toolbarItems\">\n        <button type=\"button\" [disabled]=\"item.disabled\" (click)=\"onMenuClicked(item, $event)\">\n          {{ item.label }}\n        </button>\n      </li>\n      <li *ngIf=\"dropdownItems.length\">\n        <ngx-dropdown>\n          <ngx-dropdown-toggle>\n            <button type=\"button\">\n              ...\n            </button>\n          </ngx-dropdown-toggle>\n          <ngx-dropdown-menu class=\"align-right\">\n            <ul class=\"vertical-list\">\n              <li *ngFor=\"let item of dropdownItems\">\n                <button type=\"button\" (click)=\"onMenuClicked(item, $event)\">\n                  {{ item.label }}\n                </button>\n              </li>\n            </ul>\n          </ngx-dropdown-menu>\n        </ngx-dropdown>\n      </li>\n    </ul>\n  </div>\n</header>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                class: 'ngx-toolbar'
            },
            styles: [".ngx-toolbar{display:block;height:50px;line-height:50px;color:#a0aabe;background:#1c2029;width:100%;font-size:.8rem;padding:0 20px!important}.ngx-toolbar a{color:#a0aabe}.ngx-toolbar .ngx-toolbar-title-col{overflow:hidden}.ngx-toolbar .ngx-toolbar-title{margin:0;font-weight:500;text-overflow:ellipsis;overflow:hidden;width:100%;white-space:nowrap}.ngx-toolbar .ngx-toolbar-title small{color:#5a6884;font-size:.9rem;line-height:50px;vertical-align:top}.ngx-toolbar .ngx-toolbar-menu button{color:#a0aabe}.ngx-toolbar .ngx-toolbar-content-col{text-align:right}.ngx-toolbar.page-toolbar{box-shadow:0 1px 5px 0 rgba(0,0,0,.2),0 2px 2px 0 rgba(0,0,0,.14),0 3px 1px -2px rgba(0,0,0,.12)}"]
        }),
        __metadata("design:paramtypes", [Renderer2, ElementRef])
    ], ToolbarComponent);
    return ToolbarComponent;
}());
export { ToolbarComponent };
