// @swimlane/ngx-ui/esm2015/lib/components/split/split-area.directive.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Optional, Self, Input } from '@angular/core';
import { DefaultFlexDirective, validateBasis } from '@angular/flex-layout';
let SplitAreaDirective = class SplitAreaDirective {
    constructor(flexDirective) {
        this.flexDirective = flexDirective;
        this.overflow = 'hidden';
    }
    get fxFlexFill() {
        return this.fxFlex === '';
    }
    ngOnChanges() {
        this.currentFlexBasis = this.initialFlexBasis = this.getCurrentFlexParts();
    }
    updateStyle(flexBasis) {
        const flex = this.flexDirective;
        if (typeof flexBasis === 'undefined') {
            flexBasis = flex.activatedValue || '';
        }
        if (typeof flexBasis === 'number') {
            flexBasis = this.isPercent() ? `${flexBasis}%` : `${flexBasis}px`;
        }
        if (flexBasis.indexOf(' ') < 0) {
            const grow = flex.grow;
            const shrink = flex.shrink;
            this.currentFlexBasis = [grow, shrink, flexBasis];
            flexBasis = this.currentFlexBasis.join(' ');
        }
        else {
            this.currentFlexBasis = flexBasis.split(' ');
        }
        flex.activatedValue = this.currentFlexBasis[2];
    }
    isPercent(basis) {
        if (!basis) {
            const flex = this.flexDirective;
            basis = flex.activatedValue || '1 1 1e-9px';
        }
        const hasCalc = String(basis).indexOf('calc') > -1;
        return String(basis).indexOf('%') > -1 && !hasCalc;
    }
    getCurrentFlexParts() {
        const flex = this.flexDirective;
        const basis = (flex && flex.activatedValue) || '1 1 1e-9px';
        return validateBasis(String(basis).replace(';', ''), flex.grow, flex.shrink);
    }
};
SplitAreaDirective.ctorParameters = () => [
    { type: DefaultFlexDirective, decorators: [{ type: Optional }, { type: Self }] }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], SplitAreaDirective.prototype, "minBasis", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SplitAreaDirective.prototype, "maxBasis", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SplitAreaDirective.prototype, "fxFlex", void 0);
SplitAreaDirective = __decorate([
    Directive({
        exportAs: 'ngxSplitArea',
        selector: '[ngxSplitArea]',
        host: {
            class: 'ngx-split-area',
            '[style.overflow]': 'overflow'
        }
    }),
    __param(0, Optional()),
    __param(0, Self()),
    __metadata("design:paramtypes", [DefaultFlexDirective])
], SplitAreaDirective);
export { SplitAreaDirective };
