// @swimlane/ngx-ui/esm5/lib/components/dialog/alert/alert.service.js
import { __decorate, __extends, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { InjectionService } from '../../../services/injection/injection.service';
import { OverlayService } from '../../overlay/overlay.service';
import { DialogService } from '../dialog.service';
import { AlertComponent } from './alert.component';
import { AlertTypes } from './alert-types.enum';
import { AlertStyles } from './alert-styles.enum';
var AlertService = /** @class */ (function (_super) {
    __extends(AlertService, _super);
    function AlertService(injectionService, overlayService) {
        var _a;
        var _this = _super.call(this, injectionService, overlayService) || this;
        _this.injectionService = injectionService;
        _this.overlayService = overlayService;
        _this.defaults = {
            inputs: {
                zIndex: 991,
                closeOnBlur: false,
                closeOnEscape: false,
                closeButton: false,
                showOverlay: true,
                visible: true
            }
        };
        _this.type = AlertComponent;
        _this.clsMap = (_a = {},
            _a[AlertStyles.Danger] = 'ngx-alert-danger',
            _a[AlertStyles.Warning] = 'ngx-alert-warning',
            _a[AlertStyles.Info] = 'ngx-alert-info',
            _a);
        return _this;
    }
    AlertService.prototype.alert = function (options) {
        return this.createDialog(options, AlertTypes.Alert);
    };
    AlertService.prototype.confirm = function (options) {
        return this.createDialog(options, AlertTypes.Confirm);
    };
    AlertService.prototype.prompt = function (options) {
        return this.createDialog(options, AlertTypes.Prompt);
    };
    AlertService.prototype.createDialog = function (options, type) {
        var subject = new Subject();
        var title = options.title, content = options.content, longPress = options.longPress;
        var cssClass = 'ngx-alert-dialog ' + this.clsMap[options.style];
        var component = this.create({
            title: title,
            content: content,
            longPress: longPress,
            type: type,
            cssClass: cssClass
        });
        var list = component.instance.ok.subscribe(function (data) {
            subject.next({
                type: 'ok',
                data: data
            });
            subject.complete();
            list.unsubscribe();
            list2.unsubscribe();
        });
        var list2 = component.instance.cancel.subscribe(function (data) {
            subject.next({
                type: 'cancel',
                data: data
            });
            subject.complete();
            list.unsubscribe();
            list2.unsubscribe();
        });
        return subject;
    };
    AlertService.ctorParameters = function () { return [
        { type: InjectionService },
        { type: OverlayService }
    ]; };
    AlertService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [InjectionService, OverlayService])
    ], AlertService);
    return AlertService;
}(DialogService));
export { AlertService };
