// ngx-awesome-uploader/esm5/lib/file-drop/dom.types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function FileSystemEntry() { }
if (false) {
    /** @type {?} */
    FileSystemEntry.prototype.name;
    /** @type {?} */
    FileSystemEntry.prototype.isDirectory;
    /** @type {?} */
    FileSystemEntry.prototype.isFile;
}
/**
 * @record
 */
export function FileSystemEntryMetadata() { }
if (false) {
    /** @type {?|undefined} */
    FileSystemEntryMetadata.prototype.modificationTime;
    /** @type {?|undefined} */
    FileSystemEntryMetadata.prototype.size;
}
/**
 * @record
 */
export function FileSystemDirectoryReader() { }
if (false) {
    /**
     * @param {?} successCallback
     * @param {?=} errorCallback
     * @return {?}
     */
    FileSystemDirectoryReader.prototype.readEntries = function (successCallback, errorCallback) { };
}
/**
 * @record
 */
export function FileSystemFlags() { }
if (false) {
    /** @type {?|undefined} */
    FileSystemFlags.prototype.create;
    /** @type {?|undefined} */
    FileSystemFlags.prototype.exclusive;
}
/**
 * @record
 */
export function FileSystemDirectoryEntry() { }
if (false) {
    /** @type {?} */
    FileSystemDirectoryEntry.prototype.isDirectory;
    /** @type {?} */
    FileSystemDirectoryEntry.prototype.isFile;
    /**
     * @return {?}
     */
    FileSystemDirectoryEntry.prototype.createReader = function () { };
    /**
     * @param {?=} path
     * @param {?=} options
     * @param {?=} successCallback
     * @param {?=} errorCallback
     * @return {?}
     */
    FileSystemDirectoryEntry.prototype.getFile = function (path, options, successCallback, errorCallback) { };
    /**
     * @param {?=} path
     * @param {?=} options
     * @param {?=} successCallback
     * @param {?=} errorCallback
     * @return {?}
     */
    FileSystemDirectoryEntry.prototype.getDirectory = function (path, options, successCallback, errorCallback) { };
}
/**
 * @record
 */
export function FileSystemFileEntry() { }
if (false) {
    /** @type {?} */
    FileSystemFileEntry.prototype.isDirectory;
    /** @type {?} */
    FileSystemFileEntry.prototype.isFile;
    /**
     * @param {?} callback
     * @return {?}
     */
    FileSystemFileEntry.prototype.file = function (callback) { };
}
