// @swimlane/ngx-ui/esm5/lib/services/injection-registry/injection-registry.service.js
import { __assign } from "tslib";
/* istanbul ignore next */
var InjectionRegistryService = /** @class */ (function () {
    function InjectionRegistryService(injectionService) {
        this.injectionService = injectionService;
        this.defaults = {};
        this.components = new Map();
    }
    InjectionRegistryService.prototype.getByType = function (type) {
        if (type === void 0) { type = this.type; }
        return this.components.get(type);
    };
    InjectionRegistryService.prototype.create = function (bindings) {
        return this.createByType(this.type, bindings);
    };
    InjectionRegistryService.prototype.createByType = function (type, bindings) {
        var location = bindings.parentContainer;
        bindings = this.assignDefaults(bindings);
        var component = this.injectComponent(type, bindings, location);
        this.register(type, component);
        return component;
    };
    InjectionRegistryService.prototype.destroy = function (instance) {
        var compsByType = this.components.get(instance.componentType);
        if (compsByType && compsByType.length) {
            var idx = compsByType.indexOf(instance);
            if (idx > -1) {
                var component = compsByType[idx];
                component.destroy();
                compsByType.splice(idx, 1);
            }
        }
    };
    InjectionRegistryService.prototype.destroyAll = function () {
        this.destroyByType(this.type);
    };
    InjectionRegistryService.prototype.destroyByType = function (type) {
        var comps = this.components.get(type);
        if (comps && comps.length) {
            var i = comps.length - 1;
            while (i >= 0) {
                this.destroy(comps[i--]);
            }
        }
    };
    InjectionRegistryService.prototype.injectComponent = function (type, bindings, location) {
        return this.injectionService.appendComponent(type, bindings, location);
    };
    InjectionRegistryService.prototype.assignDefaults = function (bindings) {
        var inputs = __assign({}, this.defaults.inputs);
        var outputs = __assign({}, this.defaults.outputs);
        if (!bindings.inputs && !bindings.outputs) {
            bindings = { inputs: bindings };
        }
        if (inputs) {
            bindings.inputs = __assign(__assign({}, inputs), bindings.inputs);
        }
        if (outputs) {
            bindings.outputs = __assign(__assign({}, outputs), bindings.outputs);
        }
        return bindings;
    };
    InjectionRegistryService.prototype.register = function (type, component) {
        if (!this.components.has(type)) {
            this.components.set(type, []);
        }
        var types = this.components.get(type);
        types.push(component);
    };
    return InjectionRegistryService;
}());
export { InjectionRegistryService };
