// @swimlane/ngx-ui/esm2015/lib/components/input/input-autosize.directive.js
import { __decorate, __metadata } from "tslib";
import { ElementRef, HostListener, Directive, AfterContentChecked, Renderer2 } from '@angular/core';
let AutosizeDirective = class AutosizeDirective {
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    ngAfterContentChecked() {
        this.adjust();
    }
    onInput(_) {
        this.adjust();
    }
    adjust() {
        const height = this.element.nativeElement.scrollHeight + 'px';
        this.renderer.setStyle(this.element.nativeElement, 'overflow', 'hidden');
        this.renderer.setStyle(this.element.nativeElement, 'height', height);
    }
};
AutosizeDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
__decorate([
    HostListener('input', ['$event.target']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [HTMLTextAreaElement]),
    __metadata("design:returntype", void 0)
], AutosizeDirective.prototype, "onInput", null);
AutosizeDirective = __decorate([
    Directive({
        // tslint:disable-next-line:directive-selector
        exportAs: 'ngxAutosize',
        selector: 'textarea[autosize]'
    }),
    __metadata("design:paramtypes", [ElementRef, Renderer2])
], AutosizeDirective);
export { AutosizeDirective };
