// @wkoza/ngx-upload/esm5/directives/inputfile.directive.js
import * as tslib_1 from "tslib";
import { ComponentFactoryResolver, Directive, Injector, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { InputfileComponent } from '../components/inputfile.component';
import { ngxInputFileOptions } from '../utils/configuration.model';
var NgxInputFileDirective = /** @class */ (function () {
    function NgxInputFileDirective(resolver, injector, vcRef, templateRef) {
        this.resolver = resolver;
        this.injector = injector;
        this.vcRef = vcRef;
        this.templateRef = templateRef;
    }
    NgxInputFileDirective.prototype.ngOnInit = function () {
        var _contentViewRef = this.templateRef.createEmbeddedView(null);
        var factory = this.resolver.resolveComponentFactory(InputfileComponent);
        var component = this.vcRef.createComponent(factory, 0, this.injector, [_contentViewRef.rootNodes]);
        component.instance.options = (this.ngxInputFile) ? this.ngxInputFile : ngxInputFileOptions;
        _contentViewRef.detectChanges();
    };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Object)
    ], NgxInputFileDirective.prototype, "ngxInputFile", void 0);
    NgxInputFileDirective = tslib_1.__decorate([
        Directive({
            selector: '[ngxInputFile]'
        }),
        tslib_1.__metadata("design:paramtypes", [ComponentFactoryResolver, Injector,
            ViewContainerRef, TemplateRef])
    ], NgxInputFileDirective);
    return NgxInputFileDirective;
}());
export { NgxInputFileDirective };
