// @swimlane/ngx-ui/esm5/lib/components/tabs/tab.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, TemplateRef, ContentChild, ElementRef, Renderer2, OnInit, ViewChild, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { IfTabActiveDirective } from './if-tab-active.directive';
/**
 * TODO: Remove hidden when https://github.com/angular/angular/issues/18310 is resolved
 */
var TabComponent = /** @class */ (function () {
    function TabComponent(cdr, renderer, elRef) {
        this.cdr = cdr;
        this.renderer = renderer;
        this.elRef = elRef;
        this.title = '';
        this.label = '';
        this.active = false;
        this.disabled = false;
    }
    TabComponent.prototype.ngOnInit = function () {
        // backwards compatibility
        if (this.title) {
            this.label = this.title;
            this.renderer.removeAttribute(this.elRef.nativeElement, 'title');
        }
        this.labelTemplate = typeof this.label === 'string' ? this.labelStringTemplate : this.label;
    };
    TabComponent.prototype.detectChanges = function () {
        this.cdr.detectChanges();
    };
    TabComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef },
        { type: Renderer2 },
        { type: ElementRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], TabComponent.prototype, "title", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TabComponent.prototype, "label", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TabComponent.prototype, "active", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], TabComponent.prototype, "disabled", void 0);
    __decorate([
        ViewChild('labelIsStringTmpl', { static: true }),
        __metadata("design:type", Object)
    ], TabComponent.prototype, "labelStringTemplate", void 0);
    __decorate([
        ContentChild(IfTabActiveDirective),
        __metadata("design:type", IfTabActiveDirective)
    ], TabComponent.prototype, "template", void 0);
    TabComponent = __decorate([
        Component({
            selector: 'ngx-tab',
            template: "<div *ngIf=\"template; then template_container; else content_container\"></div>\n<ng-template #template_container>\n  <div *ngIf=\"active\">\n    <ng-container [ngTemplateOutlet]=\"template.templateRef\"></ng-container>\n  </div>\n</ng-template>\n<ng-template #content_container>\n  <div [hidden]=\"!active\">\n    <ng-content></ng-content>\n  </div>\n</ng-template>\n<ng-template #labelIsStringTmpl>{{ label }}</ng-template>\n",
            host: {
                class: 'ngx-tab'
            },
            changeDetection: ChangeDetectionStrategy.OnPush
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef, Renderer2, ElementRef])
    ], TabComponent);
    return TabComponent;
}());
export { TabComponent };
