// @swimlane/ngx-ui/esm5/lib/components/tabs/tabs.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, ContentChildren, QueryList, EventEmitter, ViewEncapsulation, AfterContentInit, ChangeDetectionStrategy, ChangeDetectorRef, OnDestroy } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TabComponent } from './tab.component';
var TabsComponent = /** @class */ (function () {
    function TabsComponent(_cdr) {
        this._cdr = _cdr;
        this.selectTab = new EventEmitter();
        // For backwards compat... user selectTab instead.
        this.select = this.selectTab;
        this._destroy$ = new Subject();
    }
    Object.defineProperty(TabsComponent.prototype, "index", {
        get: function () {
            var tabs = this.tabs.toArray();
            return tabs.findIndex(function (tab) { return tab.active; });
        },
        enumerable: true,
        configurable: true
    });
    TabsComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        var tabs = this.tabs.toArray();
        var actives = this.tabs.filter(function (t) { return t.active; });
        if (actives.length > 1) {
            console.error("Multiple active tabs set 'active'");
        }
        else if (!actives.length && tabs.length) {
            setTimeout(function () {
                tabs[0].active = true;
                tabs[0].detectChanges();
                _this._cdr.markForCheck();
            });
        }
        this.tabs.changes.pipe(takeUntil(this._destroy$)).subscribe(function () { return _this._cdr.markForCheck(); });
    };
    TabsComponent.prototype.ngOnDestroy = function () {
        this._destroy$.next();
        this._destroy$.complete();
    };
    TabsComponent.prototype.tabClicked = function (activeTab) {
        this.tabs.forEach(function (tab) { return (tab.active = false); });
        activeTab.active = true;
        this.tabs.forEach(function (tab) { return tab.detectChanges(); });
        this._cdr.markForCheck();
        this.selectTab.emit(activeTab);
    };
    TabsComponent.prototype.move = function (offset) {
        var tabs = this.tabs.toArray();
        for (var i = this.index + offset; i < tabs.length && i >= 0; i += offset) {
            var tab = tabs[i];
            if (tab && !tab.disabled) {
                this.tabClicked(tabs[i]);
                return;
            }
        }
    };
    TabsComponent.prototype.next = function () {
        this.move(1);
    };
    TabsComponent.prototype.prev = function () {
        this.move(-1);
    };
    TabsComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], TabsComponent.prototype, "vertical", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TabsComponent.prototype, "selectTab", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], TabsComponent.prototype, "select", void 0);
    __decorate([
        ContentChildren(TabComponent),
        __metadata("design:type", QueryList)
    ], TabsComponent.prototype, "tabs", void 0);
    TabsComponent = __decorate([
        Component({
            exportAs: 'ngxTabs',
            selector: 'ngx-tabs',
            template: "<section>\n  <ul class=\"ngx-tabs-list list-reset\" [class.tabs-vertical]=\"vertical\" [class.tabs-horizontal]=\"!vertical\">\n    <li *ngFor=\"let tab of tabs\" class=\"ngx-tab\" [class.disabled]=\"tab.disabled\" [class.active]=\"tab.active\">\n      <button (click)=\"tabClicked(tab)\" [disabled]=\"tab.disabled\">\n        <ng-container *ngTemplateOutlet=\"tab.labelTemplate; context: { $implicit: tab }\"> </ng-container>\n      </button>\n    </li>\n  </ul>\n  <div class=\"ngx-tab-content\">\n    <ng-content></ng-content>\n  </div>\n</section>\n",
            host: {
                class: 'ngx-tabs'
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-tabs{margin-bottom:2em;display:block}.ngx-tabs .ngx-tabs-list{display:block;border-bottom:2px solid #313847}.ngx-tabs .ngx-tabs-list li{display:inline-block}.ngx-tabs .ngx-tabs-list li.active button,.ngx-tabs .ngx-tabs-list li.active button:focus,.ngx-tabs .ngx-tabs-list li.active button:hover{color:#cfcfcf;border-width:0}.ngx-tabs .ngx-tabs-list li.active button:not([disabled])::after,.ngx-tabs .ngx-tabs-list li:hover button:not([disabled])::after{-webkit-transform:scale(1);transform:scale(1)}.ngx-tabs .ngx-tabs-list li button{border:none;color:#cdd2dd;background:0 0;outline:0;box-shadow:none;box-sizing:border-box;display:inline-block;margin:0;padding:.35em .75em;position:relative;text-align:center;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;font:inherit;font-size:1em;bottom:-1px}.ngx-tabs .ngx-tabs-list li button:not([disabled]):hover{border:none;color:#cfcfcf;background:0 0;opacity:1}.ngx-tabs .ngx-tabs-list li button::after{content:'';background:#1483ff;height:2px;position:absolute;width:100%;left:0;bottom:-1px;-webkit-transition:250ms;transition:250ms;-webkit-transform:scale(0);transform:scale(0);color:#cfcfcf}.ngx-tabs .ngx-tab-content{padding:10px}"]
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], TabsComponent);
    return TabsComponent;
}());
export { TabsComponent };
