// @swimlane/ngx-ui/esm2015/lib/directives/resize-observer/resize-observer.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Output, EventEmitter, OnInit, OnDestroy, ElementRef } from '@angular/core';
import ResizeObserver from 'resize-observer-polyfill';
let ResizeObserverDirective = class ResizeObserverDirective {
    constructor(_el) {
        this._el = _el;
        this.resize = new EventEmitter();
    }
    ngOnInit() {
        this._observer = new ResizeObserver(entries => {
            for (const entry of entries) {
                this.onResize(entry.contentRect);
            }
        });
        this._observer.observe(this._el.nativeElement);
    }
    ngOnDestroy() {
        this._observer.unobserve(this._el.nativeElement);
    }
    onResize(e) {
        if (this._timer) {
            clearTimeout(this._timer);
            this._timer = undefined;
        }
        this._timer = setTimeout(() => this.resize.emit(e), 100);
    }
};
ResizeObserverDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Output('resizeObserver'),
    __metadata("design:type", Object)
], ResizeObserverDirective.prototype, "resize", void 0);
ResizeObserverDirective = __decorate([
    Directive({
        exportAs: 'resizeObserver',
        selector: '[resizeObserver]'
    }),
    __metadata("design:paramtypes", [ElementRef])
], ResizeObserverDirective);
export { ResizeObserverDirective };
