// @wkoza/ngx-upload/esm2015/services/httpClientUpload.service.js
import * as tslib_1 from "tslib";
import { Subject } from 'rxjs';
import { HttpClient, HttpErrorResponse, HttpEventType, HttpRequest, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { FileItem } from './fileItem.model';
import { NgxUploadLogger } from '../utils/logger.model';
import * as i0 from "@angular/core";
import * as i1 from "../utils/logger.model";
import * as i2 from "@angular/common/http";
// send an event for each upload event. These events can be catched by the user for call a callback
let HttpClientUploadService = class HttpClientUploadService {
    constructor(logger, httpClient) {
        this.logger = logger;
        this.httpClient = httpClient;
        this.progressTotal = 0;
        this.onCancel$ = new Subject();
        this.onError$ = new Subject();
        this.onDropError$ = new Subject();
        this.onSuccess$ = new Subject(); // TODO headers isn't `any` but `Array`
        this.onBeforeUploadItem$ = new Subject();
        this.onProgress$ = new Subject();
        this.onAddToQueue$ = new Subject();
        this.queue = new Array();
        this.headers = new Map();
    }
    /**
     * Adds files to the queue
     */
    addToQueue(files, formGroup, options) {
        this.logger.info('add to queue');
        if (options && !options.multiple) {
            if (files.length > 1) {
                this.logger.error('there is more than one file.');
                this.onDropError$.next({ errorAccept: false, errorMultiple: true });
                return;
            }
        }
        for (let i = 0; i < files.length; i++) {
            const file = files.item(i);
            this.logger.debug(files.item(i));
            if (options && options.accept) {
                const accepted = options.accept.some((type) => {
                    if (type.indexOf('/*') > -1) {
                        return type.split('/')[0] === file.type.split('/')[0];
                    }
                    else {
                        return (type === '*' || type === file.type);
                    }
                });
                if (!accepted) {
                    this.logger.error('this file is not accepted because of its type', file);
                    this.onDropError$.next({ item: file, errorAccept: true, errorMultiple: false });
                    continue;
                }
            }
            let fileItem;
            if (options && options.disableMultipart) {
                fileItem = new FileItem(file, this, this.logger, true);
            }
            else {
                fileItem = new FileItem(file, this, this.logger, false);
                if (formGroup) {
                    Object.keys(formGroup.controls).forEach((key) => {
                        fileItem.formData.append(key, formGroup.get(key).value);
                    });
                }
            }
            this.queue.push(fileItem);
            this.onAddToQueue$.next(fileItem);
        }
    }
    uploadFileItem(fileItem, endpoint, options = {}) {
        this.logger.info('enter uploadService.uploadFileItem()');
        const method = endpoint.method;
        const url = endpoint.url;
        const index = this.queue.indexOf(fileItem);
        const item = this.queue[index];
        this.onBeforeUploadItem(item);
        if (item.isCancel) {
            return;
        }
        item.uploadInProgress = true;
        let sendable;
        if (!fileItem.disableMultipart) {
            sendable = item.formData;
            sendable.append(item.alias, item.file, item.file.name);
        }
        else {
            sendable = item.file;
        }
        const req = new HttpRequest(method, url, sendable, Object.assign(options, { reportProgress: true }));
        fileItem.sub = this.httpClient.request(req).subscribe((event) => {
            if (event.type === HttpEventType.UploadProgress) {
                // This is an upload progress event. Compute and show the % done:
                const percentDone = Math.round(event.loaded * 100 / (event.total ? event.total : event.loaded));
                this.logger.debug(`File is ${percentDone}% uploaded.`);
                fileItem.ɵonProgress(percentDone);
                this.onProgressItem(item, percentDone);
            }
            else if (event instanceof HttpResponse) {
                // A successful response is delivered on the event stream.
                fileItem.ɵonSuccess();
                this.onSuccess(item, event.body, event.status, event.headers);
            }
        }, (err) => {
            if (err instanceof HttpErrorResponse) {
                if (url === 'ngx_upload_mock') {
                    item.ɵonSuccess();
                    this.onSuccess(item, err.message, err.status, err.headers);
                }
                else if (err.error instanceof Error) {
                    // A client-side or network error occurred. Handle it accordingly.
                    item.ɵonError();
                    this.onError(item, err.error.message, err.status, err.headers);
                }
                else {
                    // The backend returned an unsuccessful response code.
                    // The response body may contain clues as to what went wrong,
                    item.ɵonError();
                    this.onError(item, err.error, err.status, err.headers);
                }
            }
        });
    }
    cancelFileItem(fileItem) {
        this.progressTotal = this.computeTotalProgress();
        this.onCancel$.next(fileItem);
    }
    /**
     * Uploads all not uploaded items of queue
     */
    uploadAll(endpoint, options) {
        const items = this.queue.filter(item => (item.isReady));
        if (!items.length) {
            return;
        }
        for (const item of items) {
            item.upload(endpoint, options);
        }
    }
    /**
     * Uploads all not uploaded items of queue
     */
    cancelAll() {
        const items = this.queue.filter(item => (item.uploadInProgress));
        if (!items.length) {
            return;
        }
        for (const item of items) {
            item.cancel();
        }
        this.progressTotal = this.computeTotalProgress();
    }
    /**
     * Uploads all not uploaded items of queue
     */
    removeAllFromQueue() {
        const items = this.queue.filter(item => (!item.uploadInProgress && !item.isSuccess));
        if (!items.length) {
            return;
        }
        for (const item of items) {
            this.removeFromQueue(item);
        }
    }
    removeFromQueue(fileItem) {
        const index = this.queue.indexOf(fileItem);
        const item = this.queue[index];
        if (item.uploadInProgress) {
            item.cancel();
        }
        this.queue.splice(index, 1);
        this.progressTotal = this.computeTotalProgress();
    }
    /**
     * Returns the total progress
     * @param {Number} [value]
     * @returns {Number}
     * @private
     */
    computeTotalProgress() {
        let totalCurrent = 0;
        let total = 0;
        for (const item of this.queue) {
            if (item.uploadInProgress || item.isSuccess) {
                totalCurrent += (item.file.size / 100) * item.progress || 0;
                total += item.file.size;
                this.logger.debug(totalCurrent + ' / ' + total);
            }
        }
        return Math.round((totalCurrent * 100) / total);
    }
    /**
     * Prepares file status before upload
     * @param item
     */
    onBeforeUploadItem(item) {
        this.logger.info('enter uploadService.ɵonBeforeUploadItem()');
        item.ɵonBeforeUploadItem();
        this.onBeforeUploadItem$.next(item);
    }
    /**
     * Update status during upload progress
     * @param item
     * @param progress
     */
    onProgressItem(item, progress) {
        this.logger.info(`call onProgressItem ${item} ${progress}`);
        this.progressTotal = this.computeTotalProgress();
        item.ɵonProgress(progress);
        this.onProgress$.next({ item, progress });
    }
    /**
     * Callback called when an upload error occurs
     * @param item
     * @param xhr
     */
    onError(item, body, status, headers) {
        this.logger.info(`call onError ${item} ${body} ${status} ${headers}`);
        item.ɵonError();
        this.onError$.next({ item, body, status, headers });
    }
    /**
     * Callback called when an upload success occurs
     * @param item
     * @param xhr
     */
    onSuccess(item, body, status, headers) {
        this.logger.info(`call onSuccess ${item} ${body} ${status} ${headers}`);
        this.progressTotal = this.computeTotalProgress();
        item.ɵonSuccess();
        this.onSuccess$.next({ item, body, status, headers });
    }
};
HttpClientUploadService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function HttpClientUploadService_Factory() { return new HttpClientUploadService(i0.ɵɵinject(i1.NgxUploadLogger), i0.ɵɵinject(i2.HttpClient)); }, token: HttpClientUploadService, providedIn: "root" });
HttpClientUploadService = tslib_1.__decorate([
    Injectable({
        providedIn: 'root'
    }),
    tslib_1.__metadata("design:paramtypes", [NgxUploadLogger, HttpClient])
], HttpClientUploadService);
export { HttpClientUploadService };
