// @swimlane/ngx-ui/esm2015/lib/components/drawer/drawer.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, HostListener, ViewEncapsulation, ChangeDetectionStrategy, TemplateRef } from '@angular/core';
import { coerceNumberProperty, coerceBooleanProperty } from '@angular/cdk/coercion';
import { trigger } from '@angular/animations';
import { DRAWER_ANIMATION } from './drawer.animation';
import { DrawerDirection } from './drawer-direction.enum';
import { DrawerPosition } from './drawer-position.enum';
let DrawerComponent = class DrawerComponent {
    constructor() {
        this.cssClass = '';
        this.isRoot = true;
        this.close = new EventEmitter();
        this.position = DrawerPosition.fixed;
    }
    get size() {
        return this._size;
    }
    set size(val) {
        this._size = coerceNumberProperty(val);
    }
    get zIndex() {
        return this._zIndex;
    }
    set zIndex(val) {
        this._zIndex = coerceNumberProperty(val);
    }
    get closeOnOutsideClick() {
        return this._closeOnOutsideClick;
    }
    set closeOnOutsideClick(val) {
        this._closeOnOutsideClick = coerceBooleanProperty(val);
    }
    get cssClasses() {
        let clz = `ngx-drawer ${this.cssClass}`;
        if (this.isLeft)
            clz += ' left-drawer';
        if (this.isBottom)
            clz += ' bottom-drawer';
        return clz;
    }
    get isLeft() {
        return this.direction === DrawerDirection.Left;
    }
    get isBottom() {
        return this.direction === DrawerDirection.Bottom;
    }
    ngOnInit() {
        this.position = this.isRoot ? DrawerPosition.fixed : DrawerPosition.absolute;
        this.setDimensions(this.size);
    }
    ngOnDestroy() {
        this.close.emit(true);
    }
    setDimensions(size) {
        this.heightSize = `${this.isBottom && size ? size : 100}%`;
        this.widthSize = `${this.isLeft && size ? size : 100}%`;
    }
    onEscapeKey() {
        this.close.emit(true);
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], DrawerComponent.prototype, "cssClass", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], DrawerComponent.prototype, "direction", void 0);
__decorate([
    Input(),
    __metadata("design:type", TemplateRef)
], DrawerComponent.prototype, "template", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], DrawerComponent.prototype, "context", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean)
], DrawerComponent.prototype, "isRoot", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number),
    __metadata("design:paramtypes", [Number])
], DrawerComponent.prototype, "size", null);
__decorate([
    Input(),
    __metadata("design:type", Number),
    __metadata("design:paramtypes", [Number])
], DrawerComponent.prototype, "zIndex", null);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], DrawerComponent.prototype, "closeOnOutsideClick", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], DrawerComponent.prototype, "close", void 0);
__decorate([
    HostListener('keyup.esc'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], DrawerComponent.prototype, "onEscapeKey", null);
DrawerComponent = __decorate([
    Component({
        exportAs: 'ngxDrawer',
        selector: 'ngx-drawer',
        template: "<div class=\"ngx-drawer-content\">\n  <ng-template [ngTemplateOutlet]=\"template\" [ngTemplateOutletContext]=\"{ context: context, close: close }\">\n  </ng-template>\n</div>\n",
        animations: [trigger('drawerTransition', DRAWER_ANIMATION)],
        host: {
            role: 'dialog',
            tabindex: '-1',
            '[class]': 'cssClasses',
            '[style.width]': 'widthSize',
            '[style.height]': 'heightSize',
            '[style.zIndex]': 'zIndex',
            '[style.position]': 'position',
            '[@drawerTransition]': 'direction'
        },
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [".ngx-drawer{display:block;overflow-y:auto;overflow-x:hidden;text-align:left;z-index:998;background:#1c2029}.ngx-drawer.left-drawer{top:0;right:0;height:100%}.ngx-drawer.bottom-drawer{left:0;bottom:0;width:100%}"]
    })
], DrawerComponent);
export { DrawerComponent };
