// @swimlane/ngx-ui/esm5/lib/components/split/split-area.directive.js
import { __decorate, __metadata, __param } from "tslib";
import { Directive, Optional, Self, Input } from '@angular/core';
import { DefaultFlexDirective, validateBasis } from '@angular/flex-layout';
var SplitAreaDirective = /** @class */ (function () {
    function SplitAreaDirective(flexDirective) {
        this.flexDirective = flexDirective;
        this.overflow = 'hidden';
    }
    Object.defineProperty(SplitAreaDirective.prototype, "fxFlexFill", {
        get: function () {
            return this.fxFlex === '';
        },
        enumerable: true,
        configurable: true
    });
    SplitAreaDirective.prototype.ngOnChanges = function () {
        this.currentFlexBasis = this.initialFlexBasis = this.getCurrentFlexParts();
    };
    SplitAreaDirective.prototype.updateStyle = function (flexBasis) {
        var flex = this.flexDirective;
        if (typeof flexBasis === 'undefined') {
            flexBasis = flex.activatedValue || '';
        }
        if (typeof flexBasis === 'number') {
            flexBasis = this.isPercent() ? flexBasis + "%" : flexBasis + "px";
        }
        if (flexBasis.indexOf(' ') < 0) {
            var grow = flex.grow;
            var shrink = flex.shrink;
            this.currentFlexBasis = [grow, shrink, flexBasis];
            flexBasis = this.currentFlexBasis.join(' ');
        }
        else {
            this.currentFlexBasis = flexBasis.split(' ');
        }
        flex.activatedValue = this.currentFlexBasis[2];
    };
    SplitAreaDirective.prototype.isPercent = function (basis) {
        if (!basis) {
            var flex = this.flexDirective;
            basis = flex.activatedValue || '1 1 1e-9px';
        }
        var hasCalc = String(basis).indexOf('calc') > -1;
        return String(basis).indexOf('%') > -1 && !hasCalc;
    };
    SplitAreaDirective.prototype.getCurrentFlexParts = function () {
        var flex = this.flexDirective;
        var basis = (flex && flex.activatedValue) || '1 1 1e-9px';
        return validateBasis(String(basis).replace(';', ''), flex.grow, flex.shrink);
    };
    SplitAreaDirective.ctorParameters = function () { return [
        { type: DefaultFlexDirective, decorators: [{ type: Optional }, { type: Self }] }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], SplitAreaDirective.prototype, "minBasis", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], SplitAreaDirective.prototype, "maxBasis", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], SplitAreaDirective.prototype, "fxFlex", void 0);
    SplitAreaDirective = __decorate([
        Directive({
            exportAs: 'ngxSplitArea',
            selector: '[ngxSplitArea]',
            host: {
                class: 'ngx-split-area',
                '[style.overflow]': 'overflow'
            }
        }),
        __param(0, Optional()),
        __param(0, Self()),
        __metadata("design:paramtypes", [DefaultFlexDirective])
    ], SplitAreaDirective);
    return SplitAreaDirective;
}());
export { SplitAreaDirective };
