// @swimlane/ngx-ui/esm5/lib/components/split/split-handle.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Output, ViewEncapsulation } from '@angular/core';
import { fromEvent } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { SplitDirection } from './split-direction.enum';
var SplitHandleComponent = /** @class */ (function () {
    function SplitHandleComponent() {
        this.drag = new EventEmitter();
        this.dragStart = new EventEmitter();
        this.dragEnd = new EventEmitter();
        this.dblclick = new EventEmitter();
        this.direction = SplitDirection.Row;
    }
    Object.defineProperty(SplitHandleComponent.prototype, "isRow", {
        get: function () {
            return this.direction === SplitDirection.Row;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SplitHandleComponent.prototype, "isColumn", {
        get: function () {
            return this.direction === SplitDirection.Column;
        },
        enumerable: true,
        configurable: true
    });
    SplitHandleComponent.prototype.onMousedown = function (ev) {
        var _this = this;
        var mouseup$ = fromEvent(document, 'mouseup');
        this.subscription = mouseup$.subscribe(/* istanbul ignore next */ function (e) { return _this.onMouseup(e); });
        var mousemove$ = fromEvent(document, 'mousemove')
            .pipe(takeUntil(mouseup$))
            .subscribe(/* istanbul ignore next */ function (e) { return _this.onMouseMove(e); });
        this.subscription.add(mousemove$);
        this.dragStart.emit(ev);
    };
    SplitHandleComponent.prototype.onMouseMove = function (ev) {
        this.drag.emit(ev);
    };
    SplitHandleComponent.prototype.onMouseup = function (ev) {
        if (this.subscription) {
            this.dragEnd.emit(ev);
            this.subscription.unsubscribe();
            this.subscription = undefined;
        }
    };
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], SplitHandleComponent.prototype, "drag", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], SplitHandleComponent.prototype, "dragStart", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], SplitHandleComponent.prototype, "dragEnd", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], SplitHandleComponent.prototype, "dblclick", void 0);
    SplitHandleComponent = __decorate([
        Component({
            exportAs: 'ngxSplitHandle',
            // tslint:disable-next-line:component-selector
            selector: '[ngxSplitHandle]',
            template: "<button\n  #splitHandle\n  class=\"icon-split-handle ngx-split-button\"\n  (mousedown)=\"onMousedown($event)\"\n  (dblclick)=\"dblclick.emit($event)\"\n></button>\n",
            host: {
                class: 'ngx-split-handle',
                '[class.direction-row]': 'isRow',
                '[class.direction-column]': 'isColumn'
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-split-handle{position:relative}.ngx-split-handle .ngx-split-button{line-height:0;font-size:32px;position:absolute;display:block;padding:0}.ngx-split-handle.direction-row .ngx-split-button{top:50%;left:50%;cursor:col-resize;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%)}.ngx-split-handle.direction-column .ngx-split-button{left:50%;cursor:row-resize;top:-3px;-webkit-transform:translateX(-50%) rotate(270deg);transform:translateX(-50%) rotate(270deg)}"]
        })
    ], SplitHandleComponent);
    return SplitHandleComponent;
}());
export { SplitHandleComponent };
