// @swimlane/ngx-ui/esm2015/lib/components/json-editor/json-editor.js
import { __decorate, __metadata } from "tslib";
import { Input, Output, EventEmitter } from '@angular/core';
export class JsonEditor {
    constructor(schemaValidatorService) {
        this.schemaValidatorService = schemaValidatorService;
        this.modelChange = new EventEmitter();
        this.schemaChange = new EventEmitter();
    }
    ngOnChanges(changes) {
        if (changes.schema) {
            this.schema = JSON.parse(JSON.stringify(this.schema));
        }
        if (!this.schema) {
            this.schema = {
                type: 'object'
            };
        }
        if (!this.schema.type) {
            this.schema.type = 'object';
        }
    }
    /**
     * Model change callback. Validates the model and emits a change event
     * @param model
     */
    modelChangedCallback(model) {
        this.validate(this.schema, model);
        this.modelChange.emit(model);
    }
    /**
     * Validates the model based on the schema
     * @param schema
     * @param model
     */
    validate(schema, model) {
        this.errors = this.schemaValidator
            ? this.schemaValidator(schema, model)
            : this.schemaValidatorService.validate(schema, model);
        return this.errors && this.errors.length > 0;
    }
}
__decorate([
    Input(),
    __metadata("design:type", Object)
], JsonEditor.prototype, "model", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], JsonEditor.prototype, "schema", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], JsonEditor.prototype, "label", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], JsonEditor.prototype, "typeCheckOverrides", void 0);
__decorate([
    Input(),
    __metadata("design:type", Function)
], JsonEditor.prototype, "schemaValidator", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], JsonEditor.prototype, "modelChange", void 0);
__decorate([
    Output(),
    __metadata("design:type", EventEmitter)
], JsonEditor.prototype, "schemaChange", void 0);
