// @swimlane/ngx-ui/esm5/lib/directives/validators/pattern-validator/pattern-validator.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input } from '@angular/core';
import { NG_VALIDATORS } from '@angular/forms';
import { patternValidator } from './pattern.validator';
var PatternValidatorDirective = /** @class */ (function () {
    function PatternValidatorDirective() {
    }
    PatternValidatorDirective_1 = PatternValidatorDirective;
    PatternValidatorDirective.prototype.validate = function (control) {
        return this.pattern ? patternValidator(this.pattern)(control) : null;
    };
    var PatternValidatorDirective_1;
    __decorate([
        Input('pattern'),
        __metadata("design:type", String)
    ], PatternValidatorDirective.prototype, "pattern", void 0);
    PatternValidatorDirective = PatternValidatorDirective_1 = __decorate([
        Directive({
            selector: '[pattern]',
            providers: [
                {
                    provide: NG_VALIDATORS,
                    useExisting: PatternValidatorDirective_1,
                    multi: true
                }
            ]
        })
    ], PatternValidatorDirective);
    return PatternValidatorDirective;
}());
export { PatternValidatorDirective };
