// @swimlane/ngx-ui/esm2015/lib/components/notification/notification.service.js
var NotificationService_1;
import { __decorate, __metadata, __param } from "tslib";
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
import { InjectionService } from '../../services/injection/injection.service';
import { InjectionRegistryService } from '../../services/injection-registry/injection-registry.service';
import { NotificationType } from './notification-type.enum';
import { NotificationStyleType } from './notification-style-type.enum';
import { NotificationPermissionType } from './notification-permission-type.enum';
import { NotificationComponent } from './notification.component';
import { NotificationContainerComponent } from './notification-container.component';
/** adding dynamic to suppress `Document` type metadata error  */
/** @dynamic */
let NotificationService = NotificationService_1 = class NotificationService extends InjectionRegistryService {
    constructor(injectionService, document) {
        super(injectionService);
        this.injectionService = injectionService;
        this.document = document;
        this.defaults = {
            inputs: {
                timeout: 3000,
                rateLimit: true,
                pauseOnHover: true,
                type: NotificationType.html,
                styleType: NotificationStyleType.none,
                showClose: true,
                sound: false
            }
        };
        this.type = NotificationComponent;
    }
    get isNativeSupported() {
        return 'Notification' in window;
    }
    create(bindings) {
        // verify flood not happening
        if (bindings.rateLimit && this.isFlooded(bindings)) {
            return;
        }
        // if limit reached, remove the first one
        const compsByType = this.getByType();
        if (compsByType && compsByType.length >= NotificationService_1.limit) {
            this.destroy(compsByType[0]);
        }
        // native notifications need to be invoked
        let component;
        if (bindings.type === NotificationType.native) {
            component = this.showNative(bindings);
        }
        else {
            component = super.create(bindings);
            this.createSubscriptions(component);
            this.startTimer(component);
        }
        return component;
    }
    startTimer(component) {
        if (component.instance && component.instance.timeout !== false) {
            clearTimeout(component.instance.timer);
            component.instance.timer = setTimeout(() => {
                this.destroy(component);
            }, component.instance.timeout);
        }
    }
    pauseTimer(component) {
        clearTimeout(component.instance.timer);
    }
    requestPermissions() {
        if (this.isNativeSupported) {
            Notification.requestPermission(/* istanbul ignore next */ /* istanbul ignore next */ status => (this.permission = status));
        }
    }
    assignDefaults(options) {
        const bindings = super.assignDefaults(options);
        if (bindings.inputs && bindings.inputs.timeout === true) {
            bindings.inputs.timeout = this.defaults.inputs.timeout;
        }
        // add a timestamp for flood checks
        bindings.inputs.timestamp = +new Date();
        return bindings;
    }
    injectComponent(type, options) {
        if (!this.container || !this.document.contains(this.container.location.nativeElement)) {
            this.container = this.injectionService.appendComponent(NotificationContainerComponent);
        }
        return this.injectionService.appendComponent(type, options, this.container);
    }
    createSubscriptions(component) {
        let pauseSub;
        let resumeSub;
        let closeSub;
        const kill = () => {
            closeSub.unsubscribe();
            resumeSub.unsubscribe();
            pauseSub.unsubscribe();
            this.destroy(component);
        };
        const pause = () => {
            this.pauseTimer(component);
        };
        const resume = () => {
            this.startTimer(component);
        };
        pauseSub = component.instance.pause.subscribe(pause);
        resumeSub = component.instance.resume.subscribe(resume);
        closeSub = component.instance.close.subscribe(kill);
    }
    isFlooded(options) {
        const compsByType = this.getByType();
        for (const notification of compsByType) {
            const instance = notification.instance;
            if (instance.title === options.title &&
                instance.body === options.body &&
                instance.timestamp + 1000 > options.timestamp) {
                return true;
            }
        }
        return false;
    }
    showNative(options) {
        if (!this.isNativeSupported)
            return;
        if (!this.permission)
            this.requestPermissions();
        if (this.permission === NotificationPermissionType.denied)
            return;
        const note = new Notification(options.title, options);
        note.onerror = () => {
            console.error('Notification failed!', options);
        };
        // manually do this
        if (options && typeof options.timeout === 'number') {
            setTimeout(note.close.bind(note), options.timeout);
        }
        return note;
    }
};
NotificationService.limit = 10;
NotificationService.ctorParameters = () => [
    { type: InjectionService },
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
NotificationService = NotificationService_1 = __decorate([
    Injectable(),
    __param(1, Inject(DOCUMENT)),
    __metadata("design:paramtypes", [InjectionService, Document])
], NotificationService);
export { NotificationService };
