// @swimlane/ngx-ui/esm2015/lib/components/select/select-option.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, TemplateRef, ContentChild } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { SelectOptionTemplateDirective } from './select-option-template.directive';
import { SelectOptionInputTemplateDirective } from './select-option-input-template.directive';
let SelectOptionDirective = class SelectOptionDirective {
    constructor() {
        this.name = '';
        this._disabled = false;
        this._hidden = false;
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(disabled) {
        this._disabled = coerceBooleanProperty(disabled);
    }
    get hidden() {
        return this._hidden;
    }
    set hidden(hidden) {
        this._hidden = coerceBooleanProperty(hidden);
    }
    get optionTemplate() {
        return this._optionTemplateInput || this._optionTemplateQuery;
    }
    get inputTemplate() {
        return this._inputTemplateInput || this._inputTemplateQuery;
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectOptionDirective.prototype, "name", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], SelectOptionDirective.prototype, "value", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectOptionDirective.prototype, "disabled", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectOptionDirective.prototype, "hidden", null);
__decorate([
    Input('optionTemplate'),
    __metadata("design:type", TemplateRef)
], SelectOptionDirective.prototype, "_optionTemplateInput", void 0);
__decorate([
    ContentChild(SelectOptionTemplateDirective, { read: TemplateRef, static: true }),
    __metadata("design:type", TemplateRef)
], SelectOptionDirective.prototype, "_optionTemplateQuery", void 0);
__decorate([
    Input('inputTemplate'),
    __metadata("design:type", TemplateRef)
], SelectOptionDirective.prototype, "_inputTemplateInput", void 0);
__decorate([
    ContentChild(SelectOptionInputTemplateDirective, { read: TemplateRef, static: true }),
    __metadata("design:type", TemplateRef)
], SelectOptionDirective.prototype, "_inputTemplateQuery", void 0);
SelectOptionDirective = __decorate([
    Directive({
        // tslint:disable-next-line:directive-selector
        exportAs: 'ngxSelectOption',
        selector: 'ngx-select-option'
    })
], SelectOptionDirective);
export { SelectOptionDirective };
