// @swimlane/ngx-ui/esm2015/lib/components/select/select.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, QueryList, ContentChildren, forwardRef, ElementRef, Renderer2, OnDestroy, ViewChild, ViewEncapsulation, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
import { SelectOptionDirective } from './select-option.directive';
import { SelectInputComponent } from './select-input.component';
import { KeyboardKeys } from '../../enums';
let nextId = 0;
const SELECT_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => SelectComponent),
    multi: true
};
let SelectComponent = class SelectComponent {
    constructor(_element, _renderer, _cdr) {
        this._element = _element;
        this._renderer = _renderer;
        this._cdr = _cdr;
        this.id = `select-${++nextId}`;
        this.placeholder = '';
        this.emptyPlaceholder = 'No options available';
        this.filterEmptyPlaceholder = 'No matches...';
        this.filterPlaceholder = 'Filter options...';
        this.allowAdditionsText = 'Add Value';
        this.requiredIndicator = '*';
        this.options = [];
        this.change = new EventEmitter();
        this.keyup = new EventEmitter();
        this.toggle = new EventEmitter();
        this.focusIndex = -1;
        this.dropdownActive = false;
        this._value = [];
        this._autofocus = false;
        this._allowClear = true;
        this._allowAdditions = false;
        this._disableDropdown = false;
        this._closeOnBodyClick = true;
        this._filterable = true;
        this._filterCaseSensitive = false;
        this._tagging = false;
        this._multiple = false;
        this._disabled = false;
    }
    get minSelections() {
        return this._minSelections;
    }
    set minSelections(minSelections) {
        this._minSelections = coerceNumberProperty(minSelections);
    }
    get maxSelections() {
        return this._maxSelections;
    }
    set maxSelections(maxSelections) {
        this._maxSelections = coerceNumberProperty(maxSelections);
    }
    get autofocus() {
        return this._autofocus;
    }
    set autofocus(autofocus) {
        this._autofocus = coerceBooleanProperty(autofocus);
    }
    get allowClear() {
        return this._allowClear;
    }
    set allowClear(allowClear) {
        this._allowClear = coerceBooleanProperty(allowClear);
    }
    get allowAdditions() {
        return this._allowAdditions;
    }
    set allowAdditions(allowAdditions) {
        this._allowAdditions = coerceBooleanProperty(allowAdditions);
    }
    get disableDropdown() {
        return this._disableDropdown;
    }
    set disableDropdown(disableDropdown) {
        this._disableDropdown = coerceBooleanProperty(disableDropdown);
    }
    get closeOnSelect() {
        return this._closeOnSelect;
    }
    set closeOnSelect(closeOnSelect) {
        this._closeOnSelect = coerceBooleanProperty(closeOnSelect);
    }
    get closeOnBodyClick() {
        return this._closeOnBodyClick;
    }
    set closeOnBodyClick(closeOnBodyClick) {
        this._closeOnBodyClick = coerceBooleanProperty(closeOnBodyClick);
    }
    get filterable() {
        return this._filterable;
    }
    set filterable(filterable) {
        this._filterable = coerceBooleanProperty(filterable);
    }
    get required() {
        return this._required;
    }
    set required(required) {
        this._required = coerceBooleanProperty(required);
    }
    get filterCaseSensitive() {
        return this._filterCaseSensitive;
    }
    set filterCaseSensitive(filterCaseSensitive) {
        this._filterCaseSensitive = coerceBooleanProperty(filterCaseSensitive);
    }
    get tagging() {
        return this._tagging;
    }
    set tagging(tagging) {
        this._tagging = coerceBooleanProperty(tagging);
    }
    get multiple() {
        return this._multiple;
    }
    set multiple(multiple) {
        this._multiple = coerceBooleanProperty(multiple);
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(disabled) {
        this._disabled = coerceBooleanProperty(disabled);
    }
    get optionTemplates() {
        return this._optionTemplates;
    }
    set optionTemplates(val) {
        this._optionTemplates = val;
        if (val) {
            const arr = val.toArray();
            if (arr.length) {
                this.options = arr;
            }
        }
        this._cdr.markForCheck();
    }
    get invalid() {
        if (this.required && this.checkInvalidValue(this.value))
            return true;
        if (this.maxSelections !== undefined && (this.value && this.value.length > this.maxSelections))
            return true;
        if (this.minSelections !== undefined && (!this.value || this.value.length < this.minSelections))
            return true;
        return false;
    }
    get requiredIndicatorView() {
        const required = this.required || (this.minSelections !== undefined && this.minSelections > 0);
        if (!this.requiredIndicator || !required) {
            return '';
        }
        return this.requiredIndicator;
    }
    get isSingleSelect() {
        return !this.multiple && !this.tagging;
    }
    get hasSelections() {
        return this.value && this.value.length > 0 && typeof this.value[0] !== 'undefined';
    }
    get hasPlaceholder() {
        return this.placeholder && this.placeholder.length > 0;
    }
    get value() {
        return this._value;
    }
    set value(val) {
        if (val !== this._value) {
            this._value = val;
            this.onChangeCallback(this._value);
            this.change.emit(this._value);
            this._cdr.markForCheck();
        }
    }
    get dropdownVisible() {
        if (this.disableDropdown)
            return false;
        if (this.tagging && (!this.options || !this.options.length))
            return false;
        return this.dropdownActive;
    }
    ngOnDestroy() {
        this.toggleDropdown(false);
    }
    onDropdownSelection(selection) {
        if (selection.disabled)
            return;
        if (this.value.length === this.maxSelections)
            return;
        const idx = this.value.findIndex(o => {
            if (this.identifier) {
                return o[this.identifier] === selection.value[this.identifier];
            }
            return o === selection.value;
        });
        if (idx === -1) {
            this.value = this.multiple || this.tagging ? [...this.value, selection.value] : [selection.value];
        }
        // if tagging, we need to clear current text
        if (this.tagging) {
            this.inputComponent.inputElement.nativeElement.value = '';
        }
        const shouldClose = this.closeOnSelect || !this.multiple;
        if (shouldClose) {
            this.toggleDropdown(false);
        }
    }
    onInputSelection(selections) {
        this.value = selections;
    }
    onFocus() {
        if (this.disabled)
            return;
        this.toggleDropdown(true);
        this.onTouchedCallback();
    }
    onClear() {
        this.value = [];
    }
    onBodyClick(event) {
        if (this.dropdownActive) {
            const contains = this._element.nativeElement.contains(event.target);
            /* istanbul ignore else */
            if (!contains) {
                this.toggleDropdown(false);
            }
        }
    }
    onClose() {
        this.toggleDropdown(false);
    }
    onToggle() {
        if (this.disabled)
            return;
        this.toggleDropdown(!this.dropdownActive);
        this.onTouchedCallback();
    }
    toggleDropdown(state) {
        if (this.dropdownActive === state)
            return;
        this.dropdownActive = state;
        if (this.toggleListener)
            this.toggleListener();
        this.toggle.emit(this.dropdownActive);
        if (state && this.closeOnBodyClick) {
            this.toggleListener = this._renderer.listen(document.body, 'click', this.onBodyClick.bind(this));
        }
        this._cdr.markForCheck();
    }
    onKeyUp({ event, value }) {
        if (event && event.key === KeyboardKeys.ARROW_DOWN) {
            ++this.focusIndex;
        }
        else {
            this.filterQuery = value;
        }
        this.keyup.emit({ event, value });
    }
    writeValue(val) {
        /* istanbul ignore else */
        if (val !== this._value) {
            this._value = val;
            this._cdr.markForCheck();
        }
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    checkInvalidValue(value) {
        if (Array.isArray(value)) {
            return !this.value.length || this.checkInvalidValue(value[0]);
        }
        return value === undefined;
    }
    /* istanbul ignore next */
    onTouchedCallback() {
        // placeholder
    }
    /* istanbul ignore next */
    onChangeCallback(_) {
        // placeholder
    }
};
SelectComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 },
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "id", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "name", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "label", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "hint", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "placeholder", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "emptyPlaceholder", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "filterEmptyPlaceholder", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "filterPlaceholder", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "allowAdditionsText", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "groupBy", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "selectCaret", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], SelectComponent.prototype, "requiredIndicator", void 0);
__decorate([
    Input(),
    __metadata("design:type", Array)
], SelectComponent.prototype, "options", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], SelectComponent.prototype, "identifier", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "minSelections", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "maxSelections", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "autofocus", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "allowClear", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "allowAdditions", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "disableDropdown", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "closeOnSelect", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "closeOnBodyClick", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "filterable", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "required", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "filterCaseSensitive", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "tagging", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "multiple", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], SelectComponent.prototype, "disabled", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], SelectComponent.prototype, "change", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], SelectComponent.prototype, "keyup", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], SelectComponent.prototype, "toggle", void 0);
__decorate([
    ViewChild(SelectInputComponent, { static: true }),
    __metadata("design:type", SelectInputComponent)
], SelectComponent.prototype, "inputComponent", void 0);
__decorate([
    ContentChildren(SelectOptionDirective),
    __metadata("design:type", QueryList),
    __metadata("design:paramtypes", [QueryList])
], SelectComponent.prototype, "optionTemplates", null);
SelectComponent = __decorate([
    Component({
        exportAs: 'ngxSelect',
        selector: 'ngx-select',
        template: "<div class=\"ngx-select-wrap\">\n  <div class=\"ngx-select-flex-wrap\">\n    <div class=\"ngx-select-flex-wrap-inner\">\n      <ngx-select-input\n        [autofocus]=\"autofocus\"\n        [options]=\"options\"\n        [allowClear]=\"allowClear\"\n        [label]=\"label\"\n        [requiredIndicator]=\"requiredIndicatorView\"\n        [placeholder]=\"placeholder\"\n        [multiple]=\"multiple\"\n        [identifier]=\"identifier\"\n        [tagging]=\"tagging\"\n        [allowAdditions]=\"allowAdditions\"\n        [selectCaret]=\"selectCaret\"\n        [selected]=\"value\"\n        [hint]=\"hint\"\n        [disableDropdown]=\"disableDropdown\"\n        (keyup)=\"onKeyUp($event)\"\n        (toggle)=\"onToggle()\"\n        (activate)=\"onFocus()\"\n        (selection)=\"onInputSelection($event)\"\n      >\n      </ngx-select-input>\n    </div>\n  </div>\n  <ngx-select-dropdown\n    *ngIf=\"dropdownVisible\"\n    [focusIndex]=\"focusIndex\"\n    [filterQuery]=\"filterQuery\"\n    [filterPlaceholder]=\"filterPlaceholder\"\n    [filterCaseSensitive]=\"filterCaseSensitive\"\n    [allowAdditions]=\"allowAdditions\"\n    [allowAdditionsText]=\"allowAdditionsText\"\n    [selected]=\"value\"\n    [groupBy]=\"groupBy\"\n    [emptyPlaceholder]=\"emptyPlaceholder\"\n    [tagging]=\"tagging\"\n    [filterEmptyPlaceholder]=\"filterEmptyPlaceholder\"\n    [filterable]=\"filterable\"\n    [identifier]=\"identifier\"\n    [options]=\"options\"\n    (keyup)=\"keyup.emit($event)\"\n    (close)=\"onClose()\"\n    (selection)=\"onDropdownSelection($event)\"\n  >\n  </ngx-select-dropdown>\n</div>\n",
        host: {
            class: 'ngx-select',
            '[id]': 'id',
            '[attr.name]': 'name',
            '[class.invalid]': 'invalid',
            '[class.tagging-selection]': 'tagging',
            '[class.multi-selection]': 'multiple',
            '[class.single-selection]': 'isSingleSelect',
            '[class.disabled]': 'disabled',
            '[class.active]': 'dropdownActive',
            '[class.active-selections]': 'hasSelections',
            '[class.has-placeholder]': 'hasPlaceholder'
        },
        providers: [SELECT_VALUE_ACCESSOR],
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: ["input[type=number],input[type=password],input[type=tel],input[type=text],textarea{display:inline-block;box-sizing:border-box;outline:0}.form-input{background:#313847;border:1px solid #455066;color:#b6b6b6;-webkit-transition:box-shadow .2s;transition:box-shadow .2s;border-radius:0;font-size:13px;height:32px;line-height:32px;width:100%;padding:6px;margin-bottom:1em}.form-input::-webkit-input-placeholder{color:#647493}.form-input::-moz-placeholder{color:#647493}.form-input::-ms-input-placeholder{color:#647493}.form-input::placeholder{color:#647493}.form-input:focus{box-shadow:0 1px 5px 0 rgba(0,0,0,.2),0 2px 2px 0 rgba(0,0,0,.14),0 3px 1px -2px rgba(0,0,0,.12)}.form-input[disabled]{cursor:not-allowed;color:#363636}textarea.form-input{min-height:120px;line-height:1.3em}.ngx-input{display:block;margin-top:16px;margin-bottom:8px;padding-top:20px;padding-top:calc(.7em + 8px);padding-bottom:0}.ngx-input input:-webkit-autofill,.ngx-input input:-webkit-autofill:active,.ngx-input input:-webkit-autofill:focus,.ngx-input input:-webkit-autofill:hover{-webkit-transition:background-color 5000s ease-in-out;transition:background-color 5000s ease-in-out;-webkit-text-fill-color:#cfcfcf!important}.ngx-input .ngx-input-flex-wrap{-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;display:-webkit-box;display:flex}.ngx-input .ngx-input-flex-wrap .ngx-input-flex-wrap-inner{display:-webkit-box;display:flex;-webkit-box-flex:1;flex:1}.ngx-input .ngx-input-flex-wrap ngx-input-prefix,.ngx-input .ngx-input-flex-wrap ngx-input-suffix{-webkit-box-flex:0;flex:none;white-space:nowrap;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center}.ngx-input .ngx-input-flex-wrap ngx-input-prefix>*,.ngx-input .ngx-input-flex-wrap ngx-input-suffix>*{display:inline-fflex}.ngx-input ngx-input-prefix{margin-right:8px}.ngx-input ngx-input-suffix{margin-left:8px}.ngx-input .ngx-input-wrap{position:relative;display:block;margin-bottom:0;width:100%}.ngx-input .ngx-input-wrap .ngx-input-box-wrap{position:relative;width:100%}.ngx-input .ngx-input-wrap .ngx-input-box-wrap:focus{outline:0}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box,.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea{-webkit-box-flex:1;flex:auto;background:0 0;border:none;margin-bottom:0;padding-left:0;width:100%;color:#cdd2dd;font-size:1em;min-height:0;font-family:inherit}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box::-webkit-input-placeholder,.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea::-webkit-input-placeholder{color:#647493}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box::-moz-placeholder,.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea::-moz-placeholder{color:#647493}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box::-ms-input-placeholder,.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea::-ms-input-placeholder{color:#647493}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box::placeholder,.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea::placeholder{color:#647493}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box:focus,.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea:focus{box-shadow:none;outline:0}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box[disabled],.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea[disabled]{color:#72819f;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-box{margin:3px 0}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .ngx-input-textarea{resize:none}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .icon-eye{line-height:25px;top:0;bottom:0;position:absolute;right:10px;cursor:pointer;font-size:.8rem;color:#9c9c9c;-webkit-transition:color .1s;transition:color .1s}.ngx-input .ngx-input-wrap .ngx-input-box-wrap .icon-eye:hover{color:#b6b6b6}.ngx-input .ngx-input-wrap .ngx-input-label{pointer-events:none;position:absolute;font-size:16px}.ngx-input .ngx-input-wrap .ngx-input-underline{width:100%;height:1px;background-color:#5a6884}.ngx-input .ngx-input-wrap .ngx-input-underline .underline-fill{background-color:#1483ff;width:100%;height:2px;margin:0 auto}.ngx-input .ngx-input-wrap .ngx-input-hint{font-size:12px;color:#909cb4;margin-top:2px;min-height:1em}.ngx-input.invalid.ng-touched .ngx-input-underline,.ngx-input.ng-invalid.ng-dirty .ngx-input-underline,.ngx-input.ng-invalid.ng-touched .ngx-input-underline{background-color:#ff4514}.ngx-input.invalid.ng-touched .ngx-input-label,.ngx-input.ng-invalid.ng-dirty .ngx-input-label,.ngx-input.ng-invalid.ng-touched .ngx-input-label{font-weight:600;color:#ff4514}.ngx-input.invalid.ng-touched .ngx-input-hint,.ngx-input.ng-invalid.ng-dirty .ngx-input-hint,.ngx-input.ng-invalid.ng-touched .ngx-input-hint{color:#ff4514}.ngx-select{position:relative;display:block;min-width:300px;line-height:1.4em;margin-top:16px;margin-bottom:8px;padding-top:20px;padding-top:calc(.7em + 8px);padding-bottom:0}.ngx-select .ngx-select-flex-wrap{-webkit-box-orient:horizontal;-webkit-box-direction:normal;flex-direction:row;display:-webkit-box;display:flex}.ngx-select .ngx-select-flex-wrap .ngx-select-flex-wrap-inner{display:-webkit-box;display:flex;-webkit-box-flex:100%;flex:100%;width:100%}.ngx-select .ngx-select-wrap{position:relative;display:block;margin-bottom:0;width:100%}.ngx-select .ngx-select-wrap .ngx-select-hint{font-size:12px;color:#909cb4;margin-top:2px;line-height:1.4em;min-height:1em}.ngx-select.active .ngx-select-input .ngx-select-input-underline .underline-fill{width:100%}.ngx-select.active .ngx-select-input .ngx-select-caret{-webkit-transition:-webkit-transform .2s ease-in-out;transition:transform .2s ease-in-out;transition:transform .2s ease-in-out,-webkit-transform .2s ease-in-out;-webkit-transform:rotate(180deg) translateY(50%);transform:rotate(180deg) translateY(50%)}.ngx-select.active .ngx-select-dropdown{display:block;opacity:1;-webkit-animation:.25s openAnimation;animation:.25s openAnimation}.ngx-select.disabled .ngx-select-input .ngx-select-input-box{cursor:not-allowed}.ngx-select.active .ngx-select-input .ngx-select-label,.ngx-select.active-selections .ngx-select-input .ngx-select-label,.ngx-select.has-placeholder .ngx-select-input .ngx-select-label{font-size:.7rem;top:-25px;-webkit-transition:150ms ease-out;transition:150ms ease-out;line-height:30px}.ngx-select.multi-selection .ngx-select-input-option,.ngx-select.tagging-selection .ngx-select-input-option{background:#455066;color:#cfcfcf;border-radius:.7em;margin:0 5px;height:1.4em;white-space:nowrap;overflow:visible;text-overflow:ellipsis;cursor:text;-webkit-transition:background .1s ease-in;transition:background .1s ease-in;position:relative;top:0;padding:0 .7em;font-size:.95em}.ngx-select.multi-selection .ngx-select-input-option:hover,.ngx-select.tagging-selection .ngx-select-input-option:hover{background:#1483ff}.ngx-select.multi-selection .ngx-select-input-option .ngx-select-input-name,.ngx-select.tagging-selection .ngx-select-input-option .ngx-select-input-name{text-shadow:2px 4px 2px rgba(0,0,0,.1);max-width:300px;overflow:hidden;display:inline-block;white-space:nowrap;text-overflow:ellipsis;vertical-align:top}.ngx-select.multi-selection .ngx-select-input-option .ngx-select-clear,.ngx-select.tagging-selection .ngx-select-input-option .ngx-select-clear{display:inline;vertical-align:middle;font-size:.8em;color:#cfcfcf;-webkit-transition:color .1s ease-in;transition:color .1s ease-in;cursor:pointer;line-height:1.4em}.ngx-select.multi-selection .ngx-select-input-option .ngx-select-clear:hover,.ngx-select.tagging-selection .ngx-select-input-option .ngx-select-clear:hover{color:#1c2029}.ngx-select.multi-selection .ngx-select-input-option.disabled,.ngx-select.tagging-selection .ngx-select-input-option.disabled{padding-right:10px}.ngx-select.tagging-selection .ngx-select-input .ngx-select-input-box{cursor:text}.ngx-select.tagging-selection .ngx-select-input .ng-select-text-box{background-color:transparent;border:none;outline:0;-webkit-appearance:none;color:#b6b6b6;line-height:20px;display:inline-block;height:20px;vertical-align:bottom;margin-bottom:6px}.ngx-select.tagging-selection .ngx-select-input .ng-select-text-box:focus{outline:0}.ngx-select.single-selection .ngx-select-input .ngx-select-input-list .ngx-select-input-option{width:100%;white-space:nowrap;text-overflow:ellipsis;display:block;color:#cdd2dd;overflow-x:hidden;overflow-y:visible}.ngx-select.single-selection .ngx-select-input .ngx-select-clear{position:absolute;top:1em;right:25px;-webkit-transform:translateY(-50%);transform:translateY(-50%);cursor:pointer;color:#696969;font-size:12px;height:auto;line-height:normal}.ngx-select.single-selection .ngx-select-input .ngx-select-clear:hover{color:#cfcfcf}.ngx-select .ngx-select-input{display:block;position:relative;margin-bottom:0;width:100%}.ngx-select .ngx-select-input .ngx-select-input-box{background:0 0;border:none;margin-bottom:0;padding-left:0;width:100%;min-height:1em;cursor:pointer;display:inline-block;vertical-align:bottom}.ngx-select .ngx-select-input .ngx-select-input-underline{width:100%;height:1px;margin-top:2px;background-color:#5a6884}.ngx-select .ngx-select-input .ngx-select-input-underline .underline-fill{background-color:#1483ff;-webkit-transition:width 250ms ease-out;transition:width 250ms ease-out;width:0;height:2px;margin:0 auto}.ngx-select .ngx-select-input .ngx-select-input-list{margin-right:40px;padding-top:0;color:#b6b6b6;min-height:1.4em}.ngx-select .ngx-select-input .ngx-select-label{pointer-events:none;position:absolute;top:0}.ngx-select .ngx-select-input .ngx-select-caret{position:absolute;top:1em;right:5px;-webkit-transform:translateY(-50%);transform:translateY(-50%);cursor:pointer;color:#696969;font-size:.7em}.ngx-select .ngx-select-input .ngx-select-placeholder{position:absolute;top:0;left:0;cursor:pointer;color:#647493}.ngx-select .ngx-select-dropdown{position:absolute;clear:left;top:100%;left:0;z-index:1000;width:100%;background:#313847;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 4px 0 rgba(0,0,0,.14),0 3px 3px -2px rgba(0,0,0,.12);margin-top:10px;opacity:0;display:none;-webkit-transition:visibility,opacity .25s ease-out;transition:visibility,opacity .25s ease-out}.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options{max-height:300px;overflow-y:auto}.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option{padding:7px 15px;font-size:15px;overflow-x:hidden;overflow-y:visible;text-overflow:ellipsis}.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option:not(.disabled){cursor:pointer}.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option:not(.disabled):not(.selected).active,.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option:not(.disabled):not(.selected):hover{background:#5a6884}.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option.selected{background:#479eff;color:#cfcfcf}.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option.selected:not(.disabled).active,.ngx-select .ngx-select-dropdown .ngx-select-dropdown-options .ngx-select-dropdown-option.selected:not(.disabled):hover{background:#1483ff}.ngx-select .ngx-select-dropdown.groupings .ngx-select-option-group .ngx-select-option-group-name{padding:7px 15px;font-size:18px;display:block;background:#262c38;font-weight:700}.ngx-select .ngx-select-dropdown.groupings .ngx-select-dropdown-options .ngx-select-dropdown-option{padding:7px 25px}.ngx-select .ngx-select-dropdown.groupings .ngx-select-dropdown-options .ngx-select-dropdown-option:last-child{margin-bottom:10px}.ngx-select .ngx-select-dropdown .ngx-select-filter{padding:10px;background:#455066}.ngx-select .ngx-select-dropdown .ngx-select-filter .ngx-select-filter-input{color:#b6b6b6;background:0 0;border:none;outline:0;display:block;font-size:15px;width:100%}.ngx-select .ngx-select-dropdown .ngx-select-empty-placeholder{color:#9c9c9c;padding:7px 15px;font-size:15px;font-style:italic}.ngx-select .ngx-select-dropdown .ngx-select-empty-placeholder .ngx-select-empty-placeholder-add{font-style:normal;float:right}.ngx-select.invalid .ngx-select-label{font-weight:600;color:#ff4514}.ngx-select.invalid .ngx-select-caret,.ngx-select.invalid .ngx-select-hint{color:#ff4514}.ngx-select.invalid .ngx-select-input-underline{background-color:#ff4514}@-webkit-keyframes openAnimation{0%{-webkit-transform:translateY(-25px);transform:translateY(-25px)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}@keyframes openAnimation{0%{-webkit-transform:translateY(-25px);transform:translateY(-25px)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}"]
    }),
    __metadata("design:paramtypes", [ElementRef,
        Renderer2,
        ChangeDetectorRef])
], SelectComponent);
export { SelectComponent };
