// @swimlane/ngx-ui/esm5/lib/components/dialog/dialog.service.js
import { __decorate, __extends, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { InjectionService } from '../../services/injection/injection.service';
import { InjectionRegistryService } from '../../services/injection-registry/injection-registry.service';
import { OverlayService } from '../overlay/overlay.service';
import { DialogComponent } from './dialog.component';
var DialogService = /** @class */ (function (_super) {
    __extends(DialogService, _super);
    function DialogService(injectionService, overlayService) {
        var _this = _super.call(this, injectionService) || this;
        _this.injectionService = injectionService;
        _this.overlayService = overlayService;
        _this.defaults = {
            inputs: {
                zIndex: 991,
                closeOnBlur: true,
                closeOnEscape: true,
                closeButton: true,
                showOverlay: true,
                visible: true
            }
        };
        _this.type = DialogComponent;
        _this.zIndex = 995;
        return _this;
    }
    DialogService.prototype.create = function (options) {
        var component = _super.prototype.create.call(this, options);
        this.createSubscriptions(component);
        return component;
    };
    DialogService.prototype.destroy = function (component) {
        var _this = this;
        var hasOverlay = component.instance.showOverlay;
        this.zIndex = this.zIndex - 2;
        if (hasOverlay) {
            this.overlayService.removeTriggerComponent(component);
        }
        setTimeout(function () {
            _super.prototype.destroy.call(_this, component);
        });
    };
    DialogService.prototype.createSubscriptions = function (triggerComponent) {
        var _this = this;
        var closeSub;
        var overlaySub;
        var kill = function (c) {
            /* istanbul ignore if */
            if (c !== triggerComponent) {
                return;
            }
            closeSub.unsubscribe();
            if (overlaySub)
                overlaySub.unsubscribe();
            _this.destroy(triggerComponent);
        };
        closeSub = triggerComponent.instance.close.subscribe(kill.bind(this, triggerComponent));
        var zIndex = this.zIndex;
        if (triggerComponent.instance.showOverlay) {
            setTimeout(function () {
                _this.overlayService.show({
                    triggerComponent: triggerComponent,
                    zIndex: zIndex
                });
                if (triggerComponent.instance.closeOnBlur) {
                    overlaySub = _this.overlayService.click.subscribe(kill);
                }
            });
        }
    };
    DialogService.prototype.assignDefaults = function (options) {
        options = _super.prototype.assignDefaults.call(this, options);
        /* istanbul ignore else */
        if (!options.zIndex) {
            this.zIndex = this.overlayService.instance ? this.overlayService.instance.zIndex + 3 : this.zIndex + 2;
            options.inputs.zIndex = this.zIndex;
        }
        return options;
    };
    DialogService.ctorParameters = function () { return [
        { type: InjectionService },
        { type: OverlayService }
    ]; };
    DialogService = __decorate([
        Injectable(),
        __metadata("design:paramtypes", [InjectionService, OverlayService])
    ], DialogService);
    return DialogService;
}(InjectionRegistryService));
export { DialogService };
