// @swimlane/ngx-ui/esm5/lib/directives/long-press/long-press.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, Output, EventEmitter, HostListener } from '@angular/core';
import { coerceNumberProperty, coerceBooleanProperty } from '@angular/cdk/coercion';
// tslint:disable-next-line:directive-selector
var LongPressDirective = /** @class */ (function () {
    function LongPressDirective() {
        this.longPressStart = new EventEmitter();
        this.longPressFinish = new EventEmitter();
        this.longPressCancel = new EventEmitter();
        this._duration = 3000;
        this._disabled = false;
        this._pressed = false;
    }
    Object.defineProperty(LongPressDirective.prototype, "duration", {
        get: function () {
            return this._duration;
        },
        set: function (duration) {
            this._duration = coerceNumberProperty(duration);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LongPressDirective.prototype, "disabled", {
        get: function () {
            return this._disabled;
        },
        set: function (disabled) {
            this._disabled = coerceBooleanProperty(disabled);
        },
        enumerable: true,
        configurable: true
    });
    LongPressDirective.prototype.onPress = function (event) {
        var _this = this;
        if (this.disabled) {
            event.stopPropagation();
            event.preventDefault();
            return;
        }
        this._pressed = true;
        this.longPressStart.emit(true);
        this._pressTimeout = setTimeout(function () {
            /* istanbul ignore else */
            if (_this._pressed) {
                _this._pressed = false;
                _this.longPressFinish.emit(true);
            }
        }, this.duration);
    };
    LongPressDirective.prototype.onRelease = function () {
        this._pressed = false;
        clearTimeout(this._pressTimeout);
        this.longPressCancel.emit(true);
    };
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], LongPressDirective.prototype, "duration", null);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], LongPressDirective.prototype, "disabled", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], LongPressDirective.prototype, "longPressStart", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], LongPressDirective.prototype, "longPressFinish", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], LongPressDirective.prototype, "longPressCancel", void 0);
    __decorate([
        HostListener('mousedown', ['$event']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [MouseEvent]),
        __metadata("design:returntype", void 0)
    ], LongPressDirective.prototype, "onPress", null);
    __decorate([
        HostListener('mouseout'),
        HostListener('mouseup'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], LongPressDirective.prototype, "onRelease", null);
    LongPressDirective = __decorate([
        Directive({
            selector: '[long-press]'
        })
    ], LongPressDirective);
    return LongPressDirective;
}());
export { LongPressDirective };
