// @swimlane/ngx-ui/esm5/lib/components/overlay/overlay.component.js
import { __decorate, __metadata } from "tslib";
import { animate, state, style, transition, trigger } from '@angular/animations';
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, EventEmitter, Input, Output, ViewEncapsulation } from '@angular/core';
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
/**
 * Overlay Component for Drawer/Dialogs
 */
var OverlayComponent = /** @class */ (function () {
    function OverlayComponent(cdr) {
        this.cdr = cdr;
        this.click = new EventEmitter();
        this._visible = false;
        this._zIndex = 990;
    }
    Object.defineProperty(OverlayComponent.prototype, "visible", {
        get: function () {
            return this._visible;
        },
        set: function (val) {
            this._visible = coerceBooleanProperty(val);
            this.cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OverlayComponent.prototype, "zIndex", {
        get: function () {
            return this._zIndex;
        },
        set: function (val) {
            this._zIndex = coerceNumberProperty(val);
            this.cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(OverlayComponent.prototype, "animationState", {
        get: function () {
            return this.visible ? 'active' : 'inactive';
        },
        enumerable: true,
        configurable: true
    });
    OverlayComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], OverlayComponent.prototype, "visible", null);
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], OverlayComponent.prototype, "zIndex", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], OverlayComponent.prototype, "click", void 0);
    OverlayComponent = __decorate([
        Component({
            exportAs: 'ngxOverlay',
            selector: 'ngx-overlay',
            template: "<div (click)=\"click.emit(true)\" [style.zIndex]=\"zIndex\" [@overlayTransition]=\"animationState\" class=\"ngx-overlay\">\n  <ng-content></ng-content>\n</div>\n",
            animations: [
                trigger('overlayTransition', [
                    state('active', style({
                        opacity: 0.8,
                        visibility: 'visible'
                    })),
                    state('inactive', style({
                        visibility: 'hidden',
                        opacity: 0
                    })),
                    transition('* => active', [animate('100ms ease-in')]),
                    transition('* => inactive', [animate('100ms ease-out')]),
                    transition('* => void', [
                        style({
                            opacity: 0,
                            visibility: 'hidden',
                            'pointer-events': 'none'
                        }),
                        animate('100ms ease-out')
                    ])
                ])
            ],
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-overlay{position:fixed;top:0;left:0;bottom:0;right:0;width:100%;overflow:hidden;height:100%;background-color:#000;opacity:0;visibility:hidden;text-align:center;line-height:100vh}"]
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], OverlayComponent);
    return OverlayComponent;
}());
export { OverlayComponent };
