// @swimlane/ngx-ui/esm2015/lib/components/dropdown/dropdown-toggle.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Input, ElementRef, HostListener, EventEmitter, Output } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
let DropdownToggleDirective = class DropdownToggleDirective {
    constructor(el) {
        this.el = el;
        this.showEvent = "click" /* Click */;
        this.toggle = new EventEmitter();
        this._disabled = false;
        this.element = this.el.nativeElement;
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(disabled) {
        this._disabled = coerceBooleanProperty(disabled);
    }
    onClick(event) {
        if (this.showEvent === "click" /* Click */ && !this.disabled) {
            event.preventDefault();
            this.toggle.emit(event);
        }
    }
    onContextmenu(event) {
        if (this.showEvent === "contextmenu" /* Contextmenu */ && !this.disabled) {
            event.preventDefault();
            this.toggle.emit(event);
        }
    }
    onDblclick(event) {
        if (this.showEvent === "dblclick" /* Dblclick */ && !this.disabled) {
            event.preventDefault();
            this.toggle.emit(event);
        }
    }
};
DropdownToggleDirective.ctorParameters = () => [
    { type: ElementRef }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], DropdownToggleDirective.prototype, "showEvent", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], DropdownToggleDirective.prototype, "disabled", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], DropdownToggleDirective.prototype, "toggle", void 0);
__decorate([
    HostListener('click', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], DropdownToggleDirective.prototype, "onClick", null);
__decorate([
    HostListener('contextmenu', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], DropdownToggleDirective.prototype, "onContextmenu", null);
__decorate([
    HostListener('dblclick', ['$event']),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Event]),
    __metadata("design:returntype", void 0)
], DropdownToggleDirective.prototype, "onDblclick", null);
DropdownToggleDirective = __decorate([
    Directive({
        // tslint:disable-next-line:directive-selector
        exportAs: 'ngxDropdownToggle',
        selector: 'ngx-dropdown-toggle',
        host: {
            class: 'ngx-dropdown-toggle',
            '[class.disabled]': 'disabled'
        }
    }),
    __metadata("design:paramtypes", [ElementRef])
], DropdownToggleDirective);
export { DropdownToggleDirective };
