// @swimlane/ngx-ui/esm5/lib/components/loading/loading.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ViewEncapsulation, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
var LoadingComponent = /** @class */ (function () {
    function LoadingComponent(cdr) {
        this.cdr = cdr;
        this._visible = false;
        this._progress = 0;
    }
    Object.defineProperty(LoadingComponent.prototype, "visible", {
        get: function () {
            return this._visible;
        },
        set: function (visible) {
            this._visible = coerceBooleanProperty(visible);
            this.cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LoadingComponent.prototype, "progress", {
        get: function () {
            return this._progress;
        },
        set: function (progress) {
            this._progress = coerceNumberProperty(progress);
            this.cdr.markForCheck();
        },
        enumerable: true,
        configurable: true
    });
    LoadingComponent.ctorParameters = function () { return [
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], LoadingComponent.prototype, "visible", null);
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], LoadingComponent.prototype, "progress", null);
    LoadingComponent = __decorate([
        Component({
            exportAs: 'ngxLoadingBar',
            selector: 'ngx-loading-bar',
            template: "<div class=\"ngx-loading-bar--container\" *ngIf=\"visible\">\n  <div class=\"ngx-loading-bar--bar\" [style.width.%]=\"progress\"></div>\n</div>\n",
            host: { class: 'ngx-loading-bar' },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-loading-bar{display:block}.ngx-loading-bar .ngx-loading-bar--container{position:fixed;margin:0;padding:0;top:0;left:0;right:0;z-index:99999}.ngx-loading-bar .ngx-loading-bar--container .ngx-loading-bar--bar{background-color:#7ab9ff;color:#7ab9ff;opacity:1;z-index:99998;box-shadow:0 0 10px 0;height:2px;-webkit-transition:.5s ease-in-out;transition:.5s ease-in-out}"]
        }),
        __metadata("design:paramtypes", [ChangeDetectorRef])
    ], LoadingComponent);
    return LoadingComponent;
}());
export { LoadingComponent };
