// @swimlane/ngx-ui/esm5/lib/components/section/section.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ContentChild, Output, EventEmitter, ViewEncapsulation, ChangeDetectionStrategy } from '@angular/core';
import { SectionHeaderComponent } from './section-header.component';
var SectionComponent = /** @class */ (function () {
    function SectionComponent() {
        this.sectionCollapsed = false;
        this.sectionCollapsible = true;
        this.padding = '1.8em';
        this.toggle = new EventEmitter();
    }
    SectionComponent.prototype.onSectionClicked = function () {
        this.sectionCollapsed = !this.sectionCollapsed;
        this.toggle.emit(this.sectionCollapsed);
    };
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], SectionComponent.prototype, "sectionCollapsed", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], SectionComponent.prototype, "sectionCollapsible", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], SectionComponent.prototype, "sectionTitle", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], SectionComponent.prototype, "padding", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], SectionComponent.prototype, "toggle", void 0);
    __decorate([
        ContentChild(SectionHeaderComponent),
        __metadata("design:type", SectionHeaderComponent)
    ], SectionComponent.prototype, "headerComp", void 0);
    SectionComponent = __decorate([
        Component({
            selector: 'ngx-section',
            exportAs: 'ngxSection',
            template: "<section>\n  <header\n    [class.ngx-section-collapsible]=\"sectionCollapsible\"\n    class=\"ngx-section-header\"\n    *ngIf=\"headerComp || sectionTitle\"\n  >\n    <button\n      *ngIf=\"sectionCollapsible\"\n      class=\"ngx-section-toggle\"\n      (click)=\"onSectionClicked()\"\n      type=\"button\"\n      title=\"Toggle Content Visibility\"\n    >\n      <span [class.icon-arrow-down]=\"!sectionCollapsed\" [class.icon-arrow-right]=\"sectionCollapsed\"> </span>\n    </button>\n    <ng-content select=\"ngx-section-header\"></ng-content>\n    <h1 *ngIf=\"sectionTitle\" [innerHTML]=\"sectionTitle\"></h1>\n  </header>\n  <div class=\"ngx-section-content\" [style.padding]=\"padding\" *ngIf=\"!sectionCollapsed\">\n    <ng-content></ng-content>\n  </div>\n</section>\n",
            host: {
                class: 'ngx-section'
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-section{display:block;margin-bottom:2em;background:#171a21}.ngx-section .ngx-section-header{background:#212631;display:block;width:auto;height:40px;line-height:40px;padding:0 10px;color:#cdd2dd;position:relative}.ngx-section .ngx-section-header.ngx-section-collapsible{padding-left:30px}.ngx-section .ngx-section-header .ngx-section-toggle{font-size:14px;vertical-align:top;display:inline-block;line-height:40px;position:absolute;background:0 0;border:none;box-shadow:none;padding:0;left:10px;cursor:pointer}.ngx-section .ngx-section-header a{color:#cdd2dd}.ngx-section .ngx-section-header h1{font-size:1.1em;font-weight:600;line-height:40px;padding:0;margin:0}.ngx-section.shadow>section{box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 4px 0 rgba(0,0,0,.14),0 3px 3px -2px rgba(0,0,0,.12)}"]
        })
    ], SectionComponent);
    return SectionComponent;
}());
export { SectionComponent };
