// @swimlane/ngx-ui/esm5/lib/components/notification/notification.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, HostListener, HostBinding, ViewEncapsulation, TemplateRef, ChangeDetectionStrategy } from '@angular/core';
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
import { NotificationStyleType } from './notification-style-type.enum';
var NotificationComponent = /** @class */ (function () {
    function NotificationComponent() {
        this.cssClass = '';
        this.close = new EventEmitter();
        this.pause = new EventEmitter();
        this.resume = new EventEmitter();
        this.NotificationStyleType = NotificationStyleType;
    }
    Object.defineProperty(NotificationComponent.prototype, "showClose", {
        get: function () {
            return this._showClose;
        },
        set: function (showClose) {
            this._showClose = coerceBooleanProperty(showClose);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NotificationComponent.prototype, "pauseOnHover", {
        get: function () {
            return this._pauseOnHover;
        },
        set: function (pauseOnHover) {
            this._pauseOnHover = coerceBooleanProperty(pauseOnHover);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NotificationComponent.prototype, "timestamp", {
        get: function () {
            return this._timestamp;
        },
        set: function (timestamp) {
            this._timestamp = coerceNumberProperty(timestamp);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NotificationComponent.prototype, "cssClasses", {
        get: function () {
            var cls = "ngx-notification ngx-notification-" + this.styleType;
            if (this.cssClass)
                cls += " " + this.cssClass;
            if (this.showClose)
                cls += ' notification-closeable';
            return cls;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NotificationComponent.prototype, "animationDuration", {
        get: function () {
            if (typeof this.timeout !== 'number') {
                return '3000s';
            }
            return this.timeout + "ms";
        },
        enumerable: true,
        configurable: true
    });
    NotificationComponent.prototype.onMouseEnter = function () {
        if (this.pauseOnHover) {
            this.pause.emit();
        }
    };
    NotificationComponent.prototype.onMouseLeave = function () {
        if (this.pauseOnHover) {
            this.resume.emit();
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NotificationComponent.prototype, "cssClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NotificationComponent.prototype, "title", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NotificationComponent.prototype, "body", void 0);
    __decorate([
        Input(),
        __metadata("design:type", TemplateRef)
    ], NotificationComponent.prototype, "template", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NotificationComponent.prototype, "styleType", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NotificationComponent.prototype, "icon", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NotificationComponent.prototype, "timeout", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], NotificationComponent.prototype, "showClose", null);
    __decorate([
        Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], NotificationComponent.prototype, "pauseOnHover", null);
    __decorate([
        Input(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [Object])
    ], NotificationComponent.prototype, "timestamp", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NotificationComponent.prototype, "close", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NotificationComponent.prototype, "pause", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NotificationComponent.prototype, "resume", void 0);
    __decorate([
        HostBinding('class'),
        __metadata("design:type", String),
        __metadata("design:paramtypes", [])
    ], NotificationComponent.prototype, "cssClasses", null);
    __decorate([
        HostListener('mouseenter'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NotificationComponent.prototype, "onMouseEnter", null);
    __decorate([
        HostListener('mouseleave'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], NotificationComponent.prototype, "onMouseLeave", null);
    NotificationComponent = __decorate([
        Component({
            exportAs: 'ngxNotification',
            selector: 'ngx-notification',
            template: "<div>\n  <div class=\"notification-bar\" [class.animated]=\"timeout\">\n    <div class=\"bar\" [ngStyle]=\"{'animation-duration': animationDuration }\"></div>\n  </div>\n\n  <div *ngIf=\"styleType !== NotificationStyleType.none && !icon\" class=\"icon-container\">\n    <span *ngIf=\"styleType === NotificationStyleType.info\" class=\"icon icon-info-fulled\"></span>\n    <span *ngIf=\"styleType === NotificationStyleType.warning\" class=\"icon icon-warning-filled\"></span>\n    <span *ngIf=\"styleType === NotificationStyleType.error\" class=\"icon icon-x-filled\"></span>\n    <span *ngIf=\"styleType === NotificationStyleType.success\" class=\"icon icon-check-filled\"></span>\n  </div>\n\n  <div *ngIf=\"styleType === NotificationStyleType.none && icon\" class=\"icon-container\"><span [class]=\"'icon ' + icon\"></span></div>\n\n  <div class=\"notification-content\" [class.has-icon]=\"styleType !== NotificationStyleType.none || icon\">\n    <h2 class=\"ngx-notification-title\" [innerHTML]=\"title\"></h2>\n    <p class=\"ngx-notification-body\" [innerHTML]=\"body\"></p>\n    <ng-template *ngIf=\"template\" [ngTemplateOutlet]=\"template\"> </ng-template>\n    <button *ngIf=\"showClose\" type=\"button\" (click)=\"close.emit()\" class=\"icon-x ngx-notification-close\"></button>\n  </div>\n</div>\n\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-notification-container{position:fixed;bottom:20px;right:20px;min-width:120px;max-width:400px;z-index:9998}.ngx-notification{display:block;margin-top:.5rem;position:relative;box-shadow:0 8px 9px -5px rgba(0,0,0,.2),0 15px 22px 2px rgba(0,0,0,.14),0 6px 28px 5px rgba(0,0,0,.12);color:#fff;text-shadow:1px 1px rgba(0,0,0,.1);background:#313847}.ngx-notification .notification-bar{background-color:rgba(69,80,102,.2);width:5px;height:100%;position:absolute}.ngx-notification .notification-bar .bar{background-color:#455066;width:5px;height:100%;position:absolute;bottom:0}.ngx-notification .icon-container .icon{color:#455066;font-size:2.5rem;line-height:2.5rem}.ngx-notification .notification-bar.animated .bar{height:0%;-webkit-animation:1s ease-out heightBar;animation:1s ease-out heightBar}@-webkit-keyframes heightBar{0%{height:100%}100%{height:0%}}@keyframes heightBar{0%{height:100%}100%{height:0%}}.ngx-notification .icon-container{float:left;position:absolute;top:50%;-webkit-transform:translateY(-50%);transform:translateY(-50%);padding:0 1em;height:2.5rem}.ngx-notification .notification-content{padding:1.5rem;color:#cdd2dd;overflow:auto;max-height:300px}.ngx-notification .notification-content.has-icon{margin-left:4.5rem;border-left:1px solid #3b4457}.ngx-notification.notification-closeable{padding-right:1rem}.ngx-notification.ngx-notification-info .notification-bar{background-color:rgba(20,131,255,.2)}.ngx-notification.ngx-notification-info .notification-bar .bar{background-color:#1483ff}.ngx-notification.ngx-notification-info .icon-container .icon{color:#1483ff}.ngx-notification.ngx-notification-error .notification-bar{background-color:rgba(255,69,20,.2)}.ngx-notification.ngx-notification-error .notification-bar .bar{background-color:#ff4514}.ngx-notification.ngx-notification-error .icon-container .icon{color:#ff4514}.ngx-notification.ngx-notification-warning .notification-bar{background-color:rgba(255,168,20,.2)}.ngx-notification.ngx-notification-warning .notification-bar .bar{background-color:#ffa814}.ngx-notification.ngx-notification-warning .icon-container .icon{color:#ffa814}.ngx-notification.ngx-notification-success .notification-bar{background-color:rgba(29,222,182,.2)}.ngx-notification.ngx-notification-success .notification-bar .bar{background-color:#1ddeb6}.ngx-notification.ngx-notification-success .icon-container .icon{color:#1ddeb6}.ngx-notification .ngx-notification-title{margin:0;font-size:1.2rem;color:#fff}.ngx-notification .ngx-notification-body{font-size:.9rem;margin:0;line-height:1rem}.ngx-notification .ngx-notification-close{position:absolute;right:.5rem;top:.5rem;line-height:1em;padding:0;font-size:16px;height:16px}"]
        })
    ], NotificationComponent);
    return NotificationComponent;
}());
export { NotificationComponent };
