// @swimlane/ngx-ui/esm5/lib/components/notification/notification.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { InjectionService } from '../../services/injection/injection.service';
import { NotificationComponent } from './notification.component';
import { NotificationService } from './notification.service';
import { NotificationContainerComponent } from './notification-container.component';
var NotificationModule = /** @class */ (function () {
    function NotificationModule() {
    }
    NotificationModule = __decorate([
        NgModule({
            declarations: [NotificationComponent, NotificationContainerComponent],
            exports: [NotificationComponent, NotificationContainerComponent],
            providers: [NotificationService, InjectionService],
            imports: [CommonModule],
            entryComponents: [NotificationComponent, NotificationContainerComponent]
        })
    ], NotificationModule);
    return NotificationModule;
}());
export { NotificationModule };
