// @swimlane/ngx-ui/esm5/lib/components/hotkeys/hotkeys.component.js
import { __decorate, __metadata, __values } from "tslib";
import { trigger } from '@angular/animations';
import { Component, ChangeDetectionStrategy } from '@angular/core';
import { Subject, BehaviorSubject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { fadeInAnimation, slideDownFadeOutAnimation } from '../../animations';
import { HotkeysService } from './hotkeys.service';
import { HotkeyStatus } from './hotkey-status.enum';
var HotkeysComponent = /** @class */ (function () {
    function HotkeysComponent(_hotkeysService) {
        this._hotkeysService = _hotkeysService;
        this.hotkeys$ = new BehaviorSubject([]);
        this.visible = false;
        this._destroy = new Subject();
    }
    HotkeysComponent.prototype.ngOnInit = function () {
        this._hotkeysService.changeEvent.pipe(takeUntil(this._destroy)).subscribe(this.updateHotkeys.bind(this));
        this.updateHotkeys(this._hotkeysService.hotkeys);
    };
    HotkeysComponent.prototype.ngOnDestroy = function () {
        this._destroy.next();
        this._destroy.complete();
    };
    HotkeysComponent.prototype.updateHotkeys = function (hotkeys) {
        var e_1, _a;
        var hks = [];
        for (var comb in hotkeys) {
            try {
                for (var _b = (e_1 = void 0, __values(hotkeys[comb])), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var hotkey = _c.value;
                    if (hotkey.status === HotkeyStatus.Active && hotkey.visible) {
                        hks.push(hotkey);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        this.hotkeys$.next(hks);
    };
    HotkeysComponent.prototype.show = function () {
        this.visible = true;
    };
    HotkeysComponent.prototype.hide = function () {
        this.visible = false;
    };
    HotkeysComponent.ctorParameters = function () { return [
        { type: HotkeysService }
    ]; };
    HotkeysComponent = __decorate([
        Component({
            exportAs: 'ngxHotkeys',
            selector: 'ngx-hotkeys',
            template: "<div class=\"hotkeys-container\" *ngIf=\"(hotkeys$ | async).length > 0\">\n  <div class=\"hotkeys\" *ngIf=\"visible\" [@containerAnimationState]=\"'active'\">\n    <div *ngFor=\"let hotkey of hotkeys$ | async\" class=\"hotkey-row\">\n      {{ hotkey.description }}\n      <div class=\"combination\">\n        <span *ngFor=\"let key of hotkey.keys; let i = index\">\n          <span class=\"key\">{{ key }}</span> <span *ngIf=\"i < hotkey.keys.length - 1\"> + </span>\n        </span>\n      </div>\n    </div>\n  </div>\n  <div\n    class=\"close-icon icon icon-x-filled\"\n    *ngIf=\"visible\"\n    (click)=\"hide()\"\n    [@iconAnimationState]=\"'active'\"\n  ></div>\n  <div\n    class=\"hotkeys-icon icon icon-keyboard\"\n    *ngIf=\"!visible\"\n    (click)=\"show()\"\n    [@iconAnimationState]=\"'active'\"\n  ></div>\n</div>\n",
            animations: [
                trigger('containerAnimationState', slideDownFadeOutAnimation()),
                trigger('iconAnimationState', fadeInAnimation())
            ],
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".hotkeys-container{position:fixed;bottom:20px;right:20px;z-index:2000}.hotkeys-container .icon{cursor:pointer;width:40px;height:40px;line-height:43px;text-align:center;border-radius:20px;position:absolute;bottom:0;right:0}.hotkeys-container .close-icon{font-size:40px;color:#909cb4;float:right;line-height:1em}.hotkeys-container .hotkeys-icon{color:#909cb4;font-size:14px;background-color:#000}.hotkeys-container .icon:hover{color:#cdd2dd}.hotkeys{background:#12141a;border:1px solid #5a6884;padding:15px;box-shadow:0 1px 3px 0 rgba(0,0,0,.2),0 1px 1px 0 rgba(0,0,0,.14),0 2px 1px -1px rgba(0,0,0,.12);color:#909cb4;font-size:14px;border-radius:2px;margin-bottom:45px}.hotkeys .hotkey-row{padding:10px 0;border-bottom:1px solid #455066}.hotkeys .hotkey-row:first-child{padding-top:0}.hotkeys .hotkey-row:last-child{border-bottom:none;padding-bottom:0}.hotkeys .hotkey-row .combination{float:right;margin-left:30px}.hotkeys .hotkey-row .combination .key{background:#5a6884;color:#12141a;padding:2px 7px;border-radius:2px}"]
        }),
        __metadata("design:paramtypes", [HotkeysService])
    ], HotkeysComponent);
    return HotkeysComponent;
}());
export { HotkeysComponent };
