// @swimlane/ngx-ui/esm5/lib/components/input/input-autosize.directive.js
import { __decorate, __metadata } from "tslib";
import { ElementRef, HostListener, Directive, AfterContentChecked, Renderer2 } from '@angular/core';
var AutosizeDirective = /** @class */ (function () {
    function AutosizeDirective(element, renderer) {
        this.element = element;
        this.renderer = renderer;
    }
    AutosizeDirective.prototype.ngAfterContentChecked = function () {
        this.adjust();
    };
    AutosizeDirective.prototype.onInput = function (_) {
        this.adjust();
    };
    AutosizeDirective.prototype.adjust = function () {
        var height = this.element.nativeElement.scrollHeight + 'px';
        this.renderer.setStyle(this.element.nativeElement, 'overflow', 'hidden');
        this.renderer.setStyle(this.element.nativeElement, 'height', height);
    };
    AutosizeDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 }
    ]; };
    __decorate([
        HostListener('input', ['$event.target']),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [HTMLTextAreaElement]),
        __metadata("design:returntype", void 0)
    ], AutosizeDirective.prototype, "onInput", null);
    AutosizeDirective = __decorate([
        Directive({
            // tslint:disable-next-line:directive-selector
            exportAs: 'ngxAutosize',
            selector: 'textarea[autosize]'
        }),
        __metadata("design:paramtypes", [ElementRef, Renderer2])
    ], AutosizeDirective);
    return AutosizeDirective;
}());
export { AutosizeDirective };
