// @wkoza/ngx-upload/esm5/components/inputfile.component.js
import * as tslib_1 from "tslib";
import { Component, Injector, Optional, Renderer2, ViewChild } from '@angular/core';
import { FormGroupDirective, NgForm } from '@angular/forms';
import { HttpClientUploadService } from '../services/httpClientUpload.service';
var InputfileComponent = /** @class */ (function () {
    function InputfileComponent(injector, uploader, renderer, ngForm, formGroupDirective) {
        this.injector = injector;
        this.uploader = uploader;
        this.renderer = renderer;
        this.ngForm = ngForm;
        this.formGroupDirective = formGroupDirective;
        this.files = new Set();
        if (this.ngForm) {
            this.formGroup = ngForm.form;
        }
        else if (this.formGroupDirective) {
            this.formGroup = formGroupDirective.form;
        }
        else {
            this.formGroup = null;
        }
    }
    InputfileComponent.prototype.onFilesAdded = function () {
        this.uploader.addToQueue(this.file.nativeElement.files, this.formGroup, this.options);
        // Clear the previous input value
        this.file.nativeElement.value = '';
    };
    InputfileComponent.prototype.ngAfterViewInit = function () {
        if (this.options.multiple !== false)
            this.renderer.setProperty(this.file.nativeElement, 'multiple', 'multiple');
        if (this.options.accept)
            this.renderer.setProperty(this.file.nativeElement, 'accept', this.options.accept.join());
        if (this.options.capture)
            this.renderer.setProperty(this.file.nativeElement, 'capture', this.options.capture);
    };
    tslib_1.__decorate([
        ViewChild('file', { static: true }),
        tslib_1.__metadata("design:type", Object)
    ], InputfileComponent.prototype, "file", void 0);
    InputfileComponent = tslib_1.__decorate([
        Component({
            selector: 'ngx-upload-inputfile',
            template: "\n      <label class=\"input-file\">\n          <input type=\"file\" #file (change)=\"onFilesAdded()\">\n          <ng-content></ng-content>\n      </label>",
            styles: ['input[type="file"] { display: none; } .input-file { width: 100%; }']
        }),
        tslib_1.__param(3, Optional()), tslib_1.__param(4, Optional()),
        tslib_1.__metadata("design:paramtypes", [Injector, HttpClientUploadService, Renderer2,
            NgForm, FormGroupDirective])
    ], InputfileComponent);
    return InputfileComponent;
}());
export { InputfileComponent };
