// @swimlane/ngx-ui/esm5/lib/components/json-editor/json-editor-node.js
import { __assign, __decorate, __metadata, __read, __spread, __values } from "tslib";
import { Input, EventEmitter, Output, TemplateRef, ViewChild } from '@angular/core';
import { createValueForSchema, inferType } from './json-editor.helper';
var JsonEditorNode = /** @class */ (function () {
    function JsonEditorNode(dialogMngr) {
        this.dialogMngr = dialogMngr;
        this.required = false;
        this.inline = false;
        this.path = '';
        this.modelChange = new EventEmitter();
        this.schemaChange = new EventEmitter();
        this.requiredCache = {};
        this.expanded = true;
        this.valid = true;
        this.childrenValid = true;
        this.editorConfig = {
            lineNumbers: true,
            theme: 'dracula',
            mode: {
                label: 'Javascript',
                name: 'javascript',
                json: true
            }
        };
        this.editorModel = '';
        this.editorVisible = true;
        this.editorModes = [
            {
                label: 'Javascript',
                name: 'javascript',
                json: true
            },
            {
                label: 'YAML',
                name: 'yaml'
            },
            {
                label: 'Python',
                name: 'python'
            },
            {
                label: 'Powershell',
                name: 'powershell'
            },
            {
                label: 'HTML',
                name: 'htmlmixed'
            }
        ];
    }
    JsonEditorNode.prototype.ngOnInit = function () {
        var e_1, _a;
        var _this = this;
        if (!this.schema) {
            this.schema = __assign({}, inferType(this.model, this.typeCheckOverrides));
        }
        if (this.schema.required) {
            try {
                for (var _b = __values(this.schema.required), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var prop = _c.value;
                    this.requiredCache[prop] = true;
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        if (Array.isArray(this.schema.type) && this.schema.type.length > 0) {
            if (!this.schema.$meta) {
                this.schema.$meta = {};
            }
            this.schema.$meta.type = __spread(this.schema.type);
            if (this.model !== undefined) {
                this.schema = __assign(__assign({}, this.schema), inferType(this.model, this.typeCheckOverrides, this.schema.$meta.type));
            }
            else {
                this.schema.type = this.schema.type[0];
                this.schema.$meta.currentType = this.schema.type;
            }
        }
        setTimeout(function () {
            _this.initModel();
        });
    };
    JsonEditorNode.prototype.ngOnChanges = function (changes) {
        if (changes.errors) {
            this.processErrors();
        }
    };
    /**
     * Inits the model if it is not defined
     */
    JsonEditorNode.prototype.initModel = function () {
        if (this.model !== undefined) {
            return;
        }
        if (!this.schema) {
            return;
        }
        var value = createValueForSchema(this.schema);
        if (value !== undefined) {
            this.updateModel(value);
        }
    };
    /**
     * Process the errors input to figure out whether it or any of its children are invalid
     */
    JsonEditorNode.prototype.processErrors = function () {
        var _this = this;
        this.ownErrors = [];
        this.childrenErrors = [];
        if (this.errors && this.errors.length) {
            this.ownErrors = this.errors.filter(function (e) {
                return e.dataPath === _this.path;
            });
            this.childrenErrors = this.errors.filter(function (e) {
                return e.dataPath.startsWith(_this.path);
            });
        }
        this.childrenValid = this.childrenErrors.length === 0;
        this.valid = this.ownErrors.length === 0;
    };
    /**
     * Updates the whole model and emits the change event
     * @param value
     */
    JsonEditorNode.prototype.updateModel = function (value) {
        this.model = value;
        this.modelChange.emit(this.model);
    };
    /**
     * Expand click event
     */
    JsonEditorNode.prototype.onExpandClick = function () {
        this.expanded = !this.expanded;
    };
    /**
     * Opens the code editor dialog
     */
    JsonEditorNode.prototype.openCodeEditor = function () {
        this.editorModel = this.model;
        this.editorDialog = this.dialogMngr.create({ template: this.codeEditorTpl, class: 'code-editor-dialog' });
    };
    /**
     * Closes the code editor dialog
     */
    JsonEditorNode.prototype.closeCodeEditor = function () {
        this.dialogMngr.destroy(this.editorDialog);
    };
    /**
     * Sets the editor mode and refreshes the editor
     */
    JsonEditorNode.prototype.selectEditorMode = function (modeName) {
        var _this = this;
        this.editorConfig.mode.name = modeName;
        this.editorConfig = __assign({}, this.editorConfig);
        this.editorVisible = false;
        setTimeout(function () {
            _this.editorVisible = true;
        });
    };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], JsonEditorNode.prototype, "schema", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], JsonEditorNode.prototype, "model", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], JsonEditorNode.prototype, "required", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], JsonEditorNode.prototype, "inline", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], JsonEditorNode.prototype, "path", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], JsonEditorNode.prototype, "errors", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], JsonEditorNode.prototype, "typeCheckOverrides", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], JsonEditorNode.prototype, "modelChange", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], JsonEditorNode.prototype, "schemaChange", void 0);
    __decorate([
        ViewChild('codeEditorTpl'),
        __metadata("design:type", TemplateRef)
    ], JsonEditorNode.prototype, "codeEditorTpl", void 0);
    return JsonEditorNode;
}());
export { JsonEditorNode };
