// @swimlane/ngx-ui/esm5/lib/components/json-editor/json-editor-flat/orderable-inputs-list/orderable-inputs-list.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ViewEncapsulation, Output, EventEmitter, ChangeDetectionStrategy } from '@angular/core';
import { moveItemInArray } from '@angular/cdk/drag-drop';
import { debounceable } from '../../../../utils';
var OrderableInputsListComponent = /** @class */ (function () {
    function OrderableInputsListComponent() {
        this.onUpdate = new EventEmitter();
    }
    OrderableInputsListComponent.prototype.ngOnInit = function () {
        this.data = this.data || [];
    };
    OrderableInputsListComponent.prototype.addExample = function () {
        this.data.push('');
    };
    OrderableInputsListComponent.prototype.removeItem = function (index) {
        this.data.splice(index, 1);
        this.update();
    };
    OrderableInputsListComponent.prototype.drop = function (event) {
        moveItemInArray(this.data, event.previousIndex, event.currentIndex);
        this.update();
    };
    OrderableInputsListComponent.prototype.update = function () {
        this.onUpdate.emit(this.data);
    };
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], OrderableInputsListComponent.prototype, "data", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], OrderableInputsListComponent.prototype, "onUpdate", void 0);
    __decorate([
        debounceable(500),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], OrderableInputsListComponent.prototype, "update", null);
    OrderableInputsListComponent = __decorate([
        Component({
            selector: 'ngx-orderable-inputs-list',
            template: "<div class=\"orderable-inputs-list\" cdkDropList (cdkDropListDropped)=\"drop($event)\">\n  <div class=\"input-item\" *ngFor=\"let item of data; let index = index\" cdkDrag cdkDragLockAxis=\"y\">\n    <div class=\"drag-drop-handle\" cdkDragHandle>\n      <i class=\"ngx-icon ngx-handle\"></i>\n    </div>\n    <ngx-input type=\"text\" (change)=\"data[index] = $event\" [ngModel]=\"item\" (keyup)=\"update()\"> </ngx-input>\n    <button type=\"button\" class=\"remove-item\" (click)=\"removeItem(index)\">\n      <i class=\"ngx-icon ngx-x\"></i>\n    </button>\n  </div>\n  <div class=\"add-examples\" (click)=\"addExample()\">\n    Add Example\n  </div>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".add-examples,.input-item{height:45px;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center}.input-item{background:#3b4457;box-shadow:0 2px 7px rgba(0,0,0,.2);border-radius:2px;margin-bottom:13px}.input-item .drag-drop-handle,.input-item .remove-item{color:#818fa9}.input-item .drag-drop-handle{margin:0 9px;height:17px;font-size:14.5px;cursor:-webkit-grab;cursor:grab}.input-item .ngx-input{-webkit-box-flex:1;flex:1;margin:0}.input-item .remove-item{font-size:8px;margin:0 5px;height:8px}.add-examples{cursor:pointer;height:45px;border:1px dashed #479eff;color:#479eff;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center}.cdk-drag-placeholder{opacity:0}.cdk-drag-animating,.orderable-inputs-list.cdk-drop-list-dragging .input-item:not(.cdk-drag-placeholder){-webkit-transition:-webkit-transform 250ms cubic-bezier(0,0,.2,1);transition:transform 250ms cubic-bezier(0,0,.2,1);transition:transform 250ms cubic-bezier(0,0,.2,1),-webkit-transform 250ms cubic-bezier(0,0,.2,1)}"]
        })
    ], OrderableInputsListComponent);
    return OrderableInputsListComponent;
}());
export { OrderableInputsListComponent };
