// @swimlane/ngx-ui/esm2015/lib/components/dialog/alert/alert.component.js
import { __decorate, __metadata } from "tslib";
import { animate, state, style, transition, trigger } from '@angular/animations';
import { Component, EventEmitter, Input, Output, ViewChild, ViewEncapsulation, ElementRef, ChangeDetectionStrategy } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { DialogComponent } from '../dialog.component';
import { AlertTypes } from './alert-types.enum';
let AlertComponent = class AlertComponent extends DialogComponent {
    constructor() {
        super(...arguments);
        this.data = '';
        this.ok = new EventEmitter();
        this.cancel = new EventEmitter();
        this.AlertTypes = AlertTypes;
        this.defaults = {
            inputs: {
                zIndex: 991,
                closeOnBlur: false,
                closeOnEscape: false,
                closeButton: false,
                showOverlay: true,
                visible: true,
                class: ''
            }
        };
    }
    get longPress() {
        return this._longPress;
    }
    set longPress(longPress) {
        this._longPress = coerceBooleanProperty(longPress);
    }
    ngOnInit() {
        if (this.longPress) {
            this.closeButton = true;
        }
    }
    ngAfterViewInit() {
        if (this.type !== AlertTypes.Prompt) {
            this.dialogElm.nativeElement.focus();
        }
    }
    onOkClick() {
        this.ok.emit({ data: this.data });
        this.hide();
    }
    onCancelClick() {
        this.cancel.emit({ data: this.data });
        this.hide();
    }
    onKeydown() {
        if (!this.longPress) {
            this.ok.emit({ data: this.data });
            this.hide();
        }
    }
};
__decorate([
    Input(),
    __metadata("design:type", String)
], AlertComponent.prototype, "type", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], AlertComponent.prototype, "data", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], AlertComponent.prototype, "longPress", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AlertComponent.prototype, "ok", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], AlertComponent.prototype, "cancel", void 0);
__decorate([
    ViewChild('dialogContent', { static: true }),
    __metadata("design:type", ElementRef)
], AlertComponent.prototype, "dialogElm", void 0);
AlertComponent = __decorate([
    Component({
        exportAs: 'ngxAlertDialog',
        selector: 'ngx-alert-dialog',
        template: "<div class=\"ngx-dialog ngx-alert-dialog\" [style.zIndex]=\"zIndex\" [ngClass]=\"type\">\n  <div\n    class=\"ngx-dialog-content\"\n    [ngClass]=\"cssClass\"\n    [@visibilityTransition]=\"visibleState\"\n    [style.zIndex]=\"contentzIndex\"\n    #dialogContent\n    (keydown.escape)=\"onCancelClick()\"\n    (keydown.enter)=\"onKeydown()\"\n    tabindex=\"-1\"\n    role=\"dialog\"\n  >\n    <div class=\"ngx-dialog-header\" *ngIf=\"title || closeButton\">\n      <button *ngIf=\"closeButton\" type=\"button\" class=\"close\" (click)=\"hide()\">\n        <span class=\"icon-x\"></span>\n      </button>\n      <button\n        *ngIf=\"type === AlertTypes.Alert\"\n        type=\"button\"\n        class=\"btn close-button\"\n        (click)=\"onOkClick()\"\n      >\n        Ok\n      </button>\n      <h1 *ngIf=\"title\" [innerHTML]=\"title\"></h1>\n    </div>\n    <div *ngIf=\"content\" class=\"ngx-dialog-body\" [innerHTML]=\"content\"></div>\n\n    <div class=\"ngx-dialog-body\" *ngIf=\"type === AlertTypes.Prompt\">\n      <ngx-input type=\"text\" autofocus=\"true\" name=\"confirm_input\" [(ngModel)]=\"data\"> </ngx-input>\n    </div>\n\n    <div class=\"ngx-dialog-footer\" *ngIf=\"type !== AlertTypes.Alert\">\n      <ng-container *ngIf=\"!longPress\">\n        <button type=\"button\" class=\"btn btn-primary\" (click)=\"onOkClick()\">\n          Ok\n        </button>\n        <button type=\"button\" class=\"btn\" (click)=\"onCancelClick()\">\n          Cancel\n        </button>\n      </ng-container>\n      <ngx-long-press-button *ngIf=\"longPress\" (longPress)=\"onOkClick()\" duration=\"1000\"></ngx-long-press-button>\n    </div>\n  </div>\n</div>\n",
        animations: [
            trigger('visibilityTransition', [
                state('active', style({
                    opacity: 1,
                    transform: 'scale3d(1, 1, 1)'
                })),
                transition('void => *', [
                    style({
                        opacity: 0,
                        transform: 'scale3d(1.2, 1.2, 1.2)'
                    }),
                    animate('0.2s ease-out')
                ]),
                transition('* => inactive', [
                    style({
                        opacity: 1,
                        transform: 'scale3d(1, 1, 1)'
                    }),
                    animate('0.2s ease-out', style({
                        transform: 'scale3d(0.9, 0.9, 1)',
                        opacity: 0
                    }))
                ])
            ])
        ],
        host: {
            tabindex: '-1'
        },
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [".ngx-dialog{position:fixed;display:-webkit-box;display:flex;-webkit-box-align:center;align-items:center;-webkit-box-pack:center;justify-content:center;top:0;left:0;height:100vh;width:100%;pointer-events:none}.ngx-dialog .ngx-dialog-content{pointer-events:auto;border-radius:2px;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 4px 0 rgba(0,0,0,.14),0 3px 3px -2px rgba(0,0,0,.12);background:#1c2029;padding:1.4rem;position:relative;-webkit-animation-fill-mode:forwards;animation-fill-mode:forwards;min-width:250px;font-size:.9rem;color:#9c9c9c}.ngx-dialog .ngx-dialog-content .close{position:absolute;font-size:20px;color:#cfcfcf;right:-25px;top:-25px}.ngx-dialog .ngx-dialog-content .ngx-dialog-header .ngx-dialog-title,.ngx-dialog .ngx-dialog-content .ngx-dialog-header h1{font-size:1.8rem;font-weight:500;margin:0 0 1.4rem;color:#cfcfcf}.ngx-dialog .ngx-dialog-content .ngx-dialog-footer{text-align:right}.ngx-dialog .ngx-dialog-content .ngx-dialog-footer .btn{margin-left:5px}.ngx-dialog.wizard .ngx-dialog-content{padding:0;background:#2b3240}.ngx-dialog.wizard .ngx-dialog-content .ngx-dialog-header{padding:1.4rem;background:#262c38}.ngx-dialog.wizard .ngx-dialog-content .ngx-dialog-header h1{margin:0}.ngx-dialog.wizard .ngx-dialog-content .ngx-dialog-footer{padding:1.4rem}.ngx-dialog.wizard .ngx-dialog-content .ngx-dialog-footer .wizard-nav{float:left}.ngx-dialog.wizard .ngx-dialog-content .ngx-dialog-footer .wizard-nav .btn{margin:0 5px 0 0}.ngx-dialog.wizard .ngx-dialog-content .ngx-tabs{margin-bottom:0}.ngx-dialog.wizard .ngx-dialog-content .ngx-tabs .ngx-tabs-list{padding:0 20px;background:#212631;border-bottom:2px solid #212631}.ngx-dialog.wizard .ngx-dialog-content .ngx-tabs .ngx-tabs-list .ngx-tab{background:#212631}.ngx-dialog.wizard .ngx-dialog-content .ngx-tabs .ngx-tabs-list .ngx-tab button{padding:.6em .75em;text-transform:uppercase}.ngx-dialog.wizard .ngx-dialog-content .ngx-tabs .ngx-tabs-list .ngx-tab.active{background:#313847}.ngx-dialog.wizard .ngx-dialog-content .ngx-tabs .ngx-tab-content{padding:1.4rem 1.4rem 0}.ngx-dialog.ngx-dialog-full-screen{width:100vw;height:100%;overflow-y:scroll;position:fixed;top:0;bottom:0;left:0;right:0}.ngx-dialog.ngx-dialog-full-screen .ngx-dialog-content{box-shadow:none;width:100%;min-height:100vh}.ngx-dialog.ngx-dialog-full-screen .ngx-dialog-content button.close{color:#afb7c8;font-size:24px;opacity:1;top:56px;right:56px;text-shadow:none}@-webkit-keyframes anim-open{0%{opacity:0;-webkit-transform:scale3d(1.1,1.1,1);transform:scale3d(1.1,1.1,1)}100%{opacity:1;-webkit-transform:scale3d(1,1,1);transform:scale3d(1,1,1)}}@keyframes anim-open{0%{opacity:0;-webkit-transform:scale3d(1.1,1.1,1);transform:scale3d(1.1,1.1,1)}100%{opacity:1;-webkit-transform:scale3d(1,1,1);transform:scale3d(1,1,1)}}@-webkit-keyframes anim-close{0%{opacity:1}100%{opacity:0;-webkit-transform:scale3d(.9,.9,1);transform:scale3d(.9,.9,1)}}@keyframes anim-close{0%{opacity:1}100%{opacity:0;-webkit-transform:scale3d(.9,.9,1);transform:scale3d(.9,.9,1)}}", ".ngx-alert-dialog .ngx-dialog-content{padding:0;background:#2b3240}.ngx-alert-dialog .ngx-dialog-content .ngx-dialog-header{padding:1.4rem 2rem 0}.ngx-alert-dialog .ngx-dialog-content .ngx-dialog-header h1{font-size:18px;margin:0;color:#fff;font-weight:600}.ngx-alert-dialog .ngx-dialog-content .ngx-dialog-body{padding:.5rem 2rem;color:#909cb4}.ngx-alert-dialog .ngx-dialog-content .ngx-dialog-footer{padding:1.4rem 2rem}.ngx-alert-dialog.alert .ngx-dialog-content .ngx-dialog-header{padding:1.4rem 2rem;border-left:5px solid #1483ff;background:#262c38}.ngx-alert-dialog.alert .ngx-dialog-content .ngx-dialog-header .close-button{float:right;background:#1483ff;color:#fff}.ngx-alert-dialog.alert .ngx-dialog-content .ngx-dialog-body{padding:1.4rem 2rem}.ngx-alert-dialog.alert .ngx-dialog-content.ngx-alert-danger .ngx-dialog-header{border-left-color:#ff4514}.ngx-alert-dialog.alert .ngx-dialog-content.ngx-alert-danger .ngx-dialog-header .close-button{background:#ff4514;color:#fff}.ngx-alert-dialog.alert .ngx-dialog-content.ngx-alert-warning .ngx-dialog-header{border-left-color:#ffa814}.ngx-alert-dialog.alert .ngx-dialog-content.ngx-alert-warning .ngx-dialog-header .close-button{background:#ffa814;color:#1c2029}.ngx-alert-dialog.confirm,.ngx-alert-dialog.prompt{text-align:center}.ngx-alert-dialog.confirm .ngx-dialog-footer,.ngx-alert-dialog.prompt .ngx-dialog-footer{padding:.5rem 2rem 1.4rem;text-align:center}"]
    })
], AlertComponent);
export { AlertComponent };
