// @swimlane/ngx-ui/esm2015/lib/components/dialog/alert/alert.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
import { InjectionService } from '../../../services/injection/injection.service';
import { OverlayService } from '../../overlay/overlay.service';
import { DialogService } from '../dialog.service';
import { AlertComponent } from './alert.component';
import { AlertTypes } from './alert-types.enum';
import { AlertStyles } from './alert-styles.enum';
let AlertService = class AlertService extends DialogService {
    constructor(injectionService, overlayService) {
        super(injectionService, overlayService);
        this.injectionService = injectionService;
        this.overlayService = overlayService;
        this.defaults = {
            inputs: {
                zIndex: 991,
                closeOnBlur: false,
                closeOnEscape: false,
                closeButton: false,
                showOverlay: true,
                visible: true
            }
        };
        this.type = AlertComponent;
        this.clsMap = {
            [AlertStyles.Danger]: 'ngx-alert-danger',
            [AlertStyles.Warning]: 'ngx-alert-warning',
            [AlertStyles.Info]: 'ngx-alert-info'
        };
    }
    alert(options) {
        return this.createDialog(options, AlertTypes.Alert);
    }
    confirm(options) {
        return this.createDialog(options, AlertTypes.Confirm);
    }
    prompt(options) {
        return this.createDialog(options, AlertTypes.Prompt);
    }
    createDialog(options, type) {
        const subject = new Subject();
        const { title, content, longPress } = options;
        const cssClass = 'ngx-alert-dialog ' + this.clsMap[options.style];
        const component = this.create({
            title,
            content,
            longPress,
            type,
            cssClass
        });
        const list = component.instance.ok.subscribe((data) => {
            subject.next({
                type: 'ok',
                data
            });
            subject.complete();
            list.unsubscribe();
            list2.unsubscribe();
        });
        const list2 = component.instance.cancel.subscribe((data) => {
            subject.next({
                type: 'cancel',
                data
            });
            subject.complete();
            list.unsubscribe();
            list2.unsubscribe();
        });
        return subject;
    }
};
AlertService.ctorParameters = () => [
    { type: InjectionService },
    { type: OverlayService }
];
AlertService = __decorate([
    Injectable(),
    __metadata("design:paramtypes", [InjectionService, OverlayService])
], AlertService);
export { AlertService };
