// ngx-awesome-uploader/esm5/lib/file-preview-container/preview-lightbox/preview-lightbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
var PreviewLightboxComponent = /** @class */ (function () {
    function PreviewLightboxComponent(sanitizer) {
        this.sanitizer = sanitizer;
        this.close = new EventEmitter();
    }
    /**
     * @return {?}
     */
    PreviewLightboxComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.getSafeUrl(this.file.file);
    };
    /**
     * @param {?} file
     * @return {?}
     */
    PreviewLightboxComponent.prototype.getSafeUrl = /**
     * @param {?} file
     * @return {?}
     */
    function (file) {
        /** @type {?} */
        var url = window.URL.createObjectURL(file);
        this.safeUrl = this.sanitizer.bypassSecurityTrustResourceUrl(url);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PreviewLightboxComponent.prototype.onClose = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.close.next();
    };
    PreviewLightboxComponent.decorators = [
        { type: Component, args: [{
                    selector: 'preview-lightbox',
                    template: "<div class=\"ng-modal-backdrop\" (click)=\"onClose($event)\">\r\n\r\n</div>\r\n\r\n<div class=\"ng-modal-content\" >\r\n  <div class=\"close-icon-wrapper\" (click)=\"onClose($event)\">\r\n      <close-icon></close-icon>\r\n  </div>\r\n  <div class=\"lightbox-item\" >\r\n    <img [src]=\"safeUrl\" (load)=\"loaded = true\" [ngStyle]=\"{'visibility': loaded ? 'visible' : 'hidden'}\">\r\n  </div>\r\n </div>\r\n",
                    styles: [":host{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center;position:fixed;z-index:1040;left:0;top:0;width:100vw;height:100vh;overflow:hidden}.ng-modal-backdrop{position:fixed;top:0;right:0;bottom:0;left:0;z-index:1040;background:rgba(0,0,0,.288)}.ng-modal-content{display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center;color:rgba(0,0,0,.87);z-index:1041}.ng-modal-content .close-icon-wrapper{position:absolute;top:20px;right:20px;font-size:20px}.ng-modal-content .lightbox-item img{max-width:calc(100vw - 30px);max-height:calc(100vh - 30px);width:100%;height:auto;-o-object-fit:contain;object-fit:contain;-webkit-animation-name:zoomIn;animation-name:zoomIn;-webkit-animation-duration:.2s;animation-duration:.2s}@-webkit-keyframes zoomIn{from{opacity:0;-webkit-transform:scale3d(.9,.9,.9);transform:scale3d(.9,.9,.9)}50%{opacity:1}}@keyframes zoomIn{from{opacity:0;-webkit-transform:scale3d(.9,.9,.9);transform:scale3d(.9,.9,.9)}50%{opacity:1}}"]
                }] }
    ];
    /** @nocollapse */
    PreviewLightboxComponent.ctorParameters = function () { return [
        { type: DomSanitizer }
    ]; };
    PreviewLightboxComponent.propDecorators = {
        file: [{ type: Input }],
        close: [{ type: Output }]
    };
    return PreviewLightboxComponent;
}());
export { PreviewLightboxComponent };
if (false) {
    /** @type {?} */
    PreviewLightboxComponent.prototype.file;
    /** @type {?} */
    PreviewLightboxComponent.prototype.close;
    /** @type {?} */
    PreviewLightboxComponent.prototype.loaded;
    /** @type {?} */
    PreviewLightboxComponent.prototype.safeUrl;
    /**
     * @type {?}
     * @private
     */
    PreviewLightboxComponent.prototype.sanitizer;
}
