// @swimlane/ngx-ui/esm2015/lib/components/toggle/toggle.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ViewEncapsulation, forwardRef, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { coerceBooleanProperty, coerceNumberProperty } from '@angular/cdk/coercion';
const TOGGLE_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => ToggleComponent),
    multi: true
};
let nextId = 0;
let ToggleComponent = class ToggleComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.id = `toggle-${++nextId}`;
        this.name = null;
        this._value = false;
        this._disabled = false;
        this._required = false;
        this._tabIndex = 0;
        this.onTouchedCallback = () => {
            // placeholder
        };
        this.onChangeCallback = (_) => {
            // placeholder
        };
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(disabled) {
        this._disabled = coerceBooleanProperty(disabled);
    }
    get required() {
        return this._required;
    }
    set required(required) {
        this._required = coerceBooleanProperty(required);
    }
    get tabIndex() {
        return this._tabIndex;
    }
    set tabIndex(tabIndex) {
        this._tabIndex = coerceNumberProperty(tabIndex);
    }
    get value() {
        return this._value;
    }
    set value(value) {
        if (this.value !== value) {
            this._value = value;
            this.onChangeCallback(value);
            this.cdr.markForCheck();
        }
    }
    get getHostCssClasses() {
        return 'ngx-toggle';
    }
    get getDisabled() {
        return this.disabled ? 'disabled' : '';
    }
    toggle() {
        this.value = !this.value;
    }
    onBlur() {
        this.onTouchedCallback();
    }
    writeValue(val) {
        if (val === null || val === undefined) {
            val = false;
        }
        if (val !== this._value) {
            this.value = val;
            this.cdr.markForCheck();
        }
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
};
ToggleComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", String)
], ToggleComponent.prototype, "id", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], ToggleComponent.prototype, "name", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], ToggleComponent.prototype, "label", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], ToggleComponent.prototype, "disabled", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], ToggleComponent.prototype, "required", null);
__decorate([
    Input(),
    __metadata("design:type", Object),
    __metadata("design:paramtypes", [Object])
], ToggleComponent.prototype, "tabIndex", null);
ToggleComponent = __decorate([
    Component({
        selector: 'ngx-toggle',
        exportAs: 'ngxToggle',
        template: "<div>\n  <input\n    #input\n    class=\"ngx-toggle-input\"\n    type=\"checkbox\"\n    [id]=\"id\"\n    [(ngModel)]=\"value\"\n    [required]=\"required\"\n    [tabIndex]=\"tabIndex\"\n    [disabled]=\"disabled\"\n    [name]=\"name\"\n    (blur)=\"onBlur()\"\n  />\n  <label [attr.for]=\"id\" class=\"ngx-toggle-label\"> </label>\n  <label [attr.for]=\"id\" class=\"ngx-toggle-text\">\n    <span *ngIf=\"label\" [innerHTML]=\"label\"></span>\n    <ng-content></ng-content>\n  </label>\n</div>\n",
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [TOGGLE_VALUE_ACCESSOR],
        host: {
            class: 'ngx-toggle',
            '[class.disabled]': 'getDisabled'
        },
        styles: [".ngx-toggle{display:block;margin-bottom:1rem;margin-top:18px}.ngx-toggle.disabled .ngx-toggle-label,.ngx-toggle.disabled .ngx-toggle-text{cursor:not-allowed}.ngx-toggle .ngx-toggle-input{display:none}.ngx-toggle .ngx-toggle-input:checked~.ngx-toggle-label{background:#479eff}.ngx-toggle .ngx-toggle-input:checked~.ngx-toggle-label:after{left:15px;background:#1483ff}.ngx-toggle .ngx-toggle-label{position:relative;height:14px;width:34px;background:#313847;border-radius:100px;cursor:pointer;-webkit-transition:.3s;transition:.3s;display:inline-block;vertical-align:middle;margin-bottom:3px;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.ngx-toggle .ngx-toggle-label:after{position:absolute;left:-2px;top:-3px;display:block;height:20px;width:20px;border-radius:100px;background:#72819f;box-shadow:0 3px 3px #07080b;content:'';-webkit-transition:.3s;transition:.3s}.ngx-toggle .ngx-toggle-text{cursor:pointer;padding-left:5px}"]
    }),
    __metadata("design:paramtypes", [ChangeDetectorRef])
], ToggleComponent);
export { ToggleComponent };
