// @swimlane/ngx-ui/esm5/lib/components/json-editor/json-editor/json-editor.component.js
import { __decorate, __extends, __metadata } from "tslib";
import { Component, ViewEncapsulation, ContentChildren, QueryList, Input, ChangeDetectionStrategy } from '@angular/core';
import { JsonEditor } from '../json-editor';
import { SchemaValidatorService } from '../schema-validator.service';
import { JsonEditorNodeComponent } from './json-editor-node/json-editor-node.component';
var JsonEditorComponent = /** @class */ (function (_super) {
    __extends(JsonEditorComponent, _super);
    function JsonEditorComponent(schemaValidatorService) {
        var _this = _super.call(this, schemaValidatorService) || this;
        _this.schemaValidatorService = schemaValidatorService;
        return _this;
    }
    JsonEditorComponent.ctorParameters = function () { return [
        { type: SchemaValidatorService }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], JsonEditorComponent.prototype, "model", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], JsonEditorComponent.prototype, "schema", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], JsonEditorComponent.prototype, "typeCheckOverrides", void 0);
    __decorate([
        ContentChildren(JsonEditorNodeComponent),
        __metadata("design:type", QueryList)
    ], JsonEditorComponent.prototype, "nodeElms", void 0);
    JsonEditorComponent = __decorate([
        Component({
            selector: 'ngx-json-editor',
            template: "<div class=\"ngx-json-editor\">\n  <div class=\"editor-title\">\n    <div class=\"type-icon\">\n      <ngx-icon fontIcon=\"integration\"></ngx-icon>\n    </div>\n    <div class=\"name\">\n      <div class=\"title\" *ngIf=\"label\">{{ label }}</div>\n      <div class=\"title\" *ngIf=\"!label\">{{ schema.title ? schema.title : 'Object' }}</div>\n    </div>\n  </div>\n\n  <ngx-json-editor-node\n    [(model)]=\"model\"\n    [schema]=\"schema\"\n    (modelChange)=\"modelChangedCallback($event)\"\n    [errors]=\"errors\"\n    [typeCheckOverrides]=\"typeCheckOverrides\"\n  >\n  </ngx-json-editor-node>\n</div>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-json-editor .editor-title{border:1px solid rgba(148,198,255,.07);height:30px;display:inline-block;margin-left:20px;position:relative;background:rgba(148,198,255,.05)}.ngx-json-editor .editor-title::before{content:' ';width:15px;height:1px;border-radius:5px;background:0 0;border-top:1px dashed #5a6884;position:absolute;top:50%;left:-15px}.ngx-json-editor .editor-title .type-icon{width:28px;height:28px;position:relative;background:rgba(148,198,255,.07);margin-left:0;display:inline-block}.ngx-json-editor .editor-title .type-icon ngx-icon{color:#afb7c8;position:absolute;top:50%;left:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);width:16px;height:16px}.ngx-json-editor .editor-title .name{display:inline-block;width:170px;padding:0 5px;position:relative;-webkit-transform:translateY(-50%);transform:translateY(-50%);vertical-align:middle}.ngx-json-editor .editor-title .name .title{display:inline-block;width:100%;color:#72819f;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;position:relative;top:5px}.ngx-json-editor .ngx-dropdown-menu{white-space:nowrap;left:16px!important}.ngx-json-editor .ngx-dropdown-menu .dropdown-column{vertical-align:top;display:inline-block;min-width:150px}.ngx-json-editor .ngx-dropdown-menu .dropdown-column:nth-child(2){border-left:1px solid #505c75}.code-editor-dialog .code-editor{width:90vw;overflow:auto}.code-editor-dialog .code-editor .save-code-button{position:absolute;right:22px;top:55px}.code-editor-dialog .code-editor .CodeMirror{height:70vh}.code-editor-dialog .code-editor .mode-select{width:300px}"]
        }),
        __metadata("design:paramtypes", [SchemaValidatorService])
    ], JsonEditorComponent);
    return JsonEditorComponent;
}(JsonEditor));
export { JsonEditorComponent };
