// @swimlane/ngx-ui/esm2015/lib/components/tree/tree.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, EventEmitter, Output, ContentChild, ViewEncapsulation, ContentChildren, TemplateRef, QueryList, ChangeDetectionStrategy, ChangeDetectorRef, OnDestroy, AfterContentInit } from '@angular/core';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
import { TreeNodeComponent } from './tree-node.component';
let TreeComponent = class TreeComponent {
    constructor(_cdr) {
        this._cdr = _cdr;
        this.icons = {
            collapse: 'icon-tree-collapse',
            expand: 'icon-tree-expand'
        };
        this.expand = new EventEmitter();
        this.collapse = new EventEmitter();
        this.activate = new EventEmitter();
        this.deactivate = new EventEmitter();
        this.selectNode = new EventEmitter();
        this._destroy$ = new Subject();
    }
    get hasOneLeaf() {
        return (this.nodes && this.nodes.length === 1) || this.nodeElms.length === 1;
    }
    get template() {
        return this._templateInput || this._templateQuery;
    }
    ngAfterContentInit() {
        this.nodeElms.changes.pipe(takeUntil(this._destroy$)).subscribe(() => this._cdr.markForCheck());
    }
    ngOnDestroy() {
        this._destroy$.next();
        this._destroy$.complete();
    }
};
TreeComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", Array)
], TreeComponent.prototype, "nodes", void 0);
__decorate([
    Input('template'),
    __metadata("design:type", TemplateRef)
], TreeComponent.prototype, "_templateInput", void 0);
__decorate([
    Input(),
    __metadata("design:type", Object)
], TreeComponent.prototype, "icons", void 0);
__decorate([
    ContentChild(TemplateRef, { static: true }),
    __metadata("design:type", TemplateRef)
], TreeComponent.prototype, "_templateQuery", void 0);
__decorate([
    ContentChildren(TreeNodeComponent),
    __metadata("design:type", QueryList)
], TreeComponent.prototype, "nodeElms", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], TreeComponent.prototype, "expand", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], TreeComponent.prototype, "collapse", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], TreeComponent.prototype, "activate", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], TreeComponent.prototype, "deactivate", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], TreeComponent.prototype, "selectNode", void 0);
TreeComponent = __decorate([
    Component({
        selector: 'ngx-tree',
        template: "<div class=\"ngx-tree\" [class.one-leaf]=\"hasOneLeaf\">\n  <ul class=\"vertical-list\">\n    <ngx-tree-node\n      *ngFor=\"let node of nodes\"\n      [node]=\"node\"\n      [disabled]=\"node.disabled\"\n      [expandable]=\"node.expandable\"\n      [expanded]=\"node.expanded\"\n      [selectable]=\"node.selectable\"\n      [icons]=\"icons\"\n      [label]=\"node.label\"\n      [model]=\"node.model\"\n      [children]=\"node.children\"\n      [template]=\"template\"\n      (expand)=\"expand.emit($event)\"\n      (collapse)=\"collapse.emit($event)\"\n      (activate)=\"activate.emit($event)\"\n      (deactivate)=\"deactivate.emit($event)\"\n      (selectNode)=\"selectNode.emit($event)\"\n    >\n    </ngx-tree-node>\n    <ng-content *ngIf=\"!nodes\"></ng-content>\n  </ul>\n  <div class=\"ngx-tree-vr\" *ngIf=\"nodes?.length || nodeElms?.length\"></div>\n</div>\n",
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [".ngx-tree{position:relative}.ngx-tree .ngx-tree-node{position:relative;padding-left:20px;line-height:25px;min-height:25px}.ngx-tree .ngx-tree-node .ngx-expander{font-size:.75rem;color:#b3b6bd;position:absolute;left:0;top:3px;z-index:1;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}.ngx-tree .ngx-tree-node .ngx-expander:not(.disabled){cursor:pointer}.ngx-tree .ngx-tree-node .ngx-expander.icon-tree-collapse::before,.ngx-tree .ngx-tree-node .ngx-expander.icon-tree-expand::before{background:#1c2029;border-radius:50%}.ngx-tree .ngx-tree-node .ngx-node-label{color:#b3b6bd;font-size:.9rem}.ngx-tree .ngx-tree-node .ngx-tree{margin-left:5px}.ngx-tree .ngx-tree-node.selectable>.ngx-node-label:not(.disabled){cursor:pointer}.ngx-tree.one-leaf>.ngx-tree-vr::before{display:none}.ngx-tree .ngx-tree-vr{position:absolute;top:17px;bottom:12px;left:5px;border-left:1px dashed #667080;width:1px}.ngx-tree .ngx-tree-vr::after{content:' ';width:5px;height:5px;border-radius:5px;background:#667080;position:absolute;bottom:-3px;left:-3px}.ngx-tree .ngx-tree-vr::before{content:' ';width:5px;height:5px;border-radius:5px;background:0 0;border:1px solid #667080;position:absolute;top:-5px;left:-3px}"]
    }),
    __metadata("design:paramtypes", [ChangeDetectorRef])
], TreeComponent);
export { TreeComponent };
