// @swimlane/ngx-ui/esm5/lib/components/calendar/utils/get-weeks-for-days/get-weeks-for-days.util.js
import { __read, __spread } from "tslib";
import { getNumberRange } from '../get-number-range';
/**
 * Gets a array of days split by week
 * @param days
 * @param offset
 * @return days by week
 */
export function getWeeksForDays(days, startDay) {
    var weeks = [];
    var offset = 7;
    // fill front row
    if (startDay < 7) {
        offset = 7 - startDay;
    }
    var _loop_1 = function () {
        var wk = days.slice(0, offset);
        days.splice(0, offset);
        // fill front row
        if (offset < 7) {
            var firstDay_1 = wk[0].date;
            var fill = getNumberRange(0, startDay).map(function (_, i) {
                var date = firstDay_1.clone().subtract(startDay - i, 'd');
                return {
                    num: date.date(),
                    dayOfWeek: date.day(),
                    date: date,
                    prevMonth: true
                };
            });
            wk = __spread(fill, wk);
            offset = 7;
        }
        // fill last row
        if (!days.length && wk.length !== 7) {
            var lastDay_1 = wk[wk.length - 1].date;
            var fill = getNumberRange(wk.length, 7).map(function (_, i) {
                var date = lastDay_1.clone().add(i + 1, 'd');
                return {
                    num: date.date(),
                    dayOfWeek: date.day(),
                    date: date,
                    nextMonth: true
                };
            });
            wk = __spread(wk, fill);
        }
        wk.forEach(function (day) {
            day.classes = {
                'first-day-of-month': day.num === 1,
                'last-day-of-week': day.dayOfWeek === 6,
                today: day.today,
                'next-month': day.nextMonth,
                'prev-month': day.prevMonth
            };
        });
        weeks.push(wk);
    };
    while (days.length) {
        _loop_1();
    }
    return weeks;
}
