// @swimlane/ngx-ui/esm5/lib/components/json-editor/json-editor-flat/json-editor-node-flat/node-types/array-node-flat/array-node-flat.component.js
import { __decorate, __extends, __metadata, __read, __spread } from "tslib";
import { Component, ViewEncapsulation, Input, ViewChild, TemplateRef, ChangeDetectionStrategy } from '@angular/core';
import { ArrayNode } from '../../../../node-types/array-node.component';
import { jsonSchemaDataTypes } from '../../../../json-editor.helper';
import { DialogService } from '../../../../../dialog/dialog.service';
var ArrayNodeFlatComponent = /** @class */ (function (_super) {
    __extends(ArrayNodeFlatComponent, _super);
    function ArrayNodeFlatComponent(dialogService) {
        var _this = _super.call(this) || this;
        _this.dialogService = dialogService;
        _this.indentationArray = [];
        return _this;
    }
    ArrayNodeFlatComponent.prototype.ngOnInit = function () {
        if (this.schemaBuilderMode) {
            this.dataTypes = __spread(jsonSchemaDataTypes, this.formats);
        }
        if (this.schemaBuilderMode && !this.model.length && this.schemaRef.items && this.schemaRef.items.type) {
            this.model.push(this.schemaRef.items);
        }
        if (this.level > 0) {
            this.indentationArray = Array(this.level).fill(this.level);
        }
    };
    ArrayNodeFlatComponent.prototype.onPropertyConfig = function (item, index) {
        this.dialogService.create({
            template: this.propertyConfigTmpl,
            context: {
                property: item,
                index: index,
                schema: this.schema,
                formats: this.formats
            },
            class: 'property-config-dialog'
        });
    };
    ArrayNodeFlatComponent.prototype.updateSchema = function (options) {
        this.schema.items = options.newProperty;
        this.schemaRef.items = options.newProperty;
        this.schemaChange.emit();
    };
    ArrayNodeFlatComponent.prototype.addArrayItem = function (dataType) {
        if (this.schemaBuilderMode && dataType) {
            this.addDefaultItemForSchemaBuilder(dataType);
        }
        else {
            _super.prototype.addArrayItem.call(this, dataType);
        }
    };
    ArrayNodeFlatComponent.prototype.deleteArrayItem = function (index) {
        if (this.schemaBuilderMode) {
            this.removeDefaultItemForSchemaBuilder();
        }
        else {
            _super.prototype.deleteArrayItem.call(this, index);
        }
    };
    ArrayNodeFlatComponent.prototype.addDefaultItemForSchemaBuilder = function (dataType) {
        this.schema.items = dataType.schema;
        this.schemaRef.items = dataType.schema;
        this.model.push(this.schemaRef.items);
        this.schemaChange.emit();
    };
    ArrayNodeFlatComponent.prototype.removeDefaultItemForSchemaBuilder = function () {
        delete this.schema.items;
        delete this.schemaRef.items;
        this.model = [];
        this.schemaChange.emit();
    };
    ArrayNodeFlatComponent.ctorParameters = function () { return [
        { type: DialogService }
    ]; };
    __decorate([
        ViewChild('propertyConfigTmpl', { static: false }),
        __metadata("design:type", TemplateRef)
    ], ArrayNodeFlatComponent.prototype, "propertyConfigTmpl", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number)
    ], ArrayNodeFlatComponent.prototype, "level", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], ArrayNodeFlatComponent.prototype, "schemaBuilderMode", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], ArrayNodeFlatComponent.prototype, "formats", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], ArrayNodeFlatComponent.prototype, "compressed", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ArrayNodeFlatComponent.prototype, "hideRoot", void 0);
    ArrayNodeFlatComponent = __decorate([
        Component({
            selector: 'ngx-json-array-node-flat',
            template: "<div class=\"array-node-flat\" [hidden]=\"!expanded\">\n  <div\n    class=\"array-node-content\"\n    [class.no-margin]=\"schemaBuilderMode\"\n    *ngFor=\"let value of model; let i = index; trackBy: arrayTrackBy\"\n  >\n    <ngx-json-editor-node-flat\n      [model]=\"value\"\n      (modelChange)=\"updateArrayItem(i, $event)\"\n      [schema]=\"schemaBuilderMode ? schemaRef.items : schemas[i]\"\n      [path]=\"path + '[' + i + ']'\"\n      [errors]=\"errors\"\n      [typeCheckOverrides]=\"typeCheckOverrides\"\n      [level]=\"level\"\n      [hideRoot]=\"hideRoot\"\n      [compressed]=\"compressed\"\n      [schemaRef]=\"schemaRef?.items || null\"\n      [schemaBuilderMode]=\"schemaBuilderMode\"\n      [arrayItem]=\"schemaBuilderMode\"\n      [arrayName]=\"(schema.title ? schema.title : schema.propertyName) + '[' + (schemaBuilderMode ? '' : i) + ']'\"\n      [indentationArray]=\"indentationArray\"\n    >\n      <div class=\"node-options\" node-options>\n        <button\n          *ngIf=\"schemaBuilderMode\"\n          type=\"button\"\n          class=\"node-options-btn\"\n          (click)=\"onPropertyConfig(schemaRef.items, i)\"\n        >\n          <i class=\"ngx-icon ngx-cog\"></i>\n        </button>\n        <ngx-dropdown [showCaret]=\"true\">\n          <ngx-dropdown-toggle>\n            <button type=\"button\" class=\"node-options-btn\">\n              <i class=\"ngx-icon ngx-dots-vert-round\"></i>\n            </button>\n          </ngx-dropdown-toggle>\n          <ngx-dropdown-menu class=\"ngx-dropdown-dark-outline align-right\">\n            <ul class=\"vertical-list\">\n              <li><button type=\"button\" (click)=\"deleteArrayItem(i)\">Delete</button></li>\n\n              <ng-container *ngIf=\"schemas[i]?.$meta?.type\">\n                <li *ngFor=\"let type of schemas[i]?.$meta?.type\">\n                  <button type=\"button\" (click)=\"changeItemType(i, type)\" [disabled]=\"schemas[i].currentType === type\">\n                    Change type to {{ dataTypeMap[type].name }}\n                  </button>\n                </li>\n              </ng-container>\n            </ul>\n          </ngx-dropdown-menu>\n        </ngx-dropdown>\n      </div>\n    </ngx-json-editor-node-flat>\n  </div>\n\n  <div\n    *ngIf=\"!schemaBuilderMode || !model.length\"\n    class=\"add-button\"\n    [class.compressed]=\"compressed\"\n    [class.background]=\"hideRoot ? level > -1 : level\"\n  >\n    <span class=\"separator\" *ngFor=\"let separator of indentationArray\"></span>\n    <div class=\"indented-content\" [style.marginLeft]=\"hideRoot && level === 0 ? '20px' : '0'\">\n      <ngx-dropdown [showCaret]=\"true\">\n        <ngx-dropdown-toggle>\n          <button type=\"button\">\n            <i class=\"ngx-icon ngx-tree-expand\"></i>\n            <span>Add an item</span>\n          </button>\n        </ngx-dropdown-toggle>\n        <ngx-dropdown-menu class=\"ngx-dropdown-dark-outline\">\n          <ul class=\"vertical-list\">\n            <ng-template [ngIf]=\"!schema || !schema.items || !schema.items.type\">\n              <li *ngFor=\"let dataType of dataTypes\" (click)=\"addArrayItem(dataType)\">\n                <button type=\"button\">{{ dataType.name }}</button>\n              </li>\n            </ng-template>\n            <ng-container *ngIf=\"schema?.items?.type\">\n              <li *ngIf=\"!_array.isArray(schema.items.type)\">\n                <button type=\"button\" (click)=\"addArrayItem()\">Add</button>\n              </li>\n              <ng-container *ngIf=\"_array.isArray(schema.items.type)\">\n                <li *ngFor=\"let type of schema.items.type\">\n                  <button type=\"button\" (click)=\"addArrayItem(dataTypeMap[type])\">\n                    Add {{ dataTypeMap[type].name }}\n                  </button>\n                </li>\n              </ng-container>\n            </ng-container>\n          </ul>\n        </ngx-dropdown-menu>\n      </ngx-dropdown>\n    </div>\n  </div>\n</div>\n\n<!-- Property Config Dialog -->\n<ng-template #propertyConfigTmpl let-context=\"context\">\n  <ngx-property-config\n    [property]=\"context.property\"\n    [index]=\"context.index\"\n    [schema]=\"context.schema\"\n    [formats]=\"context.formats\"\n    [arrayItem]=\"true\"\n    (updateSchema)=\"updateSchema($event)\"\n  >\n  </ngx-property-config>\n</ng-template>\n",
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".array-node-flat .add-button .indented-content .ngx-dropdown-menu{white-space:nowrap;margin-top:10px}.array-node-flat .add-button .indented-content .ngx-dropdown-menu .dropdown-column{vertical-align:top;display:inline-block;min-width:150px}.array-node-flat .add-button .indented-content .ngx-dropdown-menu .dropdown-column:nth-child(2){border-left:1px solid #505c75}.separator{background-color:#3b4457;opacity:.5;border-radius:2px;width:1px;height:calc(100% - 4px);margin-right:20px}.separator:first-child{margin-left:20px}.array-node-flat .array-node-content{display:-webkit-box;display:flex;margin-bottom:5px;position:relative}.array-node-flat .array-node-content ngx-json-editor-node-flat{-webkit-box-flex:1;flex:1}.array-node-flat .add-button{height:100px;position:relative;display:-webkit-box;display:flex}.array-node-flat .add-button ngx-dropdown{padding-bottom:0}.array-node-flat .compressed.add-button{max-height:80px}.array-node-flat .background.add-button{background-color:rgba(49,56,71,.4);padding:7px 7px 7px 0;height:120px}.array-node-flat .add-button .indented-content{border:2px dotted rgba(59,68,87,.5);width:100%;display:-webkit-box;display:flex;-webkit-box-pack:center;justify-content:center;-webkit-box-align:center;align-items:center;color:#909cb4}.array-node-flat .add-button .indented-content button{display:-webkit-box;display:flex}.array-node-flat .add-button .indented-content button i{font-size:18px}.array-node-flat .add-button .indented-content button span{font-size:14px;margin-left:10px}.array-node-flat{margin-top:5px}.array-node-flat .array-node-content.no-margin{margin-bottom:0}"]
        }),
        __metadata("design:paramtypes", [DialogService])
    ], ArrayNodeFlatComponent);
    return ArrayNodeFlatComponent;
}(ArrayNode));
export { ArrayNodeFlatComponent };
