// @swimlane/ngx-ui/esm5/lib/components/drawer/drawer.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '../overlay/overlay.module';
import { OverlayService } from '../overlay/overlay.service';
import { InjectionService } from '../../services/injection/injection.service';
import { DrawerComponent } from './drawer.component';
import { DrawerService } from './drawer.service';
import { DrawerContainerDirective } from './drawer-container.directive';
var DrawerModule = /** @class */ (function () {
    function DrawerModule() {
    }
    DrawerModule = __decorate([
        NgModule({
            declarations: [DrawerComponent, DrawerContainerDirective],
            exports: [DrawerComponent, DrawerContainerDirective],
            providers: [DrawerService, InjectionService, OverlayService],
            imports: [CommonModule, OverlayModule],
            entryComponents: [DrawerComponent]
        })
    ], DrawerModule);
    return DrawerModule;
}());
export { DrawerModule };
