// @swimlane/ngx-ui/esm5/lib/components/drawer/drawer.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, Output, EventEmitter, HostListener, ViewEncapsulation, ChangeDetectionStrategy, TemplateRef } from '@angular/core';
import { coerceNumberProperty, coerceBooleanProperty } from '@angular/cdk/coercion';
import { trigger } from '@angular/animations';
import { DRAWER_ANIMATION } from './drawer.animation';
import { DrawerDirection } from './drawer-direction.enum';
import { DrawerPosition } from './drawer-position.enum';
var DrawerComponent = /** @class */ (function () {
    function DrawerComponent() {
        this.cssClass = '';
        this.isRoot = true;
        this.close = new EventEmitter();
        this.position = DrawerPosition.fixed;
    }
    Object.defineProperty(DrawerComponent.prototype, "size", {
        get: function () {
            return this._size;
        },
        set: function (val) {
            this._size = coerceNumberProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerComponent.prototype, "zIndex", {
        get: function () {
            return this._zIndex;
        },
        set: function (val) {
            this._zIndex = coerceNumberProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerComponent.prototype, "closeOnOutsideClick", {
        get: function () {
            return this._closeOnOutsideClick;
        },
        set: function (val) {
            this._closeOnOutsideClick = coerceBooleanProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerComponent.prototype, "cssClasses", {
        get: function () {
            var clz = "ngx-drawer " + this.cssClass;
            if (this.isLeft)
                clz += ' left-drawer';
            if (this.isBottom)
                clz += ' bottom-drawer';
            return clz;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerComponent.prototype, "isLeft", {
        get: function () {
            return this.direction === DrawerDirection.Left;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DrawerComponent.prototype, "isBottom", {
        get: function () {
            return this.direction === DrawerDirection.Bottom;
        },
        enumerable: true,
        configurable: true
    });
    DrawerComponent.prototype.ngOnInit = function () {
        this.position = this.isRoot ? DrawerPosition.fixed : DrawerPosition.absolute;
        this.setDimensions(this.size);
    };
    DrawerComponent.prototype.ngOnDestroy = function () {
        this.close.emit(true);
    };
    DrawerComponent.prototype.setDimensions = function (size) {
        this.heightSize = (this.isBottom && size ? size : 100) + "%";
        this.widthSize = (this.isLeft && size ? size : 100) + "%";
    };
    DrawerComponent.prototype.onEscapeKey = function () {
        this.close.emit(true);
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DrawerComponent.prototype, "cssClass", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], DrawerComponent.prototype, "direction", void 0);
    __decorate([
        Input(),
        __metadata("design:type", TemplateRef)
    ], DrawerComponent.prototype, "template", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], DrawerComponent.prototype, "context", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], DrawerComponent.prototype, "isRoot", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], DrawerComponent.prototype, "size", null);
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], DrawerComponent.prototype, "zIndex", null);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], DrawerComponent.prototype, "closeOnOutsideClick", null);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], DrawerComponent.prototype, "close", void 0);
    __decorate([
        HostListener('keyup.esc'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DrawerComponent.prototype, "onEscapeKey", null);
    DrawerComponent = __decorate([
        Component({
            exportAs: 'ngxDrawer',
            selector: 'ngx-drawer',
            template: "<div class=\"ngx-drawer-content\">\n  <ng-template [ngTemplateOutlet]=\"template\" [ngTemplateOutletContext]=\"{ context: context, close: close }\">\n  </ng-template>\n</div>\n",
            animations: [trigger('drawerTransition', DRAWER_ANIMATION)],
            host: {
                role: 'dialog',
                tabindex: '-1',
                '[class]': 'cssClasses',
                '[style.width]': 'widthSize',
                '[style.height]': 'heightSize',
                '[style.zIndex]': 'zIndex',
                '[style.position]': 'position',
                '[@drawerTransition]': 'direction'
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-drawer{display:block;overflow-y:auto;overflow-x:hidden;text-align:left;z-index:998;background:#1c2029}.ngx-drawer.left-drawer{top:0;right:0;height:100%}.ngx-drawer.bottom-drawer{left:0;bottom:0;width:100%}"]
        })
    ], DrawerComponent);
    return DrawerComponent;
}());
export { DrawerComponent };
