// @swimlane/ngx-ui/esm5/lib/components/json-editor/node-types/object-node.component.js
import { __assign, __decorate, __metadata, __read, __spread, __values } from "tslib";
import { Input, EventEmitter, Output } from '@angular/core';
import { createValueForSchema, jsonSchemaDataTypes, jsonSchemaDataFormats, inferType, dataTypeMap, getIcon, getCurrentType } from '../json-editor.helper';
var ObjectNode = /** @class */ (function () {
    function ObjectNode(cdr) {
        this.cdr = cdr;
        this.required = false;
        this.modelChange = new EventEmitter();
        this.schemaChange = new EventEmitter();
        this.requiredCache = {};
        this.dataTypes = __spread(jsonSchemaDataTypes, jsonSchemaDataFormats);
        this.propertyCounter = 1;
        this.propertyId = 1;
        this.propertyIndex = {};
        this.dataTypeMap = dataTypeMap;
    }
    ObjectNode.prototype.ngOnInit = function () {
        this.update();
    };
    ObjectNode.prototype.ngOnChanges = function (changes) {
        if (changes.model !== undefined || changes.schema !== undefined) {
            this.update();
        }
    };
    ObjectNode.prototype.update = function () {
        var _this = this;
        setTimeout(function () {
            for (var prop in _this.schema.properties) {
                if (Array.isArray(_this.schema.properties[prop].type) && _this.schema.properties[prop].type.length > 0) {
                    if (!_this.schema.properties[prop].$meta) {
                        _this.schema.properties[prop].$meta = {};
                    }
                    _this.schema.properties[prop].$meta.type = __spread(_this.schema.properties[prop].type);
                    if (_this.model[prop] !== undefined) {
                        _this.schema.properties[prop] = __assign(__assign({}, _this.schema.properties[prop]), inferType(_this.model[prop], _this.typeCheckOverrides, _this.schema.properties[prop].$meta.type));
                    }
                    else {
                        _this.schema.properties[prop].type = _this.schema.properties[prop].type[0];
                        _this.schema.properties[prop].$meta.currentType = getCurrentType(_this.schema.properties[prop]);
                    }
                }
            }
            _this.updateRequiredCache();
            _this.indexProperties();
            _this.addRequiredProperties();
            _this.updateIcons();
        });
    };
    /**
     * Updates a property on the model and emits the change event
     * @param propName
     * @param value
     */
    ObjectNode.prototype.updateProp = function (id, value) {
        var propName = this.propertyIndex[id].propertyName;
        this.model[propName] = value;
        this.modelChange.emit(this.model);
    };
    /**
     * Updates the name of a property
     * @param id
     * @param name
     */
    ObjectNode.prototype.updatePropertyName = function (id, name) {
        var oldName = this.propertyIndex[id].propertyName;
        this.model[name] = this.model[oldName];
        this.propertyIndex[id].propertyName = name;
        delete this.model[oldName];
        this.propertyIndex = __assign({}, this.propertyIndex);
        this.modelChange.emit(this.model);
    };
    /**
     * Adds a new property to the model
     */
    ObjectNode.prototype.addProperty = function (dataType) {
        var propName = dataType.name + " " + this.propertyCounter;
        this.propertyCounter++;
        var schema = JSON.parse(JSON.stringify(dataType.schema));
        this.model[propName] = createValueForSchema(dataType.schema);
        schema.nameEditable = !this.schemaBuilderMode;
        schema.propertyName = propName;
        schema.id = this.propertyId++;
        this.propertyIndex[schema.id] = schema;
        this.propertyIndex = __assign({}, this.propertyIndex);
        this.modelChange.emit(this.model);
        this.updateIcons();
    };
    /**
     * Adds a new property as defined in the schema
     */
    ObjectNode.prototype.addSchemaProperty = function (propName) {
        if (this.model[propName] !== undefined) {
            return;
        }
        var schema = JSON.parse(JSON.stringify(this.schema.properties[propName]));
        if (!schema.type) {
            schema.type = 'object';
        }
        var value = createValueForSchema(schema);
        this.model[propName] = value;
        schema.nameEditable = false;
        schema.propertyName = propName;
        schema.id = this.propertyId++;
        this.propertyIndex[schema.id] = schema;
        this.propertyIndex = __assign({}, this.propertyIndex);
        this.modelChange.emit(this.model);
        this.updateIcons();
    };
    /**
     * Adds a new patternProperty as defined in the schema
     */
    ObjectNode.prototype.addSchemaPatternProperty = function (propName) {
        var newPropName = "new " + this.schema.patternProperties[propName].title + " " + this.propertyCounter;
        this.propertyCounter++;
        var schema = JSON.parse(JSON.stringify(this.schema.patternProperties[propName]));
        schema.isPatternProperty = true;
        if (!schema.type) {
            schema.type = 'object';
        }
        var value = createValueForSchema(schema);
        this.model[newPropName] = value;
        schema.nameEditable = true;
        schema.propertyName = newPropName;
        schema.id = this.propertyId++;
        this.propertyIndex[schema.id] = schema;
        this.propertyIndex = __assign({}, this.propertyIndex);
        this.modelChange.emit(this.model);
        this.updateIcons();
    };
    /**
     * Deletes a property
     */
    ObjectNode.prototype.deleteProperty = function (propName) {
        delete this.model[propName];
        for (var id in this.propertyIndex) {
            if (this.propertyIndex[id].propertyName === propName) {
                delete this.propertyIndex[id];
                break;
            }
        }
        this.model = __assign({}, this.model);
        this.propertyIndex = __assign({}, this.propertyIndex);
        this.modelChange.emit(this.model);
    };
    /**
     * Returns the absolute tree path of the property
     */
    ObjectNode.prototype.getPath = function (propName) {
        var propSchema;
        for (var id in this.propertyIndex) {
            if (this.propertyIndex[id].propertyName === propName) {
                propSchema = this.propertyIndex[id];
                break;
            }
        }
        if (propSchema.isPatternProperty) {
            return "['" + propName + "']";
        }
        return "." + propName;
    };
    /**
     * Updates the required cache
     */
    ObjectNode.prototype.updateRequiredCache = function () {
        var e_1, _a;
        this.requiredCache = {};
        if (this.schema && this.schema.required) {
            try {
                for (var _b = __values(this.schema.required), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var prop = _c.value;
                    this.requiredCache[prop] = true;
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
    };
    /**
     * Creates an index out of all the properties in the model
     */
    ObjectNode.prototype.indexProperties = function () {
        var props = this.schemaBuilderMode ? this.schemaRef.properties : this.model;
        for (var prop in props) {
            if (this.isIndexed(prop)) {
                continue;
            }
            var schema = void 0;
            if (this.schema.properties && this.schema.properties[prop]) {
                schema = JSON.parse(JSON.stringify(this.schema.properties[prop]));
            }
            else {
                var matchesPattern = false;
                if (this.schema.patternProperties) {
                    for (var pattern in this.schema.patternProperties) {
                        var patternRegex = new RegExp(pattern);
                        if (patternRegex.test(prop)) {
                            schema = JSON.parse(JSON.stringify(this.schema.patternProperties[pattern]));
                            matchesPattern = true;
                        }
                    }
                }
                if (!matchesPattern) {
                    schema = __assign({}, inferType(this.model[prop], this.typeCheckOverrides));
                }
            }
            schema.id = this.propertyId++;
            schema.propertyName = prop;
            this.propertyIndex[schema.id] = schema;
            this.propertyIndex = __assign({}, this.propertyIndex);
        }
        for (var id in this.propertyIndex) {
            var schema = this.propertyIndex[id];
            if (this.model[schema.propertyName] === undefined) {
                delete this.propertyIndex[id];
            }
        }
        this.propertyIndex = __assign({}, this.propertyIndex);
        this.cdr.markForCheck();
    };
    ObjectNode.prototype.isIndexed = function (propertyName) {
        return Object.values(this.propertyIndex).findIndex(function (s) { return s.propertyName === propertyName; }) !== -1;
    };
    /**
     * Inits the required properties on the model
     */
    ObjectNode.prototype.addRequiredProperties = function () {
        if (this.schema && this.schema.properties) {
            for (var propName in this.schema.properties) {
                if (this.model[propName] !== undefined) {
                    continue;
                }
                if (this.requiredCache[propName] || this.schemaBuilderMode) {
                    // List all properties not only required if we are in schema builder mode
                    this.addSchemaProperty(propName);
                }
            }
        }
    };
    /**
     *
     * @param property
     * @param type
     */
    ObjectNode.prototype.changePropertyType = function (property, type) {
        var dataType = this.dataTypeMap[type];
        if (dataType) {
            delete property.format;
            property.type = dataType.schema.type;
            if (dataType.schema.format) {
                property.format = dataType.schema.format;
            }
            property.$meta.currentType = getCurrentType(property);
            this.schema.properties[property.propertyName] = __assign({}, property);
        }
        var value = createValueForSchema(property);
        this.model[property.propertyName] = value;
        this.modelChange.emit(this.model);
        this.updateIcons();
    };
    /**
     * Track By function for the array ittierator
     * @param index
     * @param value
     */
    ObjectNode.prototype.trackBy = function (_, value) {
        return value.id;
    };
    /**
     * Updates the icons in the schemas
     */
    ObjectNode.prototype.updateIcons = function () {
        for (var id in this.propertyIndex) {
            var schema = this.propertyIndex[id];
            if (!schema.$meta) {
                schema.$meta = {};
            }
            schema.$meta.icon = getIcon(schema);
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ObjectNode.prototype, "schema", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ObjectNode.prototype, "model", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], ObjectNode.prototype, "required", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], ObjectNode.prototype, "expanded", void 0);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], ObjectNode.prototype, "path", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Array)
    ], ObjectNode.prototype, "errors", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ObjectNode.prototype, "typeCheckOverrides", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Boolean)
    ], ObjectNode.prototype, "schemaBuilderMode", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], ObjectNode.prototype, "schemaRef", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], ObjectNode.prototype, "modelChange", void 0);
    __decorate([
        Output(),
        __metadata("design:type", EventEmitter)
    ], ObjectNode.prototype, "schemaChange", void 0);
    return ObjectNode;
}());
export { ObjectNode };
