// ngx-awesome-uploader/esm5/lib/file-preview-container/file-preview-container.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, TemplateRef } from '@angular/core';
import { FilePickerAdapter } from '../file-picker.adapter';
var FilePreviewContainerComponent = /** @class */ (function () {
    function FilePreviewContainerComponent() {
        this.removeFile = new EventEmitter();
        this.uploadSuccess = new EventEmitter();
    }
    /**
     * @return {?}
     */
    FilePreviewContainerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} file
     * @return {?}
     */
    FilePreviewContainerComponent.prototype.openLightbox = /**
     * @param {?} file
     * @return {?}
     */
    function (file) {
        this.lightboxFile = file;
    };
    /**
     * @return {?}
     */
    FilePreviewContainerComponent.prototype.closeLightbox = /**
     * @return {?}
     */
    function () {
        this.lightboxFile = undefined;
    };
    FilePreviewContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'file-preview-container',
                    template: "<preview-lightbox *ngIf=\"lightboxFile\" [file]=\"lightboxFile\" (close)=\"closeLightbox()\"></preview-lightbox>\r\n<file-preview-item\r\n*ngFor=\"let file of previewFiles\"\r\n[fileItem]=\"file\"\r\n(removeFile)=\"removeFile.next($event)\"\r\n(uploadSuccess)=\"uploadSuccess.next($event)\"\r\n(imageClicked)=\"openLightbox($event)\"\r\n[itemTemplate]=\"itemTemplate\"\r\n[adapter]=\"adapter\"\r\n[captions]=\"captions\"\r\n></file-preview-item>\r\n",
                    styles: [":host{display:-webkit-box;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-direction:column;-webkit-box-pack:start;justify-content:flex-start;width:100%;background:#fafbfd}"]
                }] }
    ];
    /** @nocollapse */
    FilePreviewContainerComponent.ctorParameters = function () { return []; };
    FilePreviewContainerComponent.propDecorators = {
        previewFiles: [{ type: Input }],
        itemTemplate: [{ type: Input }],
        removeFile: [{ type: Output }],
        uploadSuccess: [{ type: Output }],
        adapter: [{ type: Input }],
        captions: [{ type: Input }]
    };
    return FilePreviewContainerComponent;
}());
export { FilePreviewContainerComponent };
if (false) {
    /** @type {?} */
    FilePreviewContainerComponent.prototype.previewFiles;
    /** @type {?} */
    FilePreviewContainerComponent.prototype.itemTemplate;
    /** @type {?} */
    FilePreviewContainerComponent.prototype.removeFile;
    /** @type {?} */
    FilePreviewContainerComponent.prototype.uploadSuccess;
    /** @type {?} */
    FilePreviewContainerComponent.prototype.lightboxFile;
    /** @type {?} */
    FilePreviewContainerComponent.prototype.adapter;
    /** @type {?} */
    FilePreviewContainerComponent.prototype.captions;
}
