// @swimlane/ngx-ui/esm5/lib/components/dropdown/dropdown.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, ContentChild, OnDestroy, AfterContentInit, Renderer2, ViewEncapsulation, ChangeDetectorRef, ChangeDetectionStrategy, HostListener } from '@angular/core';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { DropdownMenuDirective } from './dropdown-menu.directive';
import { DropdownToggleDirective } from './dropdown-toggle.directive';
var DropdownComponent = /** @class */ (function () {
    function DropdownComponent(renderer, cd) {
        this.renderer = renderer;
        this.cd = cd;
        this._open = false;
        this._showCaret = false;
        this._closeOnClick = true;
        this._closeOnOutsideClick = true;
        this._closeOnMouseLeave = false;
        this._leaveTimeout = null;
    }
    Object.defineProperty(DropdownComponent.prototype, "open", {
        get: function () {
            return this._open;
        },
        set: function (open) {
            this._open = coerceBooleanProperty(open);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "showCaret", {
        get: function () {
            return this._showCaret;
        },
        set: function (showCaret) {
            this._showCaret = coerceBooleanProperty(showCaret);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "closeOnClick", {
        get: function () {
            return this._closeOnClick;
        },
        set: function (closeOnClick) {
            this._closeOnClick = coerceBooleanProperty(closeOnClick);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "closeOnOutsideClick", {
        get: function () {
            return this._closeOnOutsideClick;
        },
        set: function (closeOnOutsideClick) {
            this._closeOnOutsideClick = coerceBooleanProperty(closeOnOutsideClick);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DropdownComponent.prototype, "closeOnMouseLeave", {
        get: function () {
            return this._closeOnMouseLeave;
        },
        set: function (val) {
            this._closeOnMouseLeave = coerceBooleanProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    DropdownComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        if (this.dropdownToggle) {
            this.dropdownToggle.toggle.subscribe(function (ev) { return _this.onToggleClick(ev); });
        }
    };
    DropdownComponent.prototype.ngOnDestroy = function () {
        if (this._documentListener)
            this._documentListener();
    };
    DropdownComponent.prototype.onDocumentClick = function (e) {
        if (this.open && this.closeOnOutsideClick) {
            var isToggling = this.dropdownToggle.element.contains(e.target);
            var isMenuClick = !this.closeOnClick && this.dropdownMenu.element.contains(e.target);
            if (!isToggling && !isMenuClick) {
                this.close();
            }
        }
    };
    DropdownComponent.prototype.onToggleClick = function (_) {
        this.open = !this.open;
        if (this.open) {
            this._documentListener = this.renderer.listen(document, 'click', this.onDocumentClick.bind(this));
        }
        else {
            this._documentListener();
        }
    };
    DropdownComponent.prototype.onMouseLeave = function () {
        var _this = this;
        if (this.closeOnMouseLeave) {
            this._leaveTimeout = setTimeout(function () {
                _this.close();
            }, 1000);
        }
    };
    DropdownComponent.prototype.onMouseEnter = function () {
        if (this._leaveTimeout) {
            clearTimeout(this._leaveTimeout);
        }
    };
    DropdownComponent.prototype.close = function () {
        this.open = false;
        if (this._documentListener)
            this._documentListener();
        this.cd.markForCheck();
    };
    DropdownComponent.ctorParameters = function () { return [
        { type: Renderer2 },
        { type: ChangeDetectorRef }
    ]; };
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], DropdownComponent.prototype, "open", null);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], DropdownComponent.prototype, "showCaret", null);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], DropdownComponent.prototype, "closeOnClick", null);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], DropdownComponent.prototype, "closeOnOutsideClick", null);
    __decorate([
        Input(),
        __metadata("design:type", Boolean),
        __metadata("design:paramtypes", [Boolean])
    ], DropdownComponent.prototype, "closeOnMouseLeave", null);
    __decorate([
        ContentChild(DropdownToggleDirective),
        __metadata("design:type", DropdownToggleDirective)
    ], DropdownComponent.prototype, "dropdownToggle", void 0);
    __decorate([
        ContentChild(DropdownMenuDirective),
        __metadata("design:type", DropdownMenuDirective)
    ], DropdownComponent.prototype, "dropdownMenu", void 0);
    __decorate([
        HostListener('mouseleave'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DropdownComponent.prototype, "onMouseLeave", null);
    __decorate([
        HostListener('mouseenter'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DropdownComponent.prototype, "onMouseEnter", null);
    DropdownComponent = __decorate([
        Component({
            exportAs: 'ngxDropdown',
            selector: 'ngx-dropdown',
            template: "\n    <ng-content></ng-content>\n  ",
            host: {
                class: 'ngx-dropdown',
                '[class.open]': 'open',
                '[class.has-caret]': 'showCaret'
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-dropdown{position:relative;padding-bottom:10px}.ngx-dropdown.open .ngx-dropdown-menu{display:block;opacity:1;-webkit-animation:.25s openAnimation;animation:.25s openAnimation}.ngx-dropdown.has-caret .ngx-dropdown-menu:before{display:block}.ngx-dropdown .ngx-dropdown-toggle{cursor:pointer;display:inline-block}.ngx-dropdown .ngx-dropdown-toggle.disabled,.ngx-dropdown .ngx-dropdown-toggle.disabled>a,.ngx-dropdown .ngx-dropdown-toggle.disabled>button{cursor:not-allowed}.ngx-dropdown .ngx-dropdown-menu{position:absolute;clear:left;top:100%;left:0;z-index:1000;min-width:180px;background:#313847;box-shadow:0 1px 8px 0 rgba(0,0,0,.2),0 3px 4px 0 rgba(0,0,0,.14),0 3px 3px -2px rgba(0,0,0,.12);margin-top:0;opacity:0;display:none;-webkit-transition:visibility,opacity .25s ease-out;transition:visibility,opacity .25s ease-out}.ngx-dropdown .ngx-dropdown-menu::before{z-index:-1;content:' ';display:none;position:absolute;left:10px;top:-6px;width:12px;height:12px;background:#313847;-webkit-transform:rotate(45deg);transform:rotate(45deg);border-bottom:0;border-right:0}.ngx-dropdown .ngx-dropdown-menu.align-right{left:auto;right:0}.ngx-dropdown .ngx-dropdown-menu.align-right::before{left:auto;right:6px}.ngx-dropdown .ngx-dropdown-menu>ul>li>a,.ngx-dropdown .ngx-dropdown-menu>ul>li>button{padding:5px 20px;clear:both;line-height:1.42857143;white-space:nowrap;width:100%;text-align:left;background:0 0;box-shadow:none;color:#cfcfcf;font-size:.9em;display:block}.ngx-dropdown .ngx-dropdown-menu>ul>li:hover{background:#262c38}@-webkit-keyframes openAnimation{0%{-webkit-transform:translateY(-25px);transform:translateY(-25px)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}@keyframes openAnimation{0%{-webkit-transform:translateY(-25px);transform:translateY(-25px)}100%{-webkit-transform:translateY(0);transform:translateY(0)}}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline{background:#262c38;min-width:130px;font-size:13px;border:1px solid #505c75}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline:before{border-top:1px solid #505c75;border-left:1px solid #505c75;background:#262c38;z-index:10;top:-7px;-webkit-transform:translate(-1.8px,0) rotate(45deg);transform:translate(-1.8px,0) rotate(45deg)}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul{padding:9.5px 0;margin:2px 0 0;list-style:none}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li{padding-top:0!important;color:#fff}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li>a,.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li>button{cursor:pointer;height:26px;display:block;clear:both;font-weight:400;line-height:1.42857;white-space:nowrap;font-size:13px;color:#fff}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li:hover{background:#1f242f}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li.active{background:#479eff;color:#fff}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li.active>a,.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li.active>button{color:#fff}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-light-outline>ul>li.divider-label{height:initial;margin:8px 0;padding:0 20px;overflow:hidden;background-color:initial;color:#72819f}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline{background:#222732;min-width:130px;font-size:13px;border:1px solid #505c75}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline:before{border-top:1px solid #505c75;border-left:1px solid #505c75;background:#222732;z-index:10;top:-7px;-webkit-transform:translate(-1.8px,0) rotate(45deg);transform:translate(-1.8px,0) rotate(45deg)}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul{padding:9.5px 0;margin:2px 0 0;list-style:none}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li{padding-top:0!important;color:#afb7c8}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li>a,.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li>button{cursor:pointer;height:26px;display:block;clear:both;font-weight:400;line-height:1.42857;white-space:nowrap;font-size:13px;color:#afb7c8}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li:hover{background:#262c38}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li.active{background:#479eff;color:#afb7c8}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li.active>a,.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li.active>button{color:#fff}.ngx-dropdown .ngx-dropdown-menu.ngx-dropdown-dark-outline>ul>li.divider-label{height:initial;margin:8px 0;padding:0 20px;overflow:hidden;background-color:initial;color:#72819f}"]
        }),
        __metadata("design:paramtypes", [Renderer2, ChangeDetectorRef])
    ], DropdownComponent);
    return DropdownComponent;
}());
export { DropdownComponent };
