// @swimlane/ngx-ui/esm5/lib/directives/dbl-click-copy/dbl-click-copy.directive.js
import { __decorate, __metadata } from "tslib";
import { Directive, Output, EventEmitter, HostListener, ElementRef } from '@angular/core';
// tslint:disable-next-line:directive-selector
var DblClickCopyDirective = /** @class */ (function () {
    function DblClickCopyDirective(element) {
        this.element = element;
        this.onCopy = new EventEmitter();
    }
    DblClickCopyDirective.prototype.onDblClick = function () {
        var selection = getSelection();
        var range = document.createRange();
        range.selectNodeContents(this.element.nativeElement);
        selection.removeAllRanges();
        selection.addRange(range);
        document.execCommand('copy');
        this.onCopy.emit(range);
    };
    DblClickCopyDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], DblClickCopyDirective.prototype, "onCopy", void 0);
    __decorate([
        HostListener('dblclick'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", void 0)
    ], DblClickCopyDirective.prototype, "onDblClick", null);
    DblClickCopyDirective = __decorate([
        Directive({
            selector: '[dbl-click-copy]',
            host: {
                title: 'Double click to copy to clipboard'
            }
        }),
        __metadata("design:paramtypes", [ElementRef])
    ], DblClickCopyDirective);
    return DblClickCopyDirective;
}());
export { DblClickCopyDirective };
