// @swimlane/ngx-ui/esm5/lib/components/nag/nag.component.js
import { __decorate, __metadata } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, Input, Output, ViewEncapsulation } from '@angular/core';
import { coerceNumberProperty } from '@angular/cdk/coercion';
import { trigger } from '@angular/animations';
import { NAG_DRAWER_ANIMATION } from './nag.animation';
var NagComponent = /** @class */ (function () {
    function NagComponent() {
        this.cssClass = '';
        this.state = 'closed';
        this.stateChanged = new EventEmitter();
        this.nagTitle = '';
    }
    Object.defineProperty(NagComponent.prototype, "zIndex", {
        get: function () {
            return this._zIndex;
        },
        set: function (val) {
            this._zIndex = coerceNumberProperty(val);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(NagComponent.prototype, "klass", {
        get: function () {
            return "ngx-nag ngx-nag-bottom ngx-nag-" + this.state + " " + this.cssClass;
        },
        enumerable: true,
        configurable: true
    });
    NagComponent.prototype.toggle = function () {
        this.state = this.state !== 'open' ? 'open' : 'closed';
        this.stateChanged.emit(this.state);
    };
    NagComponent.prototype.ngOnDestroy = function () {
        this.stateChanged.emit(this.state);
    };
    NagComponent.prototype.ngOnChanges = function (changes) {
        var _this = this;
        if (changes.watch && this.state === 'closed') {
            this.state = 'peek';
            setTimeout(function () {
                _this.state = 'closed';
            }, 100);
        }
    };
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NagComponent.prototype, "cssClass", void 0);
    __decorate([
        HostBinding('@drawerTransition'),
        Input(),
        __metadata("design:type", String)
    ], NagComponent.prototype, "state", void 0);
    __decorate([
        Output(),
        __metadata("design:type", Object)
    ], NagComponent.prototype, "stateChanged", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Number),
        __metadata("design:paramtypes", [Number])
    ], NagComponent.prototype, "zIndex", null);
    __decorate([
        Input(),
        __metadata("design:type", String)
    ], NagComponent.prototype, "nagTitle", void 0);
    __decorate([
        Input(),
        __metadata("design:type", Object)
    ], NagComponent.prototype, "watch", void 0);
    __decorate([
        HostBinding('class'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], NagComponent.prototype, "klass", null);
    NagComponent = __decorate([
        Component({
            selector: 'ngx-nag',
            exportAs: 'ngxNag',
            template: "<div class=\"ngx-nag-content\">\n  <ngx-toolbar class=\"ngx-nag-toolbar\" (click)=\"toggle()\" [mainTitle]=\"nagTitle\">\n    <ngx-toolbar-title *ngIf=\"!nagTitle\"> <ng-content select=\"[ngx-nag-title]\"></ng-content> </ngx-toolbar-title>\n    <ngx-toolbar-content><ngx-icon class=\"ngx-nag-icon\" fontIcon=\"arrow-down\"></ngx-icon></ngx-toolbar-content>\n  </ngx-toolbar>\n  <section class=\"ngx-nag-body ngx-section-content\"><ng-content></ng-content></section>\n</div>\n",
            host: {
                role: 'dialog',
                tabindex: '-1',
                '[style.zIndex]': 'zIndex'
            },
            encapsulation: ViewEncapsulation.None,
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [trigger('drawerTransition', NAG_DRAWER_ANIMATION)],
            styles: [".ngx-nag{position:fixed;overflow-y:auto;overflow-x:hidden;text-align:left;z-index:998;background:#313847}.ngx-nag.ngx-nag-bottom{left:20%;right:20%;top:100%;width:60%}.ngx-nag.ngx-nag-bottom .ngx-nag-icon{-webkit-transform:rotateZ(180deg);transform:rotateZ(180deg);display:block;-webkit-transition:.4s ease-out;transition:.4s ease-out;background-color:#212633;height:30px;width:30px;border-radius:15px;float:right;text-align:center;margin-top:10px;line-height:34px;cursor:pointer}.ngx-nag.ngx-nag-bottom.ngx-nag-open .ngx-nag-icon{-webkit-transform:rotateZ(0);transform:rotateZ(0)}.ngx-nag .ngx-nag-body{background-color:#313847;margin-bottom:0;padding:1.8em;max-height:50vh;overflow-y:auto}.ngx-nag .ngx-nag-toolbar{background:#293040;cursor:pointer}"]
        })
    ], NagComponent);
    return NagComponent;
}());
export { NagComponent };
