// @swimlane/ngx-ui/esm2015/lib/components/checkbox/checkbox.component.js
import { __decorate, __metadata } from "tslib";
import { Component, Input, EventEmitter, Output, forwardRef, ViewEncapsulation, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { coerceNumberProperty, coerceBooleanProperty } from '@angular/cdk/coercion';
const CHKBOX_VALUE_ACCESSOR = {
    provide: NG_VALUE_ACCESSOR,
    useExisting: forwardRef(() => CheckboxComponent),
    multi: true
};
let nextId = 0;
let CheckboxComponent = class CheckboxComponent {
    constructor(cdr) {
        this.cdr = cdr;
        this.id = `checkbox-${++nextId}`;
        this.diameter = '18px';
        this.change = new EventEmitter();
        this.blur = new EventEmitter();
        this.focus = new EventEmitter();
        this._value = false;
        this._tabindex = 0;
        this._disabled = false;
        this._round = false;
        this.onTouchedCallback = () => {
            // placeholder
        };
        this.onChangeCallback = (_) => {
            // placeholder
        };
    }
    get tabindex() {
        return this._tabindex;
    }
    set tabindex(v) {
        this._tabindex = coerceNumberProperty(v);
        this.cdr.markForCheck();
    }
    get disabled() {
        return this._disabled;
    }
    set disabled(v) {
        this._disabled = coerceBooleanProperty(v);
        this.cdr.markForCheck();
    }
    get round() {
        return this._round;
    }
    set round(v) {
        this._round = coerceBooleanProperty(v);
        this.cdr.markForCheck();
    }
    set value(value) {
        if (this._value !== value) {
            this._value = value;
            this.cdr.markForCheck();
            this.onChangeCallback(this._value);
        }
    }
    get value() {
        return this._value;
    }
    onBlur(_) {
        this.onTouchedCallback();
    }
    toggle() {
        this.value = !this.value;
    }
    writeValue(value) {
        this.value = value;
    }
    registerOnChange(fn) {
        this.onChangeCallback = fn;
    }
    registerOnTouched(fn) {
        this.onTouchedCallback = fn;
    }
    setDisabledState(disabled) {
        this.disabled = disabled;
    }
};
CheckboxComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
__decorate([
    Input(),
    __metadata("design:type", Object)
], CheckboxComponent.prototype, "id", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], CheckboxComponent.prototype, "name", void 0);
__decorate([
    Input(),
    __metadata("design:type", String)
], CheckboxComponent.prototype, "diameter", void 0);
__decorate([
    Input(),
    __metadata("design:type", Number),
    __metadata("design:paramtypes", [Number])
], CheckboxComponent.prototype, "tabindex", null);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], CheckboxComponent.prototype, "disabled", null);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], CheckboxComponent.prototype, "round", null);
__decorate([
    Output(),
    __metadata("design:type", Object)
], CheckboxComponent.prototype, "change", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], CheckboxComponent.prototype, "blur", void 0);
__decorate([
    Output(),
    __metadata("design:type", Object)
], CheckboxComponent.prototype, "focus", void 0);
CheckboxComponent = __decorate([
    Component({
        selector: 'ngx-checkbox',
        exportAs: 'ngxCheckbox',
        template: "<label class=\"ngx-checkbox--label\">\n  <input\n    type=\"checkbox\"\n    [id]=\"id + '-chk'\"\n    [(ngModel)]=\"value\"\n    [disabled]=\"disabled\"\n    [name]=\"name + '-chk'\"\n    [tabIndex]=\"tabindex\"\n    (focus)=\"focus.emit($event)\"\n    (blur)=\"blur.emit($event)\"\n    (change)=\"change.emit($event)\"\n  />\n\n  <div\n    class=\"ngx-checkbox--box\"\n    [class.checked]=\"value\"\n    [style.width]=\"diameter\"\n    [style.height]=\"diameter\"\n    [style.min-width]=\"diameter\"\n    [style.min-height]=\"diameter\"\n  ></div>\n\n  <div class=\"ngx-checkbox--content\">\n    <ng-content></ng-content>\n  </div>\n</label>\n",
        host: {
            class: 'ngx-checkbox',
            '[class.disabled]': 'disabled',
            '[class.round]': 'round',
            '(blur)': 'onBlur($event)'
        },
        providers: [CHKBOX_VALUE_ACCESSOR],
        encapsulation: ViewEncapsulation.None,
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [".ngx-checkbox{display:-webkit-box;display:flex}.ngx-checkbox.disabled *{cursor:not-allowed}.ngx-checkbox.round .ngx-checkbox--box{border-radius:100%!important}.ngx-checkbox .ngx-checkbox--label{display:-webkit-box;display:flex;cursor:pointer;margin-bottom:0}.ngx-checkbox .ngx-checkbox--label input{display:none}.ngx-checkbox .ngx-checkbox--label .ngx-checkbox--box{border-radius:2px;background-color:transparent;border:2px solid #455066;-webkit-transition:-webkit-transform .4s cubic-bezier(.45,1.8,.5,.75);transition:transform .4s cubic-bezier(.45,1.8,.5,.75);transition:transform .4s cubic-bezier(.45,1.8,.5,.75),-webkit-transform .4s cubic-bezier(.45,1.8,.5,.75);-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;margin:auto 10px auto 0}.ngx-checkbox .ngx-checkbox--label .ngx-checkbox--box:after{position:absolute;top:calc(50% - 7px);left:calc(50% - 2px);width:6px;height:12px;content:'';border:solid #fff;border-width:0 2px 2px 0;-webkit-transform:rotate(0) scale(0);transform:rotate(0) scale(0);-webkit-transition:.4s cubic-bezier(.45,1.8,.5,.75);transition:.4s cubic-bezier(.45,1.8,.5,.75)}.ngx-checkbox .ngx-checkbox--label .ngx-checkbox--box.checked{background-color:#479eff;border-radius:2px;opacity:1;border:2px solid #479eff;-webkit-transform:rotate(0) scale(1);transform:rotate(0) scale(1)}.ngx-checkbox .ngx-checkbox--label .ngx-checkbox--box.checked:after{-webkit-transform:rotate(45deg) scale(1);transform:rotate(45deg) scale(1)}.ngx-checkbox .ngx-checkbox--label .ngx-checkbox--content{margin:auto 0}"]
    }),
    __metadata("design:paramtypes", [ChangeDetectorRef])
], CheckboxComponent);
export { CheckboxComponent };
