// @swimlane/ngx-ui/esm2015/lib/components/dialog/dialog.service.js
import { __decorate, __metadata } from "tslib";
import { Injectable } from '@angular/core';
import { InjectionService } from '../../services/injection/injection.service';
import { InjectionRegistryService } from '../../services/injection-registry/injection-registry.service';
import { OverlayService } from '../overlay/overlay.service';
import { DialogComponent } from './dialog.component';
let DialogService = class DialogService extends InjectionRegistryService {
    constructor(injectionService, overlayService) {
        super(injectionService);
        this.injectionService = injectionService;
        this.overlayService = overlayService;
        this.defaults = {
            inputs: {
                zIndex: 991,
                closeOnBlur: true,
                closeOnEscape: true,
                closeButton: true,
                showOverlay: true,
                visible: true
            }
        };
        this.type = DialogComponent;
        this.zIndex = 995;
    }
    create(options) {
        const component = super.create(options);
        this.createSubscriptions(component);
        return component;
    }
    destroy(component) {
        const hasOverlay = component.instance.showOverlay;
        this.zIndex = this.zIndex - 2;
        if (hasOverlay) {
            this.overlayService.removeTriggerComponent(component);
        }
        setTimeout(() => {
            super.destroy(component);
        });
    }
    createSubscriptions(triggerComponent) {
        let closeSub;
        let overlaySub;
        const kill = (c) => {
            /* istanbul ignore if */
            if (c !== triggerComponent) {
                return;
            }
            closeSub.unsubscribe();
            if (overlaySub)
                overlaySub.unsubscribe();
            this.destroy(triggerComponent);
        };
        closeSub = triggerComponent.instance.close.subscribe(kill.bind(this, triggerComponent));
        const zIndex = this.zIndex;
        if (triggerComponent.instance.showOverlay) {
            setTimeout(() => {
                this.overlayService.show({
                    triggerComponent,
                    zIndex
                });
                if (triggerComponent.instance.closeOnBlur) {
                    overlaySub = this.overlayService.click.subscribe(kill);
                }
            });
        }
    }
    assignDefaults(options) {
        options = super.assignDefaults(options);
        /* istanbul ignore else */
        if (!options.zIndex) {
            this.zIndex = this.overlayService.instance ? this.overlayService.instance.zIndex + 3 : this.zIndex + 2;
            options.inputs.zIndex = this.zIndex;
        }
        return options;
    }
};
DialogService.ctorParameters = () => [
    { type: InjectionService },
    { type: OverlayService }
];
DialogService = __decorate([
    Injectable(),
    __metadata("design:paramtypes", [InjectionService, OverlayService])
], DialogService);
export { DialogService };
